% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/all_generics.R, R/methods-bcbio.R
\docType{methods}
\name{bcbio}
\alias{bcbio}
\alias{bcbio<-}
\alias{bcbio,bcbioRNADataSet-method}
\alias{bcbio<-,bcbioRNADataSet-method}
\title{bcbioRNADataSet object accessors}
\usage{
\S4method{bcbio}{bcbioRNADataSet}(object, type = "counts")

\S4method{bcbio}{bcbioRNADataSet}(object, type = "counts") <- value
}
\arguments{
\item{object}{Object.}

\item{...}{Additional arguments.}

\item{value}{An integer matrix or other object.}

\item{type}{Type of count data to retrieve.}
}
\value{
Count matrix.
}
\description{
This method is used to access count matrices with differing dimensions from
the object. By default, we store quasi-aligned (a.k.a. lightweight) gene
expression counts generated by
\href{https://combine-lab.github.io/salmon/}{salmon} into in a
\link[SummarizedExperiment:SummarizedExperiment]{SummarizedExperiment::SummarizedExperiment} using \code{\link[=tximport]{tximport()}}.
}
\details{
Any additional transformations of those original counts, including \code{\link[=tpm]{tpm()}},
\code{\link[=rlog]{rlog()}}, or \code{\link[=tmm]{tmm()}} are stored alongside the raw counts. Reads be accessed
using the \code{counts()} generic.

Additional count matrices of interest include:
\itemize{
\item Aligned counts generated by \href{https://github.com/alexdobin/STAR}{STAR} and
\href{http://bioinf.wehi.edu.au/featureCounts/}{featureCounts}.
\item Transcript-level gene expression, for
\href{http://pachterlab.github.io/sleuth}{sleuth} analysis.
\item miRNA expression counts.
}

Since these matrices don't necessarily contain the same number of columns and
rows as the counts imported with \code{\link[=tximport]{tximport()}}, it is preferable to store this
information as alternative callers.
}
