% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/loadRNASeq.R
\name{loadRNASeq}
\alias{loadRNASeq}
\title{Load bcbio RNA-Seq Data}
\usage{
loadRNASeq(uploadDir, level = c("genes", "transcripts"),
  caller = c("salmon", "kallisto", "sailfish"),
  interestingGroups = "sampleName", samples = NULL,
  sampleMetadataFile = NULL, rowRanges = TRUE, isSpike = NULL,
  organism = NULL, ensemblRelease = NULL, genomeBuild = NULL,
  design = formula(~1), transformationLimit = 50L, ...)
}
\arguments{
\item{uploadDir}{Path to final upload directory. This path is set when
running \code{bcbio_nextgen -w template}.}

\item{level}{Import counts as "\code{genes}" (default) or "\code{transcripts}".}

\item{caller}{Expression caller. Supports "\code{salmon}" (default), "\code{kallisto}",
or "\code{sailfish}".}

\item{interestingGroups}{Character vector of interesting groups. First entry
is used for plot colors during quality control (QC) analysis. Entire vector
is used for PCA and heatmap QC functions.}

\item{samples}{\emph{Optional.} Specify a subset of samples to load. The names
must match the \code{description} specified in the bcbio YAML metadata. If a
\code{sampleMetadataFile} is provided, that will take priority for sample
selection. Typically this can be left \code{NULL}.}

\item{sampleMetadataFile}{\emph{Optional.} Custom metadata file containing
sample information. Otherwise defaults to sample metadata saved in the YAML
file. Remote URLs are supported. Typically this can be left \code{NULL}.}

\item{rowRanges}{\emph{Recommended.} User-defined genomic ranges (\code{GRanges}) that
match the rownames of the counts matrix. If set \code{TRUE} (default), the
annotations will be obtained automatically fron Ensembl using AnnotationHub
and ensembleb. These values are also accessible with the \code{\link[=rowData]{rowData()}}
function. If set \code{NULL}, then \code{\link[=rowRanges]{rowRanges()}} inside the resulting
\code{bcbioRNASeq} object will be left empty. This is recommended for projects
dealing with genes or transcripts that are poorly annotated.}

\item{isSpike}{Genes or transcripts corresponding to FASTA spike-in
sequences (e.g. ERCCs, EGFP, TDTOMATO).}

\item{organism}{\emph{Optional.} Organism name. Use the full latin name (e.g.
"Homo sapiens"). Normally this can be left \code{NULL}, and the function will
detect the organism from the gene identifiers present in the rownames of
the counts matrix using \code{\link[=detectOrganism]{detectOrganism()}}. Alternatively, automatic
organism detection can be overriden here, and the argument will be passed
to AnnotationHub.}

\item{ensemblRelease}{\emph{Optional.} Ensembl release version. If \code{NULL},
defaults to current release, and does not typically need to be
user-defined. This parameter can be useful for matching Ensembl annotations
against an outdated bcbio annotation build.}

\item{genomeBuild}{\emph{Optional.} Genome build. Normally this can be left \code{NULL}
and the build will be detected from the bcbio run data. This can be set
manually (e.g. "GRCh37" for the older \emph{Homo sapiens} reference genome).
Note that this must match the genome build identifier on Ensembl for
annotations to download correctly.}

\item{design}{DESeq2 design formula. Empty by default. Can be updated after
initial data loading using the \code{\link[=design]{design()}} function.}

\item{transformationLimit}{Maximum number of samples to calculate
\code{\link[DESeq2:rlog]{DESeq2::rlog()}} and \code{\link[DESeq2:varianceStabilizingTransformation]{DESeq2::varianceStabilizingTransformation()}} matrix.
It is not generally recommended to change this value. For large datasets,
DESeq2 will take a really long time applying variance stabilization. See
Details. Use \code{Inf} to always apply transformations and \code{0} to always skip.}

\item{...}{Additional arguments, slotted into the \code{\link[=metadata]{metadata()}} accessor.}
}
\value{
\code{bcbioRNASeq}.
}
\description{
Simply point to the final upload directory output by
\href{https://bcbio-nextgen.readthedocs.io/}{bcbio}, and this function will take
care of the rest. It automatically imports RNA-seq counts, metadata, and the
program versions used.
}
\details{
When the number of samples is bigger than the \code{transformationLimit}, \code{rlog}
and \code{vst} counts will not be slotted into \code{assays()}. In this case, we
recommend visualization using \code{tmm} counts, which are automatically
calculated by \code{\link[edgeR:tmm]{edgeR::tmm()}}.
}
\note{
When working in RStudio, we recommend connecting to the bcbio-nextgen
run directory as a remote connection over
\href{https://github.com/osxfuse/osxfuse/wiki/SSHFS}{sshfs}.
}
\examples{
uploadDir <- system.file("extdata/bcbio", package = "bcbioRNASeq")

# Gene level
loadRNASeq(uploadDir, level = "genes")

# Transcript level
loadRNASeq(uploadDir, level = "transcripts")
}
\seealso{
\code{help("bcbioRNASeq-class", "bcbioRNASeq")}.
}
\author{
Michael Steinbaugh, Lorena Pantano
}
