# Methods

RNA-seq counts were generated by [bcbio][] and [bcbioRNASeq][] [@Steinbaugh2018-rc] using [Salmon][] [@Patro2017-pi]. Counts were imported into [R][] using [tximport][] [@Soneson2016-tu] and [DESeq2][] [@Love2014-sq]. Gene annotations were obtained from [Ensembl][]. Plots were generated by [ggplot2][] [@Wickham2009-si]. Heatmaps were generated by [pheatmap][] [@Kolde2015-dy].

# R session information {.tabset}

```{r session_info}
markdownHeader("YAML params", level = 2)
print(params)

markdownHeader("`_setup.R`", level = 2)
readr::read_lines("_setup.R") %>% writeLines()

markdownHeader("`sessioninfo::session_info()`", level = 2)
sessioninfo::session_info()

markdownHeader("`utils::sessionInfo()`", level = 2)
utils::sessionInfo()
```

[bcbio]: https://github.com/bcbio/bcbio-nextgen/
[bcbioRNASeq]: https://bioinformatics.sph.harvard.edu/bcbioRNASeq/
[Bioconductor]: https://bioconductor.org/
[clusterProfiler]: https://bioconductor.org/packages/clusterProfiler/
[DESeq2]: https://bioconductor.org/packages/DESeq2/
[Ensembl]: https://www.ensembl.org/
[Entrez]: https://www.ncbi.nlm.nih.gov/search/
[Excel]: https://products.office.com/en-us/excel/
[fgsea]: http://bioconductor.org/packages/fgsea/
[ggplot2]: https://ggplot2.tidyverse.org/
[GO]: http://www.geneontology.org
[GSEA]: http://software.broadinstitute.org/gsea/
[KEGG]: https://www.genome.jp/kegg/
[LibreOffice]: https://www.libreoffice.org/
[pathview]: https://bioconductor.org/packages/pathview/
[pheatmap]: https://cran.r-project.org/package=pheatmap
[R]: https://www.r-project.org/
[RStudio]: https://www.rstudio.com/
[salmon]: https://combine-lab.github.io/salmon/
[tximport]: https://bioconductor.org/packages/tximport/
