<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Version 0.2 • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Version 0.2" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioRNASeq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bcbioRNASeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1>Changelog <small>0.2</small></h1>
      
    </div>

    <div id="bcbiornaseq-0-2-0-2018-03-22" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-2-0-2018-03-22" class="anchor"></a>bcbioRNASeq 0.2.0 (2018-03-22)</h2>
<div id="major-changes-6" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-6" class="anchor"></a>Major changes</h3>
<ul>
<li>
<code>bcbioRNASeq</code> S4 class object is now extending <code>RangedSummarizedExperiment</code> instead of <code>SummarizedExperiment</code>. Consequently, the row annotations are now stored in the <code>rowRanges</code> slot as <code>GRanges</code> class, instead of in the <code>rowData</code> slot as a <code>DataFrame</code>. The <code>rowData()</code> accessor still works and returns a data frame of gene/transcript annotations, but these are now coerced from the internally stored <code>GRanges</code>. The <code>GRanges</code> object is acquired automatically from <a href="https://www.ensembl.org/">Ensembl</a> using <code>basejump::ensembl()</code>. By default, <code>GRanges</code> are acquired from <a href="https://www.ensembl.org/">Ensembl</a> using <a href="https://bioconductor.org/packages/AnnotationHub/">AnnotationHub</a> and <a href="https://bioconductor.org/packages/ensembldb/">ensembldb</a>. Legacy GRCh37 genome build is supported using the <a href="https://bioconductor.org/packages/EnsDb.Hsapiens.v75/">EnsDb.Hsapiens.v75</a> package.</li>
<li>
<code>assays()</code> now only slot matrices. We’ve moved the <a href="https://bioconductor.org/packages/tximport/">tximport</a> data from the now defunct <code>bcbio()</code> slot to assays. This includes the <code>lengths</code> matrix from <a href="https://bioconductor.org/packages/tximport/">tximport</a>. Additionally, we are optionally slotting <a href="https://bioconductor.org/packages/DESeq2/">DESeq2</a> variance-stabilized counts (“<code>rlog</code>”, <code>"vst"</code>). <a href="https://bioconductor.org/packages/DESeq2/">DESeq2</a> normalized counts and <a href="https://bioconductor.org/packages/edgeR/">edgeR</a> TMM counts are calculated on the fly and no longer stored inside the <code>bcbioRNASeq</code> object.</li>
<li>
<code>colData()</code> now defaults to returning as <code>data.frame</code> instead of <code>DataFrame</code>, for easy piping to <a href="https://www.tidyverse.org/">tidyverse</a> functions.</li>
<li>
<code>bcbio()</code> slot is now defunct.</li>
<li>FASTA spike-ins (e.g. EGFP, ERCCs) can be defined using the <code>isSpike</code> argument during the <code><a href="../reference/deprecated.html">loadRNASeq()</a></code> data import step.</li>
<li>Melted counts are now scaled to log2 in the relevant quality control functions rather than using log10. This applies to <code><a href="../reference/plotCountsPerGene.html">plotCountsPerGene()</a></code> and <code><a href="../reference/plotCountDensity.html">plotCountDensity()</a></code>. Note that we are subsetting the nonzero genes as defined by the raw counts here.</li>
<li>Simplified internal <code>tximport()</code> code to no longer attempt to strip transcript versions. This is required for working with <em>C. elegans</em> transcripts.</li>
<li>Minimal working example dataset is now derived from GSE65267, which is also used in the F1000 paper.</li>
<li>Added <code><a href="../reference/coerce.html">as(object, "DESeqDataSet")</a></code> coercion method support for <code>bcbioRNASeq</code> class. This helps us set up the differential expression analysis easily.</li>
<li>
<code><a href="../reference/counts.html">counts()</a></code> function now returns <a href="https://bioconductor.org/packages/DESeq2/">DESeq2</a> normalized counts (<code>normalized =   TRUE</code>) and <a href="https://bioconductor.org/packages/edgeR/">edgeR</a> TMM counts (<code>normalized = "tmm"</code>) on the fly, as suggested by the F1000 reviewers.</li>
<li>Design formula can no longer be slotted into <code>bcbioRNASeq</code> object, since we’re not stashing a <code>DESeqDataSet</code> any more.</li>
<li>Updated Functional Analysis R Markdown template.</li>
</ul>
</div>
<div id="minor-changes-8" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-8" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>validObject()</code> is now required for all plotting functions. This check is also called in the R Markdown template. Legacy objects can be updated using <code><a href="../reference/updateObject.html">updateObject()</a></code>.</li>
<li>
<code><a href="../reference/metrics.html">metrics()</a></code> now returns columns sorted alphabetically.</li>
<li>Added <code><a href="../reference/contrastName.html">contrastName()</a></code> as a generic function.</li>
<li>
<code><a href="../reference/plotDEGHeatmap.html">plotDEGHeatmap()</a></code> and <code><a href="../reference/plotDEGPCA.html">plotDEGPCA()</a></code> generics no longer have <code>counts</code> defined in the signature. The <code>counts</code> argument is now only defined in the methods.</li>
<li>
<code><a href="../reference/deprecated.html">prepareRNASeqTemplate()</a></code> has been converted from a generic to a standard function.</li>
<li>Improved <code>metadata()</code> validity checks.</li>
<li>
<code><a href="../reference/plotCorrelationHeatmap.html">plotCorrelationHeatmap()</a></code> matrix method has been moved to <a href="https://basejump.acidgenomics.com/">basejump</a> package, for improved consistency with the other heatmap code.</li>
<li>
<code><a href="../reference/plotGenderMarkers.html">plotGenderMarkers()</a></code> internal code has been reworked to match <code>plotGene()</code>.</li>
<li>Default <code><a href="../reference/plotMA.html">plotMA()</a></code> appearance has changed, providing a line at the 0 y-intercept, similar to <code>DESeqDataSet</code> method.</li>
<li>Internal example datasets have been renamed (e.g. <code>bcb_small</code> instead of <code>bcb</code>).</li>
<li>Added AppVeyor CI support for code testing on Windows.</li>
<li>Made Travis CI checks stricter, added <code>BiocCheck()</code>.</li>
<li>Internal <code>.sampleDirs()</code> code is now exported in <a href="https://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a> as a generic.</li>
<li>
<code>gene2symbol()</code> and <code>interestingGroups()</code> method support are now defined for <code>SummarizedExperiment</code> in the <a href="https://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a> package.</li>
</ul>
</div>
<div id="updating-legacy-objects-v0-2-0" class="section level3">
<h3 class="hasAnchor">
<a href="#updating-legacy-objects-v0-2-0" class="anchor"></a>Updating legacy objects &lt; v0.2.0</h3>
<ul>
<li>Use <code><a href="../reference/updateObject.html">updateObject()</a></code> in combination with the <code>rowRanges</code> argument, which requires a <code>GRanges</code> object. <code>GRanges</code> can be obtained from <a href="https://www.ensembl.org/">Ensembl</a> using the <code>basejump::ensembl()</code> function or the <a href="https://bioconductor.org/packages/ensembldb/">ensembldb</a> package.</li>
</ul>
</div>
<div id="deprecations" class="section level3">
<h3 class="hasAnchor">
<a href="#deprecations" class="anchor"></a>Deprecations</h3>
<ul>
<li>
<code>bcbio()</code> slot is now defunct, since we have moved all data into the <code>SummarizedExperiment</code> container.</li>
<li>Deprecated <code>plot5x3Bias()</code> in favor of <code><a href="../reference/plot5Prime3PrimeBias.html">plot5Prime3PrimeBias()</a></code>. This is less confusing as to what this function plots.</li>
<li>
<code>flatFiles()</code> has been deprecated in favor of <code><a href="../reference/coerce.html">as(object, "list")</a></code> coercion method. See <a href="https://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a> package for <code>SummarizedExperiment</code> method support.</li>
<li>Defunct: <code>design()</code>, <code>download()</code>, <code><a href="../reference/deprecated.html">meltLog10()</a></code>, <code>txi()</code>.</li>
<li>Legacy <code>bcbioRNADataSet</code> method support has been removed.</li>
</ul>
</div>
</div>
    <div id="bcbiornaseq-0-2-1-2018-04-24" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-2-1-2018-04-24" class="anchor"></a>bcbioRNASeq 0.2.1 (2018-04-24)</h2>
<p>Last set of code fixes before F1000v2 resubmission.</p>
<div id="major-changes-5" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-5" class="anchor"></a>Major changes</h3>
<ul>
<li>Added <code>rle</code> return support for <code><a href="../reference/counts.html">counts()</a></code>, which are calculated on the fly.</li>
<li>Added <code>transgeneNames</code> and <code>spikeNames</code> support to <code><a href="../reference/deprecated.html">loadRNASeq()</a></code> function.</li>
<li>
<code><a href="../reference/deprecated.html">loadRNASeq()</a></code> now supports <code>organism = NULL</code> again, for datasets with poorly annotated genomes.</li>
<li>Primary <code>assay()</code> containing raw counts is now named <code>counts</code> instead of <code>raw</code>, for consistency with other <code>SummarizedExperiment</code> objects (e.g. <code>DESeqDataSet</code>) and the <a href="https://bioinformatics.sph.harvard.edu/bcbioSingleCell/">bcbioSingleCell</a> S4 class definition.</li>
<li>Improved internal code for <code>plotGene()</code> and <code><a href="../reference/plotGenderMarkers.html">plotGenderMarkers()</a></code>.</li>
</ul>
</div>
<div id="minor-changes-7" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-7" class="anchor"></a>Minor changes</h3>
<ul>
<li>Improved <a href="https://ci.appveyor.com/">AppVeyor CI</a> support to test against bioc-devel using <a href="https://www.r-project.org/">R</a> 3.5.</li>
<li>Improved support and unit testing for <code><a href="../reference/updateObject.html">updateObject()</a></code> method.</li>
<li>
<a href="https://bioconductor.org/packages/DESeq2/">DESeq2</a> normalized counts are always slotted in <code>assays()</code>, even when rlog and vst transformations are skipped.</li>
<li>Exporting <code>[[&lt;-</code>, <code>assays&lt;-</code>, <code>colData&lt;-</code>, <code>interestingGroups&lt;-</code>, and <code>metadata&lt;-</code> assignment methods, to avoid unwanted coercion to <code>SummarizedExperiment</code>. Objects extending <code>RangedSummarizedExperiment</code> shouldn’t be doing this, so we may need to file a bug report with Bioconductor or check our class definition in the package.</li>
<li>Now importing specific functions from S4Vectors and methods rather than importing everything.</li>
<li>Switched back to using <code><a href="https://rdrr.io/r/base/stop.html">stop()</a></code>, <code><a href="https://rdrr.io/r/base/warning.html">warning()</a></code> and <code><a href="https://rdrr.io/r/base/message.html">message()</a></code> rather than the alternate <a href="https://cran.r-project.org/package=rlang">rlang</a> functions <code>abort()</code>, <code>warn()</code>, and <code>inform()</code>.</li>
<li>Objects with invalid metadata now print which slots are invalid to the console.</li>
</ul>
</div>
</div>
    <div id="bcbiornaseq-0-2-2-2018-04-26" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-2-2-2018-04-26" class="anchor"></a>bcbioRNASeq 0.2.2 (2018-04-26)</h2>
<div id="minor-changes-6" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-6" class="anchor"></a>Minor changes</h3>
<ul>
<li>Split out assertive imports so we can pin on <a href="https://bioconda.github.io/">bioconda</a>.</li>
<li>Improved package documentation.</li>
<li>Improved label consistency in <code><a href="../reference/plotPCA.html">plotPCA()</a></code> functions to match <code><a href="../reference/deprecated.html">plotMeanAverage()</a></code> and <code><a href="../reference/plotVolcano.html">plotVolcano()</a></code>.</li>
<li>Improved automatic title labeling in <code><a href="../reference/plotDEGPCA.html">plotDEGPCA()</a></code>, matching the other DEG functions. Also added directionality to <code><a href="../reference/plotDEGPCA.html">plotDEGPCA()</a></code>.</li>
<li>Added <code>DESeqDataSet</code> method support to <code><a href="../reference/plotCorrelationHeatmap.html">plotCorrelationHeatmap()</a></code>, using the normalized counts.</li>
<li>
<code>reusltsTables()</code> now writes local files to <code><a href="https://rdrr.io/r/base/tempfile.html">tempdir()</a></code> when <a href="https://dropbox.com/">Dropbox</a> mode is enabled using <code>dropboxDir</code>.</li>
</ul>
</div>
</div>
    <div id="bcbiornaseq-0-2-3-2018-05-10" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-2-3-2018-05-10" class="anchor"></a>bcbioRNASeq 0.2.3 (2018-05-10)</h2>
<div id="major-changes-4" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-4" class="anchor"></a>Major changes</h3>
<ul>
<li>Now recommending variance stabilizing transformation (<code>vst</code>) over <code>rlog</code> counts by default in plots, where applicable.</li>
</ul>
</div>
<div id="minor-changes-5" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-5" class="anchor"></a>Minor changes</h3>
<ul>
<li>Tweaked Rory’s biotype plots in QC report to match formatting conventions in the package. These plots are now colored.</li>
<li>Added <code><a href="../reference/plotDEGPCA.html">plotDEGPCA()</a></code> to default differential expression <a href="https://rmarkdown.rstudio.com/">R Markdown</a> template.</li>
<li>
<code>colData()</code> factors are correctly releveled upon object subset with <code>[</code>. This helps avoid unwanted downstream errors when creating a <code>DESeqDataSet</code> and running differential expression with <a href="https://bioconductor.org/packages/DESeq2/">DESeq2</a>.</li>
<li>Recommending <code>facet</code> return method by default for <code>plotGene()</code>. Updated the working example to reflect this.</li>
<li>
<code><a href="../reference/metrics.html">metrics()</a></code> now returns <code>interestingGroups</code> column.</li>
<li>
<code>sample</code> label has been removed from axis title for QC plot functions.</li>
<li>Now using shared <a href="https://ggplot2.tidyverse.org/">ggplot2</a> convneience functions from <a href="https://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a> 0.2.10: <code>bcbio_geom_abline()</code>, <code>bcbio_geom_label()</code>, and <code>bcbio_geom_label_repel()</code>. These are also used by <a href="https://bioinformatics.sph.harvard.edu/bcbioSingleCell/">bcbioSingleCell</a> for improved graphical consistency.</li>
<li>Removed unused internal legacy <a href="https://ggplot2.tidyverse.org/">ggplot2</a> code.</li>
<li>Increased <a href="https://bioconductor.org/packages/DEGreport/">DEGreport</a>, <a href="https://bioconductor.org/packages/DESeq2/">DESeq2</a>, and <a href="https://bioconductor.org/packages/tximport/">tximport</a> dependency requirements.</li>
</ul>
</div>
</div>
    <div id="bcbiornaseq-0-2-4-2018-05-24" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-2-4-2018-05-24" class="anchor"></a>bcbioRNASeq 0.2.4 (2018-05-24)</h2>
<div id="major-changes-3" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-3" class="anchor"></a>Major changes</h3>
<ul>
<li>
<code><a href="../reference/aggregateReplicates.html">aggregateReplicates()</a></code> support has been added back. This function returns a <code>RangedSummarizedExperiment</code> instead of a <code>bcbioRNASeq</code> object, containing only an aggregate raw counts matrix in the <code>counts</code> slot of <code>assays()</code>.</li>
<li>The functional analysis <a href="https://rmarkdown.rstudio.com/">R Markdown</a> template has been reworked to use <code>dds_file</code> and <code>organism</code> as new parameter arguments. We’ve reduced the number of parameters required here to run <a href="https://bioconductor.org/packages/clusterProfiler/">clusterProfiler</a>.</li>
<li>Made <code><a href="../reference/alphaSummary.html">alphaSummary()</a></code> defunct for <code>bcbioRNASeq</code> object, in favor of <code>DESeqDataSet</code> only. This function is only useful when a proper design formula has been defined.</li>
</ul>
</div>
<div id="minor-changes-4" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-4" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code><a href="../reference/metrics.html">metrics()</a></code> now contains an informative error for datasets that were analyzed using the <code>fast-rnaseq</code> <a href="https://github.com/chapmanb/bcbio-nextgen/">bcbio</a> pipeline.</li>
<li>
<code>DESeqDataSet</code> coercion from <code>bcbioRNASeq</code> object doesn’t attempt to run <code>DESeq()</code> command any more, which was unnecessary and improves speed.</li>
<li>
<code>bcbioSingleCell()</code> constructor now supports <code>censorSamples</code> parameter. This is useful for removing known poor quality samples upon loading.</li>
<li>
<a href="https://ggplot2.tidyverse.org/">ggplot2</a> color and fill palettes are now set <code>NULL</code> in the quality control functions. This behavior doesn’t change the appearance of the plot colors, which will still default to <code><a href="https://rdrr.io/pkg/ggplot2/man/scale_hue.html">ggplot2::scale_colour_hue()</a></code> or <code><a href="https://rdrr.io/pkg/ggplot2/man/scale_hue.html">ggplot2::scale_fill_hue()</a></code>. The upcoming <a href="https://ggplot2.tidyverse.org/">ggplot2</a> v2.3.0 update supports global options for color and fill palettes, so these parameters may be deprecated in a future release.</li>
<li>Reworked the internal code for <code><a href="../reference/topTables.html">topTables()</a></code>.</li>
</ul>
</div>
<div id="infrastructure-changes" class="section level3">
<h3 class="hasAnchor">
<a href="#infrastructure-changes" class="anchor"></a>Infrastructure changes</h3>
<ul>
<li>Added <a href="https://www.apple.com/macos/">macOS</a> testing to <a href="https://travis-ci.org/">Travis CI</a> build checks.</li>
<li>Fixed <a href="https://bioconductor.org/packages/clusterProfiler/">clusterProfiler</a> compilation error on <a href="https://travis-ci.org/">Travis CI</a> by installing <code>libudunits2-dev</code> (Linux).</li>
</ul>
</div>
</div>
    <div id="bcbiornaseq-0-2-5-2018-06-21" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-2-5-2018-06-21" class="anchor"></a>bcbioRNASeq 0.2.5 (2018-06-21)</h2>
<div id="minor-changes-3" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-3" class="anchor"></a>Minor changes</h3>
<ul>
<li>Enable support <a href="https://github.com/chapmanb/bcbio-nextgen/">bcbio</a> integration, by modifying <code><a href="../reference/bcbioRNASeq.html">bcbioRNASeq()</a></code> constructor to work with minimal bcbio test data.</li>
<li>Switched from internal usage of <code>aes_()</code> in favor of consistent usage of <code>aes_string()</code>. This will make the transition to <a href="https://ggplot2.tidyverse.org/">ggplot2</a> v2.3.0 easier in a future update.</li>
</ul>
</div>
</div>
    <div id="bcbiornaseq-0-2-6-2018-08-19" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-2-6-2018-08-19" class="anchor"></a>bcbioRNASeq 0.2.6 (2018-08-19)</h2>
<div id="major-changes-2" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-2" class="anchor"></a>Major changes</h3>
<ul>
<li>Added <code>transform = TRUE</code> argument to <code>[</code> extraction method, allowing the user to skip automatic <a href="https://bioconductor.org/packages/DESeq2/">DESeq2</a> transformations, which can be CPU intensive for large datasets.</li>
<li>Switched back to using <code><a href="../reference/plotMA.html">plotMA()</a></code> instead of <code><a href="../reference/deprecated.html">plotMeanAverage()</a></code>. An MA-plot by definition is not a “Mean Average” plot, so this function name is misleading. We will keep the <code><a href="../reference/deprecated.html">plotMeanAverage()</a></code> working but it is now soft deprecated.</li>
<li>
<code><a href="../reference/plotGeneSaturation.html">plotGeneSaturation()</a></code> now supports <code>label</code> argument, similar to <code><a href="../reference/plotPCA.html">plotPCA()</a></code>.</li>
</ul>
</div>
<div id="minor-changes-2" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-2" class="anchor"></a>Minor changes</h3>
<ul>
<li>Improved internal handling during <code>tximport()</code> call to handle transcript version mismatch with tx2gene data.frame. This can result if the bcbio pipeline is using an old genome build.</li>
<li>Ensure <code>genomeBuild</code> is detected from <a href="https://bioconductor.org/packages/AnnotationHub/">AnnotationHub</a> <code>rowRangesMetadata</code> if applicable, and not left <code>NULL</code> in the metadata.</li>
<li>Updated internal code to use <code>aes()</code> instead of <code>aes_string()</code>, which uses <a href="http://dplyr.tidyverse.org/articles/programming.html">tidyeval</a> and quasiquotation.</li>
<li>
<code>plotGene()</code>: reduced the number of <code>return</code> options to simply “facet” and “wide”. Previously, this also supported “grid”, “list”, and “markdown”, but these were removed because they are not frequently used.</li>
<li>
<code>plotGene()</code>: Switched back to internal <code><a href="https://rdrr.io/r/base/lapply.html">lapply()</a></code> call instead of using <code><a href="https://rdrr.io/pkg/BiocParallel/man/bplapply.html">BiocParallel::bplapply()</a></code>. This doesn’t always work perfect in an HPC environment (e.g. HMS O2 cluster).</li>
<li>Soft deprecated <code><a href="../reference/deprecated.html">plotMeanAverage()</a></code> in favor of <code><a href="../reference/plotMA.html">plotMA()</a></code>.</li>
<li>All S4 generics with method support are exported for easier lookup in the reference documentation.</li>
<li>All function reexports have been consolidated in <a href="https://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a> package.</li>
</ul>
</div>
</div>
    <div id="bcbiornaseq-0-2-7-2018-08-22" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-2-7-2018-08-22" class="anchor"></a>bcbioRNASeq 0.2.7 (2018-08-22)</h2>
<div id="minor-changes-1" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-1" class="anchor"></a>Minor changes</h3>
<ul>
<li>Bug fix for <code><a href="../reference/coerce.html">as()</a></code> coercion method support. Need to ensure <code>exportMethods(coerce)</code> is included in <code>NAMESPACE</code> file, otherwise <code>bcbioRNASeq</code> to <code>DESeqDataSet</code> coercion using <code><a href="../reference/coerce.html">as(bcb, "DESeqDataSet")</a></code> won’t work when called from an Rscript without the package library loaded. Thanks <a href='https://github.com/roryk'>@roryk</a> for noticing this.</li>
<li>Only run rRNA plots if rRNA is detected.</li>
</ul>
</div>
</div>
    <div id="bcbiornaseq-0-2-8-2019-01-28" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-2-8-2019-01-28" class="anchor"></a>bcbioRNASeq 0.2.8 (2019-01-28)</h2>
<ul>
<li>Handle correlation of PCA covariates when a numerical covariate has no variation.</li>
<li>Handle <code>NA</code> in ribosomal RNA calculations.</li>
</ul>
</div>
    <div id="bcbiornaseq-0-2-9-2019-05-03" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-2-9-2019-05-03" class="anchor"></a>bcbioRNASeq 0.2.9 (2019-05-03)</h2>
<div id="major-changes-1" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-1" class="anchor"></a>Major changes</h3>
<ul>
<li>Fixed a bug in internal tximport code (refer to <code>tximport-internal.R</code> file) where sample names can get associated with the wrong samples in experiments, causing samples with numbers roll over from 1 -&gt; 10. Thanks to Richard from AstraZeneca for catching this.</li>
</ul>
</div>
</div>
    <div id="bcbiornaseq-0-2-10-2019-05-12" class="section level2">
<h2 class="hasAnchor">
<a href="#bcbiornaseq-0-2-10-2019-05-12" class="anchor"></a>bcbioRNASeq 0.2.10 (2019-05-12)</h2>
<p>This is a maintenance release to support the latest release of basejump.</p>
<div id="updated-basejump-release-series-support" class="section level3">
<h3 class="hasAnchor">
<a href="#updated-basejump-release-series-support" class="anchor"></a>Updated basejump release series support</h3>
<p>This is a significant update to bcbioRNASeq prior to the upcoming v0.3 release series. This update provides a compatibility update for the <a href="https://basejump.acidgenomics.com/">basejump</a> v0.10 release series, whereas <a href="https://bioinformatics.sph.harvard.edu/bcbioRNASeq/">bcbioRNASeq</a> was previously pinned to basejump v0.7 for stability. No major changes to the code base have been applied here, and hopefully no breaking code changes are observed in this update. If you encounter any breaking changes, please file an issue on GitHub.</p>
</div>
<div id="major-changes" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes" class="anchor"></a>Major changes</h3>
<ul>
<li>Now importing bioverbs package for generics.</li>
<li>Use <code>plotCounts</code> instead of <code>plotGene</code>, which is now deprecated.</li>
<li>
<code>prepareRNASeqTemplate</code> is deprecated in favor of <code>prepareTemplate</code>.</li>
<li>
<code>sampleData</code> and <code>tpm</code> method support have been removed in favor of importing improved code from basejump.</li>
<li>Now using Docker images for Travis CI tests, since they are much faster.</li>
</ul>
</div>
<div id="minor-changes" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes" class="anchor"></a>Minor changes</h3>
<ul>
<li>All S4 methods have been split out into internal functions.</li>
<li>Moved bcbioRNASeq validity method into <code>setClass</code> call.</li>
<li>Miscellaneous documentation improvements. Function titles have been renamed into sentence case, matching Acid Genomics conventions.</li>
<li>Removed defunct/deprecated functions prior to v0.2 release.</li>
<li>Added periods to comments, where applicable.</li>
<li>Switched back to using <code>str</code> instead of <code>glimpse</code> in working examples, since <code>glimpse</code> requires tibble import.</li>
<li>Reworked some plots to inherit new <a href="https://acidplots.acidgenomics.com/">acidplots</a> code.</li>
<li>Files have been renamed in kebab/dash case instead of snake case where applicable. This applies mainly to the unit test files.</li>
</ul>
</div>
<div id="r-markdown-templates" class="section level3">
<h3 class="hasAnchor">
<a href="#r-markdown-templates" class="anchor"></a>R Markdown templates</h3>
<ul>
<li>The <code>_setup.R</code> file no longer includes <code>library</code> loads in the file. Instead these should be declared per R Markdown template, where appropriate. Alternatively, the user can modify <code>_setup.R</code> manually per project, but we are no longer recommending default library loads in this file.</li>
<li>Updated quality control template to use “object” instead of “bcb” as name.</li>
<li>Updated differential expression template to support loading of either rda or rds files, similar to QC template.</li>
<li>Updated functional analysis template. Note that GSEA ranked list now uses Wald test statistic instead of BH adjusted P value, which provides better directional ranks with fewer ties.</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
        <li><a href="#bcbiornaseq-0-2-0-2018-03-22">0.2.0</a></li>
        <li><a href="#bcbiornaseq-0-2-1-2018-04-24">0.2.1</a></li>
        <li><a href="#bcbiornaseq-0-2-2-2018-04-26">0.2.2</a></li>
        <li><a href="#bcbiornaseq-0-2-3-2018-05-10">0.2.3</a></li>
        <li><a href="#bcbiornaseq-0-2-4-2018-05-24">0.2.4</a></li>
        <li><a href="#bcbiornaseq-0-2-5-2018-06-21">0.2.5</a></li>
        <li><a href="#bcbiornaseq-0-2-6-2018-08-19">0.2.6</a></li>
        <li><a href="#bcbiornaseq-0-2-7-2018-08-22">0.2.7</a></li>
        <li><a href="#bcbiornaseq-0-2-8-2019-01-28">0.2.8</a></li>
        <li><a href="#bcbiornaseq-0-2-9-2019-05-03">0.2.9</a></li>
        <li><a href="#bcbiornaseq-0-2-10-2019-05-12">0.2.10</a></li>
      </ul>
    </div>
  </div>

</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io'>Lorena Pantano</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/chapmanb'>Brad Chapman</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/mistrm82'>Meeta Mistry</a>, <a href='https://github.com/rkhetani'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson'>John Hutchinson</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9100.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

