---
title: "Small RNA-seq quality control"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
bibliography: bcbioRnaseq.bib
params:
    upload_dir: "data/final"
    interesting_groups: !r c("genotype", "treatment")
    output_dir: "."
---

```{r setup, include=FALSE}
# Run [bcbioRnaseq::downloads()] to download `setup.R`
source("setup.R")
library(isomiRs)

# Directory paths
output_dir <- params$output_dir
data_dir <- file.path(output_dir, "data")
counts_dir <- file.path(output_dir, "results", "counts")
```

```{r header, child="header.Rmd"}
```



```{r bcb}
# bcbioRnaDataSet
if (file.exists(file.path(data_dir, "bcb.rda"))) {
    load(file.path(data_dir, "bcb.rda"))
} else {
    bcb <- load_run(
        upload_dir = params$upload_dir,
        interesting_groups = params$interesting_groups,
        analysis = "srnaseq")
    save_data(bcb, dir = data_dir)
}
```



# Exploratory analysis

In this section we will see descriptive figures about quality of the data, reads with adapter, reads mapped to miRNAs, reads mapped to other small RNAs.


## Size distribution

After adapter removal, we can plot the size distribution of the small RNAs.

```{r adapter}
plot_size_distribution(bcb)
```


## miRNA

### Total miRNA expression annotated with mirbase

```{r plot_mirna_counts, fig.height=9}
plot_mirna_counts(bcb)
```

```{r plot_srna_clusters}
plot_srna_clusters(bcb)
```

### isomiRs

```{r isomiRs, eval=ismirbase}
ids <- bcbio(bcb, "isomirs")
isoPlot(ids, type = "all", column = "group")
```



```{r footer, child="footer.Rmd"}
```
