<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>bcbioRNASeq • bcbioRNASeq</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="bcbioRNASeq">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioRNASeq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.17</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bcbioRNASeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>bcbioRNASeq</h1>
            
            <h4 class="date">25 April 2019</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/hbc/bcbioRNASeq/blob/master/vignettes/bcbioRNASeq.Rmd"><code>vignettes/bcbioRNASeq.Rmd</code></a></small>
      <div class="hidden name"><code>bcbioRNASeq.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>RNA-seq analysis involves multiple steps from processing raw sequencing data to identifying, organizing, annotating, and reporting differentially expressed genes. bcbio is an open source, community-maintained framework providing automated and scalable RNA-seq methods for identifying gene abundance counts. We have developed bcbioRNASeq, a Bioconductor package that provides ready-to-render templates and wrapper functions to post-process bcbio output data. bcbioRNASeq automates the generation of high-level RNA-seq reports, including identification of differentially expressed genes, functional enrichment analysis and quality control analysis.</p>
    </div>
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>For a high-level overview of our <a href="https://bcbio-nextgen.readthedocs.io/">bcbio</a> RNA-seq analysis pipeline, including detailed explanation of the <code>bcbioRNASeq</code> S4 class definition, first consult our <a href="http://dx.doi.org/10.12688/f1000research.12093.2">workflow paper</a> published in <em>F1000 Research</em> <span class="citation">(Steinbaugh et al. 2018)</span>. This vignette is focused on more advanced usage and edge cases that a user may encounter when attempting to load a bcbio dataset and perform downstream quality control analysis.</p>
<p><strong>Note:</strong> if you use <a href="https://bioinformatics.sph.harvard.edu/bcbioRNASeq/">bcbioRNASeq</a> in published research, please include this citation:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/citation">citation</a></span>(<span class="st">"bcbioRNASeq"</span>)</code></pre>
<pre><code>## 
## To cite bcbioRNASeq in publications use:
## 
##   Steinbaugh MJ, Pantano L, Kirchner RD, Barrera V, Chapman BA,
##   Piper ME, Mistry M, Khetani RS, Rutherford KD, Hoffman O,
##   Hutchinson JN, Ho Sui SJ. (2018). bcbioRNASeq: R package for
##   bcbio RNA-seq analysis. F1000Research, 6:1976. URL
##   https://f1000research.com/articles/6-1976/v2. DOI
##   10.12688/f1000research.12093.2.
## 
## A BibTeX entry for LaTeX users is
## 
##   @Article{,
##     title = {{bcbioRNASeq}: {R} package for bcbio {RNA-seq} analysis},
##     author = {Michael J. Steinbaugh and Lorena Pantano and Rory D. Kirchner and Victor Barrera and Brad A. Chapman and Mary E. Piper and Meeta Mistry and Radhika S. Khetani and Kayleigh D. Rutherford and Oliver Hofmann and John N. Hutchinson and Shannan J. Ho Sui},
##     journal = {F1000Research},
##     volume = {6},
##     number = {1976},
##     year = {2018},
##     url = {https://f1000research.com/articles/6-1976/v2},
##     doi = {10.12688/f1000research.12093.2},
##   }</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(bcbioRNASeq)</code></pre>
</div>
<div id="loading-bcbio-data" class="section level1">
<h1 class="hasAnchor">
<a href="#loading-bcbio-data" class="anchor"></a>Loading bcbio data</h1>
<p>The <code>bcbioRNASeq</code> constructor function is the main interface connecting bcbio] output data to interactive use in <a href="https://www.r-project.org/">R</a>. It is highly customizable and supports a number of options for advanced use cases. Consult the documentation for additional details.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/help">help</a></span>(<span class="dt">topic =</span> <span class="st">"bcbioRNASeq"</span>, <span class="dt">package =</span> <span class="st">"bcbioRNASeq"</span>)</code></pre>
<div id="upload-directory" class="section level2">
<h2 class="hasAnchor">
<a href="#upload-directory" class="anchor"></a>Upload directory</h2>
<p>We have designed the constructor to work as simply as possible by default. The only required argument is <code>uploadDir</code>, the path to the bcbio final upload directory specified with <code>upload:</code> in the YAML configuration. Refer to the <a href="https://bcbio-nextgen.readthedocs.io/en/latest/contents/configuration.html">bcbio configuration</a> documentation for detailed information on how to set up a bcbio run, which is outside the scope of this vignette.</p>
<p>For example, let’s load up the example bcbio dataset stored internally in the package.</p>
<pre class="sourceCode r"><code class="sourceCode r">uploadDir &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"bcbio"</span>, <span class="dt">package =</span> <span class="st">"bcbioRNASeq"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/stopifnot">stopifnot</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/files">file.exists</a></span>(uploadDir))</code></pre>
<p>bcbio outputs RNA-seq data in a standardized directory structure, which is described in detail in our workflow paper.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">path =</span> uploadDir, <span class="dt">full.names =</span> <span class="ot">FALSE</span>, <span class="dt">recursive =</span> <span class="ot">TRUE</span>)</code></pre>
<pre><code>##  [1] "2018-03-18_GSE65267-merged/bcbio-nextgen-commands.log"
##  [2] "2018-03-18_GSE65267-merged/bcbio-nextgen.log"         
##  [3] "2018-03-18_GSE65267-merged/combined.counts"           
##  [4] "2018-03-18_GSE65267-merged/programs.txt"              
##  [5] "2018-03-18_GSE65267-merged/project-summary.yaml"      
##  [6] "2018-03-18_GSE65267-merged/tx2gene.csv"               
##  [7] "control_rep1/salmon/quant.sf"                         
##  [8] "control_rep2/salmon/quant.sf"                         
##  [9] "control_rep3/salmon/quant.sf"                         
## [10] "fa_day7_rep1/salmon/quant.sf"                         
## [11] "fa_day7_rep2/salmon/quant.sf"                         
## [12] "fa_day7_rep3/salmon/quant.sf"                         
## [13] "sample_metadata.csv"</code></pre>
</div>
<div id="counts-level" class="section level2">
<h2 class="hasAnchor">
<a href="#counts-level" class="anchor"></a>Counts level</h2>
<p>By default, <a href="https://bioinformatics.sph.harvard.edu/bcbioRNASeq/">bcbioRNASeq</a> imports counts at gene level, which are required for standard differential expression analysis (<code>level = "genes"</code>). For pseudo-aligned counts (e.g. <a href="https://combine-lab.github.io/salmon/">Salmon</a>, <a href="https://pachterlab.github.io/kallisto/">Kallisto</a>, <a href="http://www.cs.cmu.edu/~ckingsf/software/sailfish/">Sailfish</a>) <span class="citation">(Bray et al. 2016; Patro, Mount, and Kingsford 2014; Patro et al. 2017)</span>, <a href="https://bioconductor.org/packages/tximport/">tximport</a> <span class="citation">(Soneson, Love, and Robinson 2016)</span> is used internally to aggregate transcript-level counts to gene-level counts, and generates length-scaled transcripts per million (TPM) values. For aligned counts processed with <a href="http://bioinf.wehi.edu.au/featureCounts/">featureCounts</a> <span class="citation">(Liao, Smyth, and Shi 2014)</span> (e.g. <a href="https://github.com/alexdobin/STAR/">STAR</a>, <a href="https://ccb.jhu.edu/software/hisat2/">HISAT2</a>) <span class="citation">(Dobin et al. 2013; Dobin and Gingeras 2016; Kim, Langmead, and Salzberg 2015)</span>, these values are already returned at gene level, and therefore not handled by <a href="https://bioconductor.org/packages/tximport/">tximport</a>. Once the gene-level counts are imported during the <code>bcbioRNASeq</code> call, the <a href="https://bioconductor.org/packages/DESeq2/">DESeq2</a> package <span class="citation">(Love, Huber, and Anders 2014)</span> is then used to generate an internal <code>DESeqDataSet</code> from which we derive normalized and variance-stabilized counts.</p>
<pre class="sourceCode r"><code class="sourceCode r">bcb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bcbioRNASeq-class.html">bcbioRNASeq</a></span>(<span class="dt">uploadDir =</span> uploadDir, <span class="dt">level =</span> <span class="st">"genes"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(bcb)</code></pre>
<pre><code>## bcbioRNASeq 0.3.17
## uploadDir: /data00/R/x86_64-redhat-linux-gnu-library/3.5/bioc-3.8/bcbioRNASeq/extdata/bcbio
## dates(2): [bcbio] 2018-03-18; [R] 2019-04-25
## level: genes
## caller: salmon
## organism: Mus musculus
## interestingGroups: sampleName
## class: RangedSummarizedExperiment 
## dim: 100 6 
## metadata(26): version level ... wd sessionInfo
## assays(5): counts tpm avgTxLength normalized vst
## rownames(100): ENSMUSG00000000001 ENSMUSG00000000003 ...
##   ENSMUSG00000062661 ENSMUSG00000074340
## rowData names(0):
## colnames(6): control_rep1 control_rep2 ... fa_day7_rep2
##   fa_day7_rep3
## colData names(26): averageInsertSize averageReadLength ... tissue
##   treatment</code></pre>
<p>Alternatively, if you want to perform transcript-aware analysis, such as differential exon usage or splicing analysis, transcript-level counts can be obtained using <code>level = "transcripts"</code>. Note that when counts are loaded at transcript level, TPMs are generated with tximport internally, but no additional normalizations or transformations normally calculated for gene-level counts with DESeq2 are generated.</p>
<pre class="sourceCode r"><code class="sourceCode r">bcb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bcbioRNASeq-class.html">bcbioRNASeq</a></span>(<span class="dt">uploadDir =</span> uploadDir, <span class="dt">level =</span> <span class="st">"transcripts"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(bcb)</code></pre>
<pre><code>## bcbioRNASeq 0.3.17
## uploadDir: /data00/R/x86_64-redhat-linux-gnu-library/3.5/bioc-3.8/bcbioRNASeq/extdata/bcbio
## dates(2): [bcbio] 2018-03-18; [R] 2019-04-25
## level: transcripts
## caller: salmon
## organism: Mus musculus
## interestingGroups: sampleName
## class: RangedSummarizedExperiment 
## dim: 100 6 
## metadata(26): version level ... wd sessionInfo
## assays(3): counts tpm avgTxLength
## rownames(100): ENSMUST00000000001 ENSMUST00000000003 ...
##   ENSMUST00000000674 ENSMUST00000000687
## rowData names(0):
## colnames(6): control_rep1 control_rep2 ... fa_day7_rep2
##   fa_day7_rep3
## colData names(26): averageInsertSize averageReadLength ... tissue
##   treatment</code></pre>
</div>
<div id="expression-callers" class="section level2">
<h2 class="hasAnchor">
<a href="#expression-callers" class="anchor"></a>Expression callers</h2>
<p>Since bcbio is flexible and supports a number of expression callers, we have provided advanced options in the <code>bcbioRNASeq</code> constructor to support a variety of workflows using the <code>caller</code> argument. Salmon, Kallisto, and Sailfish counts are supported at either gene or transcript level. Internally, these are loaded using tximport.</p>
<p>STAR and HISAT2 aligned counts processed with featureCounts are also supported, but only at gene level.</p>
<pre class="sourceCode r"><code class="sourceCode r">bcb_star &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bcbioRNASeq-class.html">bcbioRNASeq</a></span>(<span class="dt">uploadDir =</span> uploadDir, <span class="dt">caller =</span> <span class="st">"star"</span>)
<span class="kw">assayNames</span>(bcb_star)</code></pre>
<pre><code>## [1] "counts"     "normalized" "vst"</code></pre>
</div>
<div id="sample-selection-and-metadata" class="section level2">
<h2 class="hasAnchor">
<a href="#sample-selection-and-metadata" class="anchor"></a>Sample selection and metadata</h2>
<p>If you’d like to load up only a subset of samples, this can be done easily using the <code>samples</code> argument. Note that the <code>character</code> vector declared here must match the <code>description</code> column specified in the sample metadata. Conversely, if you’re working with a large dataset and you simply want to drop a few samples, this can be accomplished with the <code>censorSamples</code> argument.</p>
<p>When working with a bcbio run that has incorrect or outdated metadata, the simplest way to fix this issue is to pass in new metadata from an external spreadsheet (CSV or Excel) using the <code>sampleMetadataFile</code> argument. Note that this can also be used to subset the bcbio dataset, similar to the <code>samples</code> argument (see above), based on the rows that are included in the spreadsheet.</p>
<pre class="sourceCode r"><code class="sourceCode r">sampleMetadataFile &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/file.path">file.path</a></span>(uploadDir, <span class="st">"sample_metadata.csv"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/stopifnot">stopifnot</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/files">file.exists</a></span>(sampleMetadataFile))
<span class="kw"><a href="https://www.rdocumentation.org/packages/bcbioRNASeq/topics/reexports">import</a></span>(<span class="dt">file =</span> sampleMetadataFile)</code></pre>
<pre><code>##                   fileName  description  treatment day
## 1 control_rep1_R1.fastq.gz control_rep1    control   0
## 2 control_rep2_R1.fastq.gz control_rep2    control   0
## 3 control_rep3_R1.fastq.gz control_rep3    control   0
## 4 fa_day7_rep1_R1.fastq.gz fa_day7_rep1 folic acid   7
## 5 fa_day7_rep2_R1.fastq.gz fa_day7_rep2 folic acid   7
## 6 fa_day7_rep3_R1.fastq.gz fa_day7_rep3 folic acid   7</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">bcb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bcbioRNASeq-class.html">bcbioRNASeq</a></span>(
    <span class="dt">uploadDir =</span> uploadDir,
    <span class="dt">sampleMetadataFile =</span> sampleMetadataFile
)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(bcb)</code></pre>
<pre><code>## bcbioRNASeq 0.3.17
## uploadDir: /data00/R/x86_64-redhat-linux-gnu-library/3.5/bioc-3.8/bcbioRNASeq/extdata/bcbio
## dates(2): [bcbio] 2018-03-18; [R] 2019-04-25
## level: genes
## caller: salmon
## sampleMetadataFile: /data00/R/x86_64-redhat-linux-gnu-library/3.5/bioc-3.8/bcbioRNASeq/extdata/bcbio/sample_metadata.csv
## organism: Mus musculus
## interestingGroups: sampleName
## class: RangedSummarizedExperiment 
## dim: 100 6 
## metadata(26): version level ... wd sessionInfo
## assays(5): counts tpm avgTxLength normalized vst
## rownames(100): ENSMUSG00000000001 ENSMUSG00000000003 ...
##   ENSMUSG00000062661 ENSMUSG00000074340
## rowData names(0):
## colnames(6): control_rep1 control_rep2 ... fa_day7_rep2
##   fa_day7_rep3
## colData names(22): averageInsertSize averageReadLength ...
##   treatment day</code></pre>
</div>
<div id="genome-annotations" class="section level2">
<h2 class="hasAnchor">
<a href="#genome-annotations" class="anchor"></a>Genome annotations</h2>
<p>When analyzing a dataset against a well-annotated genome, we recommend importing the corresponding metadata using <a href="https://bioconductor.org/packages/AnnotationHub/">AnnotationHub</a> and <a href="https://bioconductor.org/packages/ensembldb/">ensembldb</a>. This functionality is natively supported in the <code>bcbioRNASeq</code> constructor with using the <code>organism</code>, <code>ensemblRelease</code>, and <code>genomeBuild</code> arguments. This will return a <code>GRanges</code> object using the <a href="https://bioconductor.org/packages/GenomicRanges/">GenomicRanges</a> package <span class="citation">(Lawrence et al. 2013)</span>, which contains coordinates and rich metadata for each gene or transcript. These annotations are accessible with the <code>rowRanges</code> and <code>rowData</code> functions defined in the <a href="https://bioconductor.org/packages/SummarizedExperiment/">SummarizedExperiment</a> package <span class="citation">(Huber et al. 2015)</span>.</p>
<p>For example, with our internal bcbio dataset, we’re analyzing counts generated against the <a href="https://www.ensembl.org/">Ensembl</a> <em>Mus musculus</em> GRCm38 genome build (release 87). These parameters can be defined in the object load call to ensure that the annotations match up exactly with the genome used.</p>
<pre class="sourceCode r"><code class="sourceCode r">bcb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bcbioRNASeq-class.html">bcbioRNASeq</a></span>(
    <span class="dt">uploadDir =</span> uploadDir,
    <span class="dt">level =</span> <span class="st">"genes"</span>,
    <span class="dt">organism =</span> <span class="st">"Mus musculus"</span>,
    <span class="dt">genomeBuild =</span> <span class="st">"GRCm38"</span>,
    <span class="dt">ensemblRelease =</span> <span class="dv">87</span>
)
<span class="kw">rowRanges</span>(bcb)</code></pre>
<pre><code>## GRanges object with 100 ranges and 7 metadata columns:
##                      seqnames              ranges strand | broadClass
##                         &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |      &lt;Rle&gt;
##   ENSMUSG00000000001        3 108107280-108146146      - |     coding
##   ENSMUSG00000000003        X   77837901-77853623      - |     coding
##   ENSMUSG00000000028       16   18780447-18811987      - |     coding
##   ENSMUSG00000000049       11 108343354-108414396      + |     coding
##   ENSMUSG00000000058        6   17281185-17289115      + |     coding
##                  ...      ...                 ...    ... .        ...
##   ENSMUSG00000048583        7 142650766-142666816      - |     coding
##   ENSMUSG00000055022       15   92051165-92341967      + |     coding
##   ENSMUSG00000061689        2 156613705-156764363      + |     coding
##   ENSMUSG00000062661        2   31245823-31295989      + |     coding
##   ENSMUSG00000074340        3 105973711-105987423      + |     coding
##                                                                                                               description
##                                                                                                                     &lt;Rle&gt;
##   ENSMUSG00000000001 guanine nucleotide binding protein (G protein), alpha inhibiting 3 [Source:MGI Symbol;Acc:MGI:95773]
##   ENSMUSG00000000003                                                         probasin [Source:MGI Symbol;Acc:MGI:1860484]
##   ENSMUSG00000000028                                           cell division cycle 45 [Source:MGI Symbol;Acc:MGI:1338073]
##   ENSMUSG00000000049                                                   apolipoprotein H [Source:MGI Symbol;Acc:MGI:88058]
##   ENSMUSG00000000058                                                        caveolin 2 [Source:MGI Symbol;Acc:MGI:107571]
##                  ...                                                                                                  ...
##   ENSMUSG00000048583                                       insulin-like growth factor 2 [Source:MGI Symbol;Acc:MGI:96434]
##   ENSMUSG00000055022                                                       contactin 1 [Source:MGI Symbol;Acc:MGI:105980]
##   ENSMUSG00000061689           discs, large homolog-associated protein 4 (Drosophila) [Source:MGI Symbol;Acc:MGI:2138865]
##   ENSMUSG00000062661                                         neuronal calcium sensor 1 [Source:MGI Symbol;Acc:MGI:109166]
##   ENSMUSG00000074340                                          oviductal glycoprotein 1 [Source:MGI Symbol;Acc:MGI:106661]
##                      entrezID    geneBiotype             geneID geneName
##                        &lt;list&gt;          &lt;Rle&gt;              &lt;Rle&gt;    &lt;Rle&gt;
##   ENSMUSG00000000001    14679 protein_coding ENSMUSG00000000001    Gnai3
##   ENSMUSG00000000003    54192 protein_coding ENSMUSG00000000003     Pbsn
##   ENSMUSG00000000028    12544 protein_coding ENSMUSG00000000028    Cdc45
##   ENSMUSG00000000049    11818 protein_coding ENSMUSG00000000049     Apoh
##   ENSMUSG00000000058    12390 protein_coding ENSMUSG00000000058     Cav2
##                  ...      ...            ...                ...      ...
##   ENSMUSG00000048583    16002 protein_coding ENSMUSG00000048583     Igf2
##   ENSMUSG00000055022    12805 protein_coding ENSMUSG00000055022    Cntn1
##   ENSMUSG00000061689   228836 protein_coding ENSMUSG00000061689   Dlgap4
##   ENSMUSG00000062661    14299 protein_coding ENSMUSG00000062661     Ncs1
##   ENSMUSG00000074340    12659 protein_coding ENSMUSG00000074340    Ovgp1
##                      seqCoordSystem
##                               &lt;Rle&gt;
##   ENSMUSG00000000001     chromosome
##   ENSMUSG00000000003     chromosome
##   ENSMUSG00000000028     chromosome
##   ENSMUSG00000000049     chromosome
##   ENSMUSG00000000058     chromosome
##                  ...            ...
##   ENSMUSG00000048583     chromosome
##   ENSMUSG00000055022     chromosome
##   ENSMUSG00000061689     chromosome
##   ENSMUSG00000062661     chromosome
##   ENSMUSG00000074340     chromosome
##   -------
##   seqinfo: 86 sequences from GRCm38 genome</code></pre>
<p>Alternatively, transcript-level annotations can also be obtained automatically using this method.</p>
<pre class="sourceCode r"><code class="sourceCode r">bcb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bcbioRNASeq-class.html">bcbioRNASeq</a></span>(
    <span class="dt">uploadDir =</span> uploadDir,
    <span class="dt">level =</span> <span class="st">"transcripts"</span>,
    <span class="dt">organism =</span> <span class="st">"Mus musculus"</span>,
    <span class="dt">genomeBuild =</span> <span class="st">"GRCm38"</span>,
    <span class="dt">ensemblRelease =</span> <span class="dv">87</span>
)
<span class="kw">rowRanges</span>(bcb)</code></pre>
<pre><code>## GRanges object with 100 ranges and 13 metadata columns:
##                      seqnames              ranges strand | broadClass
##                         &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |      &lt;Rle&gt;
##   ENSMUST00000000001        3 108107280-108146146      - |     coding
##   ENSMUST00000000003        X   77837901-77853623      - |     coding
##   ENSMUST00000000010       11   96271457-96276595      + |     coding
##   ENSMUST00000000028       16   18780447-18811972      - |     coding
##   ENSMUST00000000033        7 142650766-142657496      - |     coding
##                  ...      ...                 ...    ... .        ...
##   ENSMUST00000000642        6   82725025-82774454      - |     coding
##   ENSMUST00000000645       11   77777265-77865980      + |     coding
##   ENSMUST00000000646       11   77931012-77979040      + |     coding
##   ENSMUST00000000674        6   17743592-17942966      + |     coding
##   ENSMUST00000000687       17   83831356-83846790      - |     coding
##                                                                                                               description
##                                                                                                                     &lt;Rle&gt;
##   ENSMUST00000000001 guanine nucleotide binding protein (G protein), alpha inhibiting 3 [Source:MGI Symbol;Acc:MGI:95773]
##   ENSMUST00000000003                                                         probasin [Source:MGI Symbol;Acc:MGI:1860484]
##   ENSMUST00000000010                                                        homeobox B9 [Source:MGI Symbol;Acc:MGI:96190]
##   ENSMUST00000000028                                           cell division cycle 45 [Source:MGI Symbol;Acc:MGI:1338073]
##   ENSMUST00000000033                                       insulin-like growth factor 2 [Source:MGI Symbol;Acc:MGI:96434]
##                  ...                                                                                                  ...
##   ENSMUST00000000642                                                     hexokinase 2 [Source:MGI Symbol;Acc:MGI:1315197]
##   ENSMUST00000000645                                                    myosin XVIIIA [Source:MGI Symbol;Acc:MGI:2667185]
##   ENSMUST00000000646                                            seizure related gene 6 [Source:MGI Symbol;Acc:MGI:104745]
##   ENSMUST00000000674                                  suppression of tumorigenicity 7 [Source:MGI Symbol;Acc:MGI:1927450]
##   ENSMUST00000000687                            3-hydroxyanthranilate 3,4-dioxygenase [Source:MGI Symbol;Acc:MGI:1349444]
##                      entrezID    geneBiotype             geneID geneName
##                        &lt;list&gt;          &lt;Rle&gt;              &lt;Rle&gt;    &lt;Rle&gt;
##   ENSMUST00000000001    14679 protein_coding ENSMUSG00000000001    Gnai3
##   ENSMUST00000000003    54192 protein_coding ENSMUSG00000000003     Pbsn
##   ENSMUST00000000010    15417 protein_coding ENSMUSG00000020875    Hoxb9
##   ENSMUST00000000028    12544 protein_coding ENSMUSG00000000028    Cdc45
##   ENSMUST00000000033    16002 protein_coding ENSMUSG00000048583     Igf2
##                  ...      ...            ...                ...      ...
##   ENSMUST00000000642    15277 protein_coding ENSMUSG00000000628      Hk2
##   ENSMUST00000000645   360013 protein_coding ENSMUSG00000000631   Myo18a
##   ENSMUST00000000646    20370 protein_coding ENSMUSG00000000632     Sez6
##   ENSMUST00000000674    64213 protein_coding ENSMUSG00000029534      St7
##   ENSMUST00000000687   107766 protein_coding ENSMUSG00000000673     Haao
##                      seqCoordSystem transcriptBiotype transcriptCdsSeqEnd
##                               &lt;Rle&gt;             &lt;Rle&gt;               &lt;Rle&gt;
##   ENSMUST00000000001     chromosome    protein_coding           108146005
##   ENSMUST00000000003     chromosome    protein_coding            77853483
##   ENSMUST00000000010     chromosome    protein_coding            96274859
##   ENSMUST00000000028     chromosome    protein_coding            18811565
##   ENSMUST00000000033     chromosome    protein_coding           142655810
##                  ...            ...               ...                 ...
##   ENSMUST00000000642     chromosome    protein_coding            82773980
##   ENSMUST00000000645     chromosome    protein_coding            77864835
##   ENSMUST00000000646     chromosome    protein_coding            77978058
##   ENSMUST00000000674     chromosome    protein_coding            17942765
##   ENSMUST00000000687     chromosome    protein_coding            83846716
##                      transcriptCdsSeqStart       transcriptID
##                                      &lt;Rle&gt;              &lt;Rle&gt;
##   ENSMUST00000000001             108109422 ENSMUST00000000001
##   ENSMUST00000000003              77841883 ENSMUST00000000003
##   ENSMUST00000000010              96271542 ENSMUST00000000010
##   ENSMUST00000000028              18781898 ENSMUST00000000028
##   ENSMUST00000000033             142653816 ENSMUST00000000033
##                  ...                   ...                ...
##   ENSMUST00000000642              82727310 ENSMUST00000000642
##   ENSMUST00000000645              77777316 ENSMUST00000000645
##   ENSMUST00000000646              77931060 ENSMUST00000000646
##   ENSMUST00000000674              17743682 ENSMUST00000000674
##   ENSMUST00000000687              83831643 ENSMUST00000000687
##                          transcriptName transcriptSupportLevel
##                                   &lt;Rle&gt;                  &lt;Rle&gt;
##   ENSMUST00000000001 ENSMUST00000000001                      1
##   ENSMUST00000000003 ENSMUST00000000003                      1
##   ENSMUST00000000010 ENSMUST00000000010                      1
##   ENSMUST00000000028 ENSMUST00000000028                      1
##   ENSMUST00000000033 ENSMUST00000000033                      1
##                  ...                ...                    ...
##   ENSMUST00000000642 ENSMUST00000000642                      1
##   ENSMUST00000000645 ENSMUST00000000645                      5
##   ENSMUST00000000646 ENSMUST00000000646                      1
##   ENSMUST00000000674 ENSMUST00000000674                      1
##   ENSMUST00000000687 ENSMUST00000000687                      1
##   -------
##   seqinfo: 86 sequences from GRCm38 genome</code></pre>
<p>When working with a dataset generated against a poorly-annotated or non-standard genome, we provide a fallback method for loading gene annotations from a <a href="https://useast.ensembl.org/info/website/upload/gff.html">general feature format (GFF)</a> file with the <code>gffFile</code> argument. If possible, we recommend providing a general transfer format (GTF) file, which is identical to GFF version 2. GFFv3 is more complicated and non-standard, but Ensembl GFFv3 files are also supported.</p>
<p>If your dataset contains transgenes (e.g. EGFP, TDTOMATO) or spike-ins (e.g. ERCCs), these features can be defined with the <code>transgeneNames</code> and <code>spikeNames</code> arguments, which will automatically populate the <code>rowRanges</code> slot with placeholder metadata.</p>
<p>We recommend loading up data per genome in its own <code>bcbioRNASeq</code> object when possible, so that rich metadata can be imported easily. In the edge case where you need to look at multiple genomes simultaneously, set <code>organism = NULL</code>, and bcbioRNASeq will skip the gene annotation acquisition step.</p>
<pre class="sourceCode r"><code class="sourceCode r">bcb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bcbioRNASeq-class.html">bcbioRNASeq</a></span>(<span class="dt">uploadDir =</span> uploadDir, <span class="dt">organism =</span> <span class="ot">NULL</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(bcb)</code></pre>
<pre><code>## bcbioRNASeq 0.3.17
## uploadDir: /data00/R/x86_64-redhat-linux-gnu-library/3.5/bioc-3.8/bcbioRNASeq/extdata/bcbio
## dates(2): [bcbio] 2018-03-18; [R] 2019-04-25
## level: genes
## caller: salmon
## organism: Mus musculus
## interestingGroups: sampleName
## class: RangedSummarizedExperiment 
## dim: 100 6 
## metadata(26): version level ... wd sessionInfo
## assays(5): counts tpm avgTxLength normalized vst
## rownames(100): ENSMUSG00000000001 ENSMUSG00000000003 ...
##   ENSMUSG00000062661 ENSMUSG00000074340
## rowData names(0):
## colnames(6): control_rep1 control_rep2 ... fa_day7_rep2
##   fa_day7_rep3
## colData names(26): averageInsertSize averageReadLength ... tissue
##   treatment</code></pre>
</div>
<div id="variance-stabilization" class="section level2">
<h2 class="hasAnchor">
<a href="#variance-stabilization" class="anchor"></a>Variance stabilization</h2>
<p>During the <code>bcbioRNASeq</code> constructor call, variance stabilizaton of gene-level counts can be calculated automatically, and is recommended. This is performed internally by the DESeq2 package. These transformations will be slotted into <code>assays</code> as <code>vst</code> and/or <code>rlog</code> matrices. The <code>vst</code> matrix is calculated by <code><a href="https://www.rdocumentation.org/packages/DESeq2/topics/varianceStabilizingTransformation">DESeq2::varianceStabilizingTransformation</a></code> and <code>rlog</code> by <code><a href="https://www.rdocumentation.org/packages/DESeq2/topics/rlog">DESeq2::rlog</a></code>. For large datasets, <code>rlog</code> calculation can take a long time, so we are currently recommending calculation of only <code>vst</code> by default.</p>
<pre class="sourceCode r"><code class="sourceCode r">bcb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bcbioRNASeq-class.html">bcbioRNASeq</a></span>(<span class="dt">uploadDir =</span> uploadDir, <span class="dt">vst =</span> <span class="ot">TRUE</span>, <span class="dt">rlog =</span> <span class="ot">TRUE</span>)
<span class="kw">assayNames</span>(bcb)</code></pre>
<pre><code>## [1] "counts"      "tpm"         "avgTxLength" "normalized"  "vst"        
## [6] "rlog"</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(<span class="kw"><a href="../reference/counts.html">counts</a></span>(bcb, <span class="dt">normalized =</span> <span class="st">"vst"</span>))</code></pre>
<pre><code>##   control_rep1     control_rep2     control_rep3     fa_day7_rep1   
##  Min.   : 4.919   Min.   : 4.919   Min.   : 4.919   Min.   : 4.919  
##  1st Qu.: 4.919   1st Qu.: 4.919   1st Qu.: 4.919   1st Qu.: 4.919  
##  Median : 6.051   Median : 5.658   Median : 6.009   Median : 6.237  
##  Mean   : 6.713   Mean   : 6.706   Mean   : 6.715   Mean   : 6.762  
##  3rd Qu.: 7.990   3rd Qu.: 8.239   3rd Qu.: 7.815   3rd Qu.: 8.239  
##  Max.   :12.215   Max.   :12.972   Max.   :11.818   Max.   :11.764  
##   fa_day7_rep2     fa_day7_rep3   
##  Min.   : 4.919   Min.   : 4.919  
##  1st Qu.: 5.101   1st Qu.: 4.919  
##  Median : 6.216   Median : 6.185  
##  Mean   : 6.754   Mean   : 6.789  
##  3rd Qu.: 8.033   3rd Qu.: 7.930  
##  Max.   :12.086   Max.   :11.997</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(<span class="kw"><a href="../reference/counts.html">counts</a></span>(bcb, <span class="dt">normalized =</span> <span class="st">"rlog"</span>))</code></pre>
<pre><code>##   control_rep1     control_rep2     control_rep3     fa_day7_rep1   
##  Min.   :-1.970   Min.   :-1.980   Min.   :-1.970   Min.   :-1.976  
##  1st Qu.: 1.519   1st Qu.: 1.433   1st Qu.: 1.384   1st Qu.: 1.475  
##  Median : 3.934   Median : 3.720   Median : 3.954   Median : 4.478  
##  Mean   : 4.425   Mean   : 4.408   Mean   : 4.411   Mean   : 4.476  
##  3rd Qu.: 7.390   3rd Qu.: 7.856   3rd Qu.: 7.322   3rd Qu.: 7.838  
##  Max.   :12.175   Max.   :12.772   Max.   :11.777   Max.   :11.826  
##   fa_day7_rep2     fa_day7_rep3   
##  Min.   :-1.976   Min.   :-1.975  
##  1st Qu.: 1.504   1st Qu.: 1.459  
##  Median : 3.894   Median : 4.090  
##  Mean   : 4.454   Mean   : 4.485  
##  3rd Qu.: 7.599   3rd Qu.: 7.637  
##  Max.   :12.075   Max.   :12.006</code></pre>
</div>
</div>
<div id="use-case-dataset" class="section level1">
<h1 class="hasAnchor">
<a href="#use-case-dataset" class="anchor"></a>Use case dataset</h1>
<p>To demonstrate the functionality and configuration of the package, we have taken an experiment from the <a href="https://www.ncbi.nlm.nih.gov/geo/">Gene Expression Omnibus (GEO)</a> public repository of expression data to use as an example use case. The RNA-seq data is from a study of acute kidney injury in a mouse model (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE65267">GSE65267</a>) <span class="citation">(Craciun et al. 2016)</span>. The study aims to identify differentially expressed genes in progressive kidney fibrosis and contains samples from mouse kidneys at several time points (n = 3, per time point) after folic acid treatment. From this dataset, we are using a subset of the samples for our use case: before folic acid treatment, and 1, 3, 7 days after treatment.</p>
<p>For the vignette, we are loading a pre-computed version of the example <code>bcbioRNASeq</code> object used in our workflow paper.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rm">rm</a></span>(bcb)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/bcbioRNASeq/topics/reexports">loadRemoteData</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/bcbioRNASeq/topics/reexports">pasteURL</a></span>(
  <span class="st">"github.com"</span>, <span class="st">"hbc"</span>, <span class="st">"bcbioRNASeq"</span>,
  <span class="st">"raw"</span>, <span class="st">"f1000v2"</span>, <span class="st">"data"</span>, <span class="st">"bcb.rda"</span>,
  <span class="dt">protocol =</span> <span class="st">"https"</span>
))
bcb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/updateObject.html">updateObject</a></span>(bcb)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(bcb)</code></pre>
<pre><code>## bcbioRNASeq 0.3.17
## uploadDir: /n/data1/cores/bcbio/bcbioRNASeq/F1000v2/GSE65267-merged/final
## dates(2): [bcbio] 2018-03-18; [R] 2019-04-25
## level: genes
## caller: salmon
## organism: Mus musculus
## interestingGroups: day
## class: RangedSummarizedExperiment 
## dim: 51652 12 
## metadata(29): version level ... sessionInfo previousVersion
## assays(6): counts tpm ... vst avgTxLength
## rownames(51652): ENSMUSG00000000001 ENSMUSG00000000003 ...
##   ENSMUSG00000114967 ENSMUSG00000114968
## rowData names(7): geneID geneName ... entrezID broadClass
## colnames(12): control_rep1 control_rep2 ... fa_day7_rep2
##   fa_day7_rep3
## colData names(25): sampleName day ... x5x3Bias xGC</code></pre>
</div>
<div id="sample-metadata" class="section level1">
<h1 class="hasAnchor">
<a href="#sample-metadata" class="anchor"></a>Sample metadata</h1>
<p>For reference, let’s take a look at the sample metadata. By comparison, using <code>colData</code> will return all sample-level metadata, including our quality control metrics generated by bcbio. We recommend instead using <code>sampleData</code> with the <code>clean = TRUE</code> argument in reports, which only returns <code>factor</code> columns of interest.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/bcbioRNASeq/topics/reexports">sampleData</a></span>(bcb)</code></pre>
<pre><code>## DataFrame with 12 rows and 7 columns
##                sampleName      day replicate   strain   tissue  treatment
##                  &lt;factor&gt; &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;   &lt;factor&gt;
## control_rep1 control_rep1        0         1   Balb/c   kidney    control
## control_rep2 control_rep2        0         2   Balb/c   kidney    control
## control_rep3 control_rep3        0         3   Balb/c   kidney    control
## fa_day1_rep1 fa_day1_rep1        1         1   Balb/c   kidney folic acid
## fa_day1_rep2 fa_day1_rep2        1         2   Balb/c   kidney folic acid
## ...                   ...      ...       ...      ...      ...        ...
## fa_day3_rep2 fa_day3_rep2        3         2   Balb/c   kidney folic acid
## fa_day3_rep3 fa_day3_rep3        3         3   Balb/c   kidney folic acid
## fa_day7_rep1 fa_day7_rep1        7         1   Balb/c   kidney folic acid
## fa_day7_rep2 fa_day7_rep2        7         2   Balb/c   kidney folic acid
## fa_day7_rep3 fa_day7_rep3        7         3   Balb/c   kidney folic acid
##              interestingGroups
##                       &lt;factor&gt;
## control_rep1                 0
## control_rep2                 0
## control_rep3                 0
## fa_day1_rep1                 1
## fa_day1_rep2                 1
## ...                        ...
## fa_day3_rep2                 3
## fa_day3_rep3                 3
## fa_day7_rep1                 7
## fa_day7_rep2                 7
## fa_day7_rep3                 7</code></pre>
</div>
<div id="interesting-groups" class="section level1">
<h1 class="hasAnchor">
<a href="#interesting-groups" class="anchor"></a>Interesting groups</h1>
<p>Groups of interest to be used for coloring and sample grouping in the quality control plots can be defined in the <code>bcbioRNASeq</code> object using the <code>interestingGroups</code> argument in the <code>bcbioRNASeq</code> constructor call. Assignment method support with the <code>interestingGroups</code> function is also provided, which can modify the groups of interest after the object has been created.</p>
<p>The <code>interestingGroups</code> definition defaults to <code>sampleName</code>, which is automatically generated from the bcbio <code>description</code> metadata for demultiplexed bulk RNA-seq samples. In this case, the samples will be grouped and colored uniquely.</p>
<p>Interesting groups must be defined using a <code>character</code> vector and refer to column names defined in the <code>colData</code> slot of the object. Note that the bcbioRNASeq package uses lower camel case formatting for column names (e.g. “sampleName”), so the interesting groups should be defined using camel case, not snake (e.g. “sample_name”) or dotted case (e.g. “sample.name”).</p>
<p>This approach was inspired by the DESeq2 package, which uses the argument <code>intgroup</code> in some functions, such as <code>plotPCA</code> for labeling groups of interest. We took this idea and ran with it for a number of our quality control functions, which are described in detail below.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/bcbioRNASeq/topics/reexports">interestingGroups</a></span>(bcb) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"treatment"</span>, <span class="st">"day"</span>)</code></pre>
</div>
<div id="quality-control" class="section level1">
<h1 class="hasAnchor">
<a href="#quality-control" class="anchor"></a>Quality control</h1>
<p>Our workflow paper describes the quality control process in detail. In addition, our <strong>Quality Control</strong> <a href="https://rmarkdown.rstudio.com/">R Markdown</a> template contains notes detailing each metric. Here were are going into more technical detail regarding how to customize the appearance of the plots. Note that all quality control plotting functions inherit the <code>interestingGroups</code> defined inside the <code>bcbioRNASeq</code> object. You can also change this dynamially for each function call using the <code>interestingGroups</code> argument.</p>
<div id="read-counts" class="section level2">
<h2 class="hasAnchor">
<a href="#read-counts" class="anchor"></a>Read counts</h2>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotTotalReads.html">plotTotalReads</a></span>(bcb)</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotMappedReads.html">plotMappedReads</a></span>(bcb)</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-19-2.png" width="700"></p>
</div>
<div id="mapping-rates" class="section level2">
<h2 class="hasAnchor">
<a href="#mapping-rates" class="anchor"></a>Mapping rates</h2>
<p>Note that the overall mapping rate is relatively low per sample, while the exonic mapping rate is acceptable. High quality samples should have low intronic mapping rate, and high values are indicative of sample degradation and/or contamination.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotMappingRate.html">plotMappingRate</a></span>(bcb)</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotExonicMappingRate.html">plotExonicMappingRate</a></span>(bcb)</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-20-2.png" width="700"></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotIntronicMappingRate.html">plotIntronicMappingRate</a></span>(bcb)</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-20-3.png" width="700"></p>
</div>
<div id="rrna-mapping-rate" class="section level2">
<h2 class="hasAnchor">
<a href="#rrna-mapping-rate" class="anchor"></a>rRNA mapping rate</h2>
<p>Note that the samples here have a high rRNA mapping rate. This can be indicative of the polyA enrichment or ribo depletion protocol not having removed all ribosomal RNA (rRNA) transcripts. This will reduce the number of biologically meaningful reads in the experiment and is best avoided.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotRRNAMappingRate.html">plotRRNAMappingRate</a></span>(bcb)</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
</div>
<div id="bias" class="section level2">
<h2 class="hasAnchor">
<a href="#bias" class="anchor"></a>5’-&gt;3’ bias</h2>
<p>RNA-seq data can have specific biases at either the 5’ or 3’ end of sequenced fragments. It is common to see a small amount of bias, especially if polyA enrichment was performed, or if there is any sample degradation. If a large amount of bias is observed here, be sure to analyze the samples with a Bioanalyzer and check the RIN scores.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plot5Prime3PrimeBias.html">plot5Prime3PrimeBias</a></span>(bcb)</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
</div>
<div id="gene-distributions" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-distributions" class="anchor"></a>Gene distributions</h2>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/reexports-S4.html">plotFeaturesDetected</a></span>(bcb)</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotGeneSaturation.html">plotGeneSaturation</a></span>(bcb, <span class="dt">label =</span> <span class="ot">FALSE</span>)</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotGeneSaturation.html">plotGeneSaturation</a></span>(bcb, <span class="dt">label =</span> <span class="ot">TRUE</span>)</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-24-2.png" width="700"></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotCountsPerFeature.html">plotCountsPerFeature</a></span>(bcb, <span class="dt">geom =</span> <span class="st">"boxplot"</span>)</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotCountsPerFeature.html">plotCountsPerFeature</a></span>(bcb, <span class="dt">geom =</span> <span class="st">"density"</span>)</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-25-2.png" width="700"></p>
</div>
<div id="variance-stabilization-1" class="section level2">
<h2 class="hasAnchor">
<a href="#variance-stabilization-1" class="anchor"></a>Variance stabilization</h2>
<p>These plots show the standard deviation of normalized counts using <code>log2</code>, <code>varianceStabilizingTransform</code>, <code>rlog</code>, and <code>tmm</code> by <code><a href="https://www.rdocumentation.org/packages/base/topics/rank">rank(mean)</a></code>. Note that all counts shown are log2 scale.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotMeanSD.html">plotMeanSD</a></span>(bcb)</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
</div>
<div id="sample-similarity" class="section level2">
<h2 class="hasAnchor">
<a href="#sample-similarity" class="anchor"></a>Sample similarity</h2>
<p>We can visualize sample similarity with principal component analysis (PCA) and hierarchical clustering of sample correlations. These functions support multiple normalization methods with the <code>normalized</code> argument.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotPCA.html">plotPCA</a></span>(bcb, <span class="dt">label =</span> <span class="ot">FALSE</span>)</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotPCA.html">plotPCA</a></span>(bcb, <span class="dt">label =</span> <span class="ot">TRUE</span>)</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-27-2.png" width="700"></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotCorrelationHeatmap.html">plotCorrelationHeatmap</a></span>(bcb, <span class="dt">method =</span> <span class="st">"pearson"</span>)</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-28-1.png" width="700"></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotCorrelationHeatmap.html">plotCorrelationHeatmap</a></span>(bcb, <span class="dt">method =</span> <span class="st">"spearman"</span>)</code></pre>
<p><img src="bcbioRNASeq_files/figure-html/unnamed-chunk-28-2.png" width="700"></p>
</div>
</div>
<div id="saving-counts" class="section level1">
<h1 class="hasAnchor">
<a href="#saving-counts" class="anchor"></a>Saving counts</h1>
<p>The <code>counts</code> accessor function returns raw counts by default. Multiple normalization methods are accessible with the <code>normalized</code> argument:</p>
<ul>
<li>
<code>FALSE</code>: Raw counts. When importing using a tximport-compatible caller (e.g. salmon, kallisto), these are length scaled by default (see <code>countsFromAbundance</code>). When importing using a featureCounts-compatible caller (e.g. HISAT2, STAR), these are integer counts.</li>
<li>
<code>TRUE</code>: DESeq2 size factor adjusted counts.</li>
<li>
<code>"tpm"</code>: Transcripts per million, calculated by tximport.</li>
<li>
<code>"tmm"</code>: Trimmed mean of M-values normalization method.</li>
<li>
<code>"vst"</code>: Variance stabilization transformation.</li>
<li>
<code>"rlog"</code>: Regularized log transformation.</li>
<li>
<code>"rle"</code>: Relative log expression transformation.</li>
</ul>
<p>We are providing <code>bcbioRNASeq</code> method support for the <code>export</code> function, which can automatically write these counts and corresponding metadata (e.g. <code>rowData</code>, <code>colData</code>) to disk in a single call.</p>
<pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/bcbioRNASeq/topics/reexports">export</a></span>(bcb, <span class="dt">dir =</span> <span class="st">"export"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(x)</code></pre>
<pre><code>## $assays
## $assays$counts
## [1] "/data00/packages/bcbioRNASeq/vignettes/export/bcb/assays/counts.csv"
## 
## $assays$tpm
## [1] "/data00/packages/bcbioRNASeq/vignettes/export/bcb/assays/tpm.csv"
## 
## $assays$normalized
## [1] "/data00/packages/bcbioRNASeq/vignettes/export/bcb/assays/normalized.csv"
## 
## $assays$rlog
## [1] "/data00/packages/bcbioRNASeq/vignettes/export/bcb/assays/rlog.csv"
## 
## $assays$vst
## [1] "/data00/packages/bcbioRNASeq/vignettes/export/bcb/assays/vst.csv"
## 
## $assays$avgTxLength
## [1] "/data00/packages/bcbioRNASeq/vignettes/export/bcb/assays/avgTxLength.csv"
## 
## 
## $colData
## [1] "/data00/packages/bcbioRNASeq/vignettes/export/bcb/colData.csv"
## 
## $rowData
## [1] "/data00/packages/bcbioRNASeq/vignettes/export/bcb/rowData.csv"</code></pre>
</div>
<div id="differential-expression" class="section level1">
<h1 class="hasAnchor">
<a href="#differential-expression" class="anchor"></a>Differential expression</h1>
<p>bcbioRNASeq integrates with DESeq2, which we recommend for differential expression. To prepare our dataset for analysis, we need to coerce the <code>bcbioRNASeq</code> object to a <code>DESeqDataSet</code>. Use the <code>as</code> function or refer to the DESeq2 documentation for the <code>DESeqDataSet</code> constructor.</p>
<pre class="sourceCode r"><code class="sourceCode r">dds &lt;-<span class="st"> </span><span class="kw">as</span>(bcb, <span class="st">"DESeqDataSet"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(dds)</code></pre>
<pre><code>## class: DESeqDataSet 
## dim: 51652 12 
## metadata(11): version caller ... date sessionInfo
## assays(1): counts
## rownames(51652): ENSMUSG00000000001 ENSMUSG00000000003 ...
##   ENSMUSG00000114967 ENSMUSG00000114968
## rowData names(7): geneID geneName ... entrezID broadClass
## colnames(12): control_rep1 control_rep2 ... fa_day7_rep2
##   fa_day7_rep3
## colData names(25): sampleName day ... x5x3Bias xGC</code></pre>
<p>Since both <code>bcbioRNASeq</code> and <code>DESeqDataSet</code> classes extend <code>RangedSummarizedExperiment</code>, internally we coerce the original <code>bcbioRNASeq</code> object to a <code>RangedSummarizedExperiment</code> and then use the <code>DESeqDataSet</code> constructor, which requires a <code>SummarizedExperiment</code> with integer counts.</p>
<p>The source code for our <code>bcbioRNASeq</code> to <code>DESeqDataSet</code> coercion is accessible with <code>getMethod</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">getMethod</span>(
    <span class="dt">f =</span> <span class="st">"coerce"</span>,
    <span class="dt">signature =</span> <span class="kw">signature</span>(<span class="dt">from =</span> <span class="st">"bcbioRNASeq"</span>, <span class="dt">to =</span> <span class="st">"DESeqDataSet"</span>)
)</code></pre>
<p>Now you can perform differential expression analysis with DESeq2. The authors of that package have provided a number of detailed, well documented references online:</p>
<ul>
<li>
<a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">Analyzing RNA-seq data with DESeq2</a>.</li>
<li>
<a href="http://dx.doi.org/10.12688/f1000research.7035.2">RNA-Seq workflow: gene-level exploratory analysis and differential expression</a>.</li>
<li>
<a href="https://f1000research.com/articles/4-1521/v2">Differential analyses for RNA-seq: transcript-level estimates improve gene-level inferences</a>.</li>
</ul>
<p>We also provide a parameterized starter R Markdown template for standard DESeq2] differential expression, supporting analysis and visualization of multiple contrasts inside a single report. In our workflow paper, we also describe an example LRT analysis.</p>
</div>
<div id="functional-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#functional-analysis" class="anchor"></a>Functional analysis</h1>
<p>We provide a starter R Markdown template for <a href="http://software.broadinstitute.org/gsea/">gene set enrichment analysis (GSEA)</a> and <a href="https://www.genome.jp/kegg/">Kyoto Encyclopedia of Genes and Genomes (KEGG)</a> pathway enrichment analysis <span class="citation">(Kanehisa and Goto 2000; Subramanian et al. 2005)</span>, which leverages the functionality of the <a href="https://bioconductor.org/packages/clusterProfiler/">clusterProfiler</a> package <span class="citation">(Yu et al. 2012)</span>. This workflow is described in detail in our workflow paper, and is outside the scope of this advanced use vignette. For more information on functional analysis, consult the <a href="https://bioconductor.org/packages/clusterProfiler/">clusterProfiler</a> vignette or Stephen Turner’s excellent <a href="https://stephenturner.github.io/deseq-to-fgsea/">DESeq2 to fgsea workflow</a>.</p>
</div>
<div id="r-session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#r-session-information" class="anchor"></a>R session information</h1>
<pre class="sourceCode r"><code class="sourceCode r">utils::<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/sessionInfo">sessionInfo</a></span>()</code></pre>
<pre><code>## R version 3.5.2 (2018-12-20)
## Platform: x86_64-redhat-linux-gnu (64-bit)
## Running under: Red Hat Enterprise Linux
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib64/R/lib/libRblas.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    datasets  utils     grDevices graphics  stats    
## [8] methods   base     
## 
## other attached packages:
##  [1] hexbin_1.27.2               bcbioRNASeq_0.3.17         
##  [3] SummarizedExperiment_1.12.0 DelayedArray_0.8.0         
##  [5] BiocParallel_1.16.6         matrixStats_0.54.0         
##  [7] Biobase_2.42.0              GenomicRanges_1.34.0       
##  [9] GenomeInfoDb_1.18.2         IRanges_2.16.0             
## [11] S4Vectors_0.20.1            BiocGenerics_0.28.0        
## [13] BiocStyle_2.10.0           
## 
## loaded via a namespace (and not attached):
##   [1] basejump_0.10.5               freerange_0.1.3              
##   [3] R.utils_2.8.0                 tidyselect_0.2.5             
##   [5] RSQLite_2.1.1                 AnnotationDbi_1.44.0         
##   [7] htmlwidgets_1.3               grid_3.5.2                   
##   [9] munsell_0.5.0                 preprocessCore_1.44.0        
##  [11] withr_2.1.2                   colorspace_1.4-1             
##  [13] knitr_1.22                    rstudioapi_0.10              
##  [15] SingleCellExperiment_1.4.1    rdrop2_0.8.1                 
##  [17] acidplots_0.2.1               labeling_0.3                 
##  [19] lasso2_1.2-20                 tximport_1.10.1              
##  [21] GenomeInfoDbData_1.2.0        mnormt_1.5-5                 
##  [23] bit64_0.9-7                   pheatmap_1.0.12              
##  [25] rprojroot_1.3-2               Matrix.utils_0.9.7           
##  [27] generics_0.0.2                xfun_0.6                     
##  [29] R6_2.4.0                      locfit_1.5-9.1               
##  [31] AnnotationFilter_1.6.0        bitops_1.0-6                 
##  [33] reshape_0.8.8                 assertthat_0.2.1             
##  [35] promises_1.0.1                scales_1.0.0                 
##  [37] nnet_7.3-12                   gtable_0.3.0                 
##  [39] affy_1.60.0                   ensembldb_2.6.8              
##  [41] rlang_0.3.4                   genefilter_1.64.0            
##  [43] GlobalOptions_0.1.0           splines_3.5.2                
##  [45] rtracklayer_1.42.2            lazyeval_0.2.2               
##  [47] acepack_1.4.1                 broom_0.5.2                  
##  [49] checkmate_1.9.1               BiocManager_1.30.4           
##  [51] yaml_2.2.0                    reshape2_1.4.3               
##  [53] GenomicFeatures_1.34.8        backports_1.1.4              
##  [55] httpuv_1.5.1                  syntactic_0.1.8              
##  [57] Hmisc_4.2-0                   tools_3.5.2                  
##  [59] bookdown_0.9                  psych_1.8.12                 
##  [61] logging_0.9-107               ggplot2_3.1.1                
##  [63] affyio_1.52.0                 RColorBrewer_1.1-2           
##  [65] goalie_0.2.15                 ggdendro_0.1-20              
##  [67] sessioninfo_1.1.1             Rcpp_1.0.1                   
##  [69] plyr_1.8.4                    base64enc_0.1-3              
##  [71] progress_1.2.0                zlibbioc_1.28.0              
##  [73] purrr_0.3.2                   RCurl_1.95-4.12              
##  [75] prettyunits_1.0.2             rpart_4.1-15                 
##  [77] GetoptLong_0.1.7              cowplot_0.9.4                
##  [79] haven_2.1.0                   grr_0.9.5                    
##  [81] ggrepel_0.8.0                 cluster_2.0.8                
##  [83] fs_1.2.7                      magrittr_1.5                 
##  [85] data.table_1.12.2             openxlsx_4.1.0               
##  [87] circlize_0.4.6                ProtGenerics_1.14.0          
##  [89] hms_0.4.2                     mime_0.6                     
##  [91] evaluate_0.13                 bioverbs_0.1.14              
##  [93] xtable_1.8-4                  XML_3.98-1.19                
##  [95] rio_0.5.16                    readxl_1.3.1                 
##  [97] gridExtra_2.3                 shape_1.4.4                  
##  [99] compiler_3.5.2                biomaRt_2.38.0               
## [101] tibble_2.1.1                  crayon_1.3.4                 
## [103] transformer_0.1.9             R.oo_1.22.0                  
## [105] htmltools_0.3.6               later_0.8.0                  
## [107] Formula_1.2-3                 tidyr_0.8.3                  
## [109] geneplotter_1.60.0            DBI_1.0.0                    
## [111] ComplexHeatmap_1.20.0         MASS_7.3-51.3                
## [113] Matrix_1.2-17                 readr_1.3.1                  
## [115] brio_0.1.19                   cli_1.1.0                    
## [117] vsn_3.50.0                    R.methodsS3_1.7.1            
## [119] forcats_0.4.0                 pkgconfig_2.0.2              
## [121] pkgdown_1.3.0                 GenomicAlignments_1.18.1     
## [123] foreign_0.8-71                xml2_1.2.0                   
## [125] roxygen2_6.1.1                annotate_1.60.1              
## [127] XVector_0.22.0                stringr_1.4.0                
## [129] digest_0.6.18                 ConsensusClusterPlus_1.46.0  
## [131] Biostrings_2.50.2             cellranger_1.1.0             
## [133] rmarkdown_1.12                htmlTable_1.13.1             
## [135] edgeR_3.24.3                  curl_3.3                     
## [137] shiny_1.3.2                   Rsamtools_1.34.1             
## [139] commonmark_1.7                rjson_0.2.20                 
## [141] nlme_3.1-139                  jsonlite_1.6                 
## [143] DEGreport_1.18.1              desc_1.2.0                   
## [145] limma_3.38.3                  pillar_1.3.1                 
## [147] lattice_0.20-38               Nozzle.R1_1.1-1              
## [149] httr_1.4.0                    survival_2.44-1.1            
## [151] interactiveDisplayBase_1.20.0 glue_1.3.1                   
## [153] zip_2.0.1                     bit_1.1-14                   
## [155] stringi_1.4.3                 bcbioBase_0.6.2              
## [157] blob_1.1.1                    DESeq2_1.22.2                
## [159] AnnotationHub_2.14.5          latticeExtra_0.6-28          
## [161] memoise_1.1.0                 dplyr_0.8.0.1</code></pre>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<p>The papers and software cited in our workflows are available as a <a href="https://paperpile.com/shared/e1q8fn">shared library</a> on <a href="https://paperpile.com/">Paperpile</a>.</p>
<div class="references">
<p>Bray, Nicolas L, Harold Pimentel, P<span>á</span>ll Melsted, and Lior Pachter. 2016. “Near-Optimal Probabilistic RNA-seq Quantification.” <em>Nat. Biotechnol.</em> 34 (5) (May): 525–527. doi:<a href="http://dx.doi.org/10.1038/nbt.3519">10.1038/nbt.3519</a>. <a href="http://dx.doi.org/10.1038/nbt.3519">http://dx.doi.org/10.1038/nbt.3519</a>.</p>
<p>Craciun, Florin L, Vanesa Bijol, Amrendra K Ajay, Poornima Rao, Ramya K Kumar, John Hutchinson, Oliver Hofmann, et al. 2016. “RNA Sequencing Identifies Novel Translational Biomarkers of Kidney Fibrosis.” <em>J. Am. Soc. Nephrol.</em> 27 (6) (June): 1702–1713. doi:<a href="http://dx.doi.org/10.1681/ASN.2015020225">10.1681/ASN.2015020225</a>. <a href="http://dx.doi.org/10.1681/ASN.2015020225">http://dx.doi.org/10.1681/ASN.2015020225</a>.</p>
<p>Dobin, Alexander, and Thomas R Gingeras. 2016. “Optimizing RNA-Seq Mapping with STAR.” <em>Methods Mol. Biol.</em> 1415: 245–262. doi:<a href="http://dx.doi.org/10.1007/978-1-4939-3572-7%5C_13">10.1007/978-1-4939-3572-7\_13</a>. <a href="http://dx.doi.org/10.1007/978-1-4939-3572-7_13">http://dx.doi.org/10.1007/978-1-4939-3572-7_13</a>.</p>
<p>Dobin, Alexander, Carrie A Davis, Felix Schlesinger, Jorg Drenkow, Chris Zaleski, Sonali Jha, Philippe Batut, Mark Chaisson, and Thomas R Gingeras. 2013. “STAR: ultrafast Universal RNA-seq Aligner.” <em>Bioinformatics</em> 29 (1) (January): 15–21. doi:<a href="http://dx.doi.org/10.1093/bioinformatics/bts635">10.1093/bioinformatics/bts635</a>. <a href="http://dx.doi.org/10.1093/bioinformatics/bts635">http://dx.doi.org/10.1093/bioinformatics/bts635</a>.</p>
<p>Huber, Wolfgang, Vincent J Carey, Robert Gentleman, Simon Anders, Marc Carlson, Benilton S Carvalho, Hector Corrada Bravo, et al. 2015. “Orchestrating High-Throughput Genomic Analysis with Bioconductor.” <em>Nat. Methods</em> 12 (2) (February): 115–121. doi:<a href="http://dx.doi.org/10.1038/nmeth.3252">10.1038/nmeth.3252</a>. <a href="http://dx.doi.org/10.1038/nmeth.3252">http://dx.doi.org/10.1038/nmeth.3252</a>.</p>
<p>Kanehisa, M, and S Goto. 2000. “KEGG: kyoto Encyclopedia of Genes and Genomes.” <em>Nucleic Acids Res.</em> 28 (1) (January): 27–30. <a href="https://www.ncbi.nlm.nih.gov/pubmed/10592173">https://www.ncbi.nlm.nih.gov/pubmed/10592173</a>.</p>
<p>Kim, Daehwan, Ben Langmead, and Steven L Salzberg. 2015. “HISAT: a Fast Spliced Aligner with Low Memory Requirements.” <em>Nat. Methods</em> 12 (4) (April): 357–360. doi:<a href="http://dx.doi.org/10.1038/nmeth.3317">10.1038/nmeth.3317</a>. <a href="http://dx.doi.org/10.1038/nmeth.3317">http://dx.doi.org/10.1038/nmeth.3317</a>.</p>
<p>Lawrence, Michael, Wolfgang Huber, Herv<span>é</span> Pag<span>è</span>s, Patrick Aboyoun, Marc Carlson, Robert Gentleman, Martin T Morgan, and Vincent J Carey. 2013. “Software for Computing and Annotating Genomic Ranges.” <em>PLoS Comput. Biol.</em> 9 (8) (August): e1003118. doi:<a href="http://dx.doi.org/10.1371/journal.pcbi.1003118">10.1371/journal.pcbi.1003118</a>. <a href="http://dx.doi.org/10.1371/journal.pcbi.1003118">http://dx.doi.org/10.1371/journal.pcbi.1003118</a>.</p>
<p>Liao, Yang, Gordon K Smyth, and Wei Shi. 2014. “featureCounts: an Efficient General Purpose Program for Assigning Sequence Reads to Genomic Features.” <em>Bioinformatics</em> 30 (7) (April): 923–930. doi:<a href="http://dx.doi.org/10.1093/bioinformatics/btt656">10.1093/bioinformatics/btt656</a>. <a href="http://dx.doi.org/10.1093/bioinformatics/btt656">http://dx.doi.org/10.1093/bioinformatics/btt656</a>.</p>
<p>Love, Michael I, Wolfgang Huber, and Simon Anders. 2014. “Moderated Estimation of Fold Change and Dispersion for RNA-seq Data with DESeq2.” <em>Genome Biol.</em> 15 (12): 550. doi:<a href="http://dx.doi.org/10.1186/s13059-014-0550-8">10.1186/s13059-014-0550-8</a>. <a href="http://dx.doi.org/10.1186/s13059-014-0550-8">http://dx.doi.org/10.1186/s13059-014-0550-8</a>.</p>
<p>Patro, Rob, Geet Duggal, Michael I Love, Rafael A Irizarry, and Carl Kingsford. 2017. “Salmon Provides Fast and Bias-Aware Quantification of Transcript Expression.” <em>Nat. Methods</em> 14 (4) (April): 417–419. doi:<a href="http://dx.doi.org/10.1038/nmeth.4197">10.1038/nmeth.4197</a>. <a href="http://dx.doi.org/10.1038/nmeth.4197">http://dx.doi.org/10.1038/nmeth.4197</a>.</p>
<p>Patro, Rob, Stephen M Mount, and Carl Kingsford. 2014. “Sailfish Enables Alignment-Free Isoform Quantification from RNA-seq Reads Using Lightweight Algorithms.” <em>Nat. Biotechnol.</em> 32 (5) (May): 462–464. doi:<a href="http://dx.doi.org/10.1038/nbt.2862">10.1038/nbt.2862</a>. <a href="http://dx.doi.org/10.1038/nbt.2862">http://dx.doi.org/10.1038/nbt.2862</a>.</p>
<p>Soneson, Charlotte, Michael I Love, and Mark D Robinson. 2016. “Differential Analyses for RNA-seq: transcript-Level Estimates Improve Gene-Level Inferences.” <em>F1000Res.</em> 4 (December). doi:<a href="http://dx.doi.org/10.12688/f1000research.7563.1">10.12688/f1000research.7563.1</a>. <a href="https://f1000research.com/articles/4-1521/v1/pdf">https://f1000research.com/articles/4-1521/v1/pdf</a>.</p>
<p>Steinbaugh, Michael J, Lorena Pantano, Rory D Kirchner, Victor Barrera, Brad A Chapman, Mary E Piper, Meeta Mistry, et al. 2018. “bcbioRNASeq: R Package for Bcbio RNA-seq Analysis.” <em>F1000Res.</em> 6 (June). doi:<a href="http://dx.doi.org/10.12688/f1000research.12093.2">10.12688/f1000research.12093.2</a>. <a href="https://f1000research.com/articles/6-1976/v2/pdf">https://f1000research.com/articles/6-1976/v2/pdf</a>.</p>
<p>Subramanian, Aravind, Pablo Tamayo, Vamsi K Mootha, Sayan Mukherjee, Benjamin L Ebert, Michael A Gillette, Amanda Paulovich, et al. 2005. “Gene Set Enrichment Analysis: a Knowledge-Based Approach for Interpreting Genome-Wide Expression Profiles.” <em>Proc. Natl. Acad. Sci. U. S. A.</em> 102 (43) (October): 15545–15550. doi:<a href="http://dx.doi.org/10.1073/pnas.0506580102">10.1073/pnas.0506580102</a>. <a href="http://dx.doi.org/10.1073/pnas.0506580102">http://dx.doi.org/10.1073/pnas.0506580102</a>.</p>
<p>Yu, Guangchuang, Li-Gen Wang, Yanyan Han, and Qing-Yu He. 2012. “clusterProfiler: an R Package for Comparing Biological Themes Among Gene Clusters.” <em>OMICS</em> 16 (5) (May): 284–287. doi:<a href="http://dx.doi.org/10.1089/omi.2011.0118">10.1089/omi.2011.0118</a>. <a href="http://dx.doi.org/10.1089/omi.2011.0118">http://dx.doi.org/10.1089/omi.2011.0118</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li>
<a href="#loading-bcbio-data">Loading bcbio data</a><ul class="nav nav-pills nav-stacked">
<li><a href="#upload-directory">Upload directory</a></li>
      <li><a href="#counts-level">Counts level</a></li>
      <li><a href="#expression-callers">Expression callers</a></li>
      <li><a href="#sample-selection-and-metadata">Sample selection and metadata</a></li>
      <li><a href="#genome-annotations">Genome annotations</a></li>
      <li><a href="#variance-stabilization">Variance stabilization</a></li>
      </ul>
</li>
      <li><a href="#use-case-dataset">Use case dataset</a></li>
      <li><a href="#sample-metadata">Sample metadata</a></li>
      <li><a href="#interesting-groups">Interesting groups</a></li>
      <li>
<a href="#quality-control">Quality control</a><ul class="nav nav-pills nav-stacked">
<li><a href="#read-counts">Read counts</a></li>
      <li><a href="#mapping-rates">Mapping rates</a></li>
      <li><a href="#rrna-mapping-rate">rRNA mapping rate</a></li>
      <li><a href="#bias">5’-&gt;3’ bias</a></li>
      <li><a href="#gene-distributions">Gene distributions</a></li>
      <li><a href="#variance-stabilization-1">Variance stabilization</a></li>
      <li><a href="#sample-similarity">Sample similarity</a></li>
      </ul>
</li>
      <li><a href="#saving-counts">Saving counts</a></li>
      <li><a href="#differential-expression">Differential expression</a></li>
      <li><a href="#functional-analysis">Functional analysis</a></li>
      <li><a href="#r-session-information">R session information</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://steinbaugh.com/">Michael Steinbaugh</a>, <a href="https://lpantano.github.io/">Lorena Pantano</a>, <a href="https://github.com/roryk/">Rory Kirchner</a>, <a href="https://github.com/vbarrera/">Victor Barrera</a>, <a href="https://github.com/chapmanb/">Brad Chapman</a>, <a href="https://github.com/marypiper/">Mary Piper</a>, <a href="https://github.com/mistrm82/">Meeta Mistry</a>, <a href="https://github.com/rkhetani/">Radhika Khetani</a>, <a href="https://github.com/KayleighRutherford/">Kayleigh Rutherford</a>, <a href="https://github.com/ohofmann/">Oliver Hofmann</a>, <a href="https://github.com/jnhutchinson/">John Hutchinson</a>, <a href="https://github.com/sjhosui/">Shannan Ho Sui</a>, <a href="https://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
