<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Alpha Level Cutoff Summary Statistics — alphaSummary • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  <link href="../extra.css" rel="stylesheet">
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bcbioRNASeq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Alpha Level Cutoff Summary Statistics</h1>
    </div>

    
    <p>Alpha Level Cutoff Summary Statistics</p>
    

    <pre class="usage"><span class='fu'>alphaSummary</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S4 method for DESeqDataSet</span>
<span class='fu'>alphaSummary</span>(<span class='no'>object</span>, <span class='kw'>alpha</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='fl'>0.1</span>, <span class='fl'>0.05</span>, <span class='fl'>0.01</span>,
  <span class='fl'>0.001</span>, <span class='fl'>1e-06</span>), <span class='kw'>caption</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>...</span>)

<span class='co'># S4 method for bcbioRNASeq</span>
<span class='fu'>alphaSummary</span>(<span class='no'>object</span>, <span class='kw'>alpha</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='fl'>0.1</span>, <span class='fl'>0.05</span>, <span class='fl'>0.01</span>, <span class='fl'>0.001</span>,
  <span class='fl'>1e-06</span>), <span class='kw'>caption</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p><em>Optional.</em> Passthrough arguments to <code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/results'>DESeq2::results()</a></code>. Use
either <code>contrast</code> or <code>name</code> arguments to define the desired contrast.</p></td>
    </tr>
    <tr>
      <th>alpha</th>
      <td><p>Numeric vector of multiple alpha cutoffs.</p></td>
    </tr>
    <tr>
      <th>caption</th>
      <td><p><em>Optional.</em> Character string to use as a caption.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>kable</code>.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p><code>bcbioRNASeq</code> class does currently support contrast definitions, since
the object contains an internal <code>DESeqDataSet</code> with an empty design
formula.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p><code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/results'>DESeq2::results()</a></code>.</p>
<p>Other Differential Expression Functions: <code>contrastName</code>,
  <code>plotDEGHeatmap</code>, <code>plotDEGPCA</code>,
  <code><a href='plotMA.html'>plotMA</a></code>, <code>plotVolcano</code>,
  <code>resultsTables</code></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>load</span>(<span class='fu'>system.file</span>(<span class='st'>"extdata/bcb_small.rda"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"bcbioRNASeq"</span>))
<span class='fu'>load</span>(<span class='fu'>system.file</span>(<span class='st'>"extdata/dds_small.rda"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"bcbioRNASeq"</span>))

<span class='co'># bcbioRNASeq ====</span>
<span class='fu'>alphaSummary</span>(<span class='no'>bcb_small</span>)</div><div class='output co'>#&gt; <span class='message'>using just counts from tximport</span></div><div class='output co'>#&gt; <span class='message'>estimating size factors</span></div><div class='output co'>#&gt; <span class='message'>estimating dispersions</span></div><div class='output co'>#&gt; <span class='message'>gene-wise dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>mean-dispersion relationship</span></div><div class='output co'>#&gt; <span class='message'>final dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>fitting model and testing</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Empty design formula detected</span></div><div class='output co'>#&gt; 
#&gt; 
#&gt; |               |0.1               |0.05              |0.01              |0.001             |1e-06             |
#&gt; |:--------------|:-----------------|:-----------------|:-----------------|:-----------------|:-----------------|
#&gt; |LFC &gt; 0 (up)   |494, 99%          |494, 99%          |494, 99%          |494, 99%          |494, 99%          |
#&gt; |LFC &lt; 0 (down) |0, 0%             |0, 0%             |0, 0%             |0, 0%             |0, 0%             |
#&gt; |outliers       |6, 1.2%           |6, 1.2%           |6, 1.2%           |6, 1.2%           |6, 1.2%           |
#&gt; |low counts     |0, 0%             |0, 0%             |0, 0%             |0, 0%             |0, 0%             |
#&gt; |cutoff         |(mean count &lt; 28) |(mean count &lt; 28) |(mean count &lt; 28) |(mean count &lt; 28) |(mean count &lt; 28) |</div><div class='input'>
<span class='co'># DESeqDataSet ====</span>
<span class='fu'>resultsNames</span>(<span class='no'>dds_small</span>)</div><div class='output co'>#&gt; [1] "Intercept"                       "treatment_folic_acid_vs_control"</div><div class='input'><span class='fu'>alphaSummary</span>(<span class='no'>dds_small</span>, <span class='kw'>contrast</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"treatment"</span>, <span class='st'>"folic_acid"</span>, <span class='st'>"control"</span>))</div><div class='output co'>#&gt; 
#&gt; 
#&gt; |               |0.1               |0.05              |0.01              |0.001              |1e-06             |
#&gt; |:--------------|:-----------------|:-----------------|:-----------------|:------------------|:-----------------|
#&gt; |LFC &gt; 0 (up)   |131, 26%          |112, 22%          |77, 15%           |40, 8%             |21, 4.2%          |
#&gt; |LFC &lt; 0 (down) |146, 29%          |128, 26%          |104, 21%          |67, 13%            |16, 3.2%          |
#&gt; |outliers       |6, 1.2%           |6, 1.2%           |6, 1.2%           |6, 1.2%            |6, 1.2%           |
#&gt; |low counts     |0, 0%             |0, 0%             |0, 0%             |30, 6%             |0, 0%             |
#&gt; |cutoff         |(mean count &lt; 28) |(mean count &lt; 28) |(mean count &lt; 28) |(mean count &lt; 432) |(mean count &lt; 28) |</div><div class='input'><span class='fu'>alphaSummary</span>(<span class='no'>dds_small</span>, <span class='kw'>name</span> <span class='kw'>=</span> <span class='st'>"treatment_folic_acid_vs_control"</span>)</div><div class='output co'>#&gt; 
#&gt; 
#&gt; |               |0.1               |0.05              |0.01              |0.001              |1e-06             |
#&gt; |:--------------|:-----------------|:-----------------|:-----------------|:------------------|:-----------------|
#&gt; |LFC &gt; 0 (up)   |131, 26%          |112, 22%          |77, 15%           |40, 8%             |21, 4.2%          |
#&gt; |LFC &lt; 0 (down) |146, 29%          |128, 26%          |104, 21%          |67, 13%            |16, 3.2%          |
#&gt; |outliers       |6, 1.2%           |6, 1.2%           |6, 1.2%           |6, 1.2%            |6, 1.2%           |
#&gt; |low counts     |0, 0%             |0, 0%             |0, 0%             |30, 6%             |0, 0%             |
#&gt; |cutoff         |(mean count &lt; 28) |(mean count &lt; 28) |(mean count &lt; 28) |(mean count &lt; 432) |(mean count &lt; 28) |</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh, Lorena Patano

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io'>Lorena Pantano</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/mistrm82'>Meeta Mistry</a>, <a href='https://github.com/rkhetani'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson'>John Hutchinson</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
