<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>bcbio RNA-Seq Data Set — bcbioRNASeq • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="bcbio RNA-Seq Data Set — bcbioRNASeq" />

<meta property="og:description" content="Simply point to the final upload directory generated by bcbio,
and this generator function will take care of the rest." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioRNASeq</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bcbioRNASeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>bcbio RNA-Seq Data Set</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioRNASeq/blob/master/R/bcbioRNASeq-generator.R'><code>R/bcbioRNASeq-generator.R</code></a></small>
    <div class="hidden name"><code>bcbioRNASeq.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Simply point to the final upload directory generated by bcbio,
and this generator function will take care of the rest.</p>
    
    </div>

    <pre class="usage"><span class='fu'>bcbioRNASeq</span>(<span class='no'>uploadDir</span>, <span class='kw'>level</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"genes"</span>, <span class='st'>"transcripts"</span>),
  <span class='kw'>caller</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"salmon"</span>, <span class='st'>"kallisto"</span>, <span class='st'>"sailfish"</span>, <span class='st'>"star"</span>, <span class='st'>"hisat2"</span>),
  <span class='kw'>samples</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>censorSamples</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>sampleMetadataFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>organism</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>genomeBuild</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>ensemblRelease</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>gffFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>transgeneNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>spikeNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>countsFromAbundance</span> <span class='kw'>=</span> <span class='st'>"lengthScaledTPM"</span>, <span class='kw'>vst</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>rlog</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>interestingGroups</span> <span class='kw'>=</span> <span class='st'>"sampleName"</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>uploadDir</th>
      <td><p><code>string</code>. Path to final upload directory. This path is set
when running "<code>bcbio_nextgen -w template</code>".</p></td>
    </tr>
    <tr>
      <th>level</th>
      <td><p><code>string</code>. Import counts at gene level ("<code>genes</code>"; <em>default</em>) or
transcript level ("<code>transcripts</code>"; <em>advanced use</em>). Only
tximport-compatible callers (e.g. salmon, kallisto, sailfish) can be loaded
at transcript level. Aligned counts from featureCounts-compatible callers
(e.g. STAR, HISAT2) can only be loaded at gene level.</p></td>
    </tr>
    <tr>
      <th>caller</th>
      <td><p><code>string</code>. Expression caller:</p><ul>
<li><p>"<code>salmon</code>" (<em>default</em>): <a href='https://combine-lab.github.io/salmon'>Salmon</a>
alignment-free, quasi-mapped counts.</p></li>
<li><p>"<code>kallisto</code>". <a href='https://pachterlab.github.io/kallisto'>Kallisto</a>
alignment-free, pseudo-aligned counts.</p></li>
<li><p>"<code>sailfish</code>".
<a href='http://www.cs.cmu.edu/~ckingsf/software/sailfish'>Sailfish</a>
alignment-free, lightweight counts.</p></li>
<li><p>"<code>star</code>": <a href='https://github.com/alexdobin/STAR'>STAR</a>
(Spliced Transcripts Alignment to a Reference) aligned counts.</p></li>
<li><p>"<code>hisat2</code>": <a href='https://ccb.jhu.edu/software/hisat2'>HISAT2</a>
(Hierarchical Indexing for Spliced Alignment of Transcripts) graph-based
aligned counts.</p></li>
</ul></td>
    </tr>
    <tr>
      <th>samples</th>
      <td><p><code>character</code> or <code>NULL</code>. Specify a subset of samples to load.
The names must match the <code>description</code> specified in the bcbio YAML
metadata. If a <code>sampleMetadataFile</code> is provided, that will take priority
for sample selection. Typically this can be left unset.</p></td>
    </tr>
    <tr>
      <th>censorSamples</th>
      <td><p><code>character</code> or <code>NULL</code>. Samples to exclude from the
analysis.</p></td>
    </tr>
    <tr>
      <th>sampleMetadataFile</th>
      <td><p><code>string</code> or <code>NULL</code>. Custom metadata file containing
sample information. Otherwise defaults to sample metadata saved in the YAML
file. Remote URLs are supported. Typically this can be left unset.</p></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p><code>string</code> or <code>NULL</code>. Organism name. Use the full Latin name
(e.g. "Homo sapiens"), since this will be input downstream to AnnotationHub
and ensembldb, unless <code>gffFile</code> is set. If left <code>NULL</code> (<em>not recommended</em>),
the function call will skip loading gene/transcript-level annotations into
<code>rowRanges()</code>. This can be useful for poorly annotation genomes or
experiments involving multiple genomes.</p></td>
    </tr>
    <tr>
      <th>genomeBuild</th>
      <td><p><code>string</code> or <code>NULL</code>. Ensembl genome build name (e.g.
"GRCh38"). This will be passed to AnnotationHub for <code>EnsDb</code> annotation
matching, unless <code>gffFile</code> is set.</p></td>
    </tr>
    <tr>
      <th>ensemblRelease</th>
      <td><p><code>scalar integer</code> or <code>NULL</code>. Ensembl release version. If
unset, defaults to current release, and does not typically need to be
user-defined. Passed to AnnotationHub for <code>EnsDb</code> annotation matching,
unless <code>gffFile</code> is set.</p></td>
    </tr>
    <tr>
      <th>gffFile</th>
      <td><p><code>string</code> or <code>NULL</code>. By default, we recommend leaving this
<code>NULL</code> for genomes that are supported on Ensembl. In this case, the row
annotations (<code>rowRanges()</code>) will be obtained automatically from Ensembl by
passing the <code>organism</code>, <code>genomeBuild</code>, and <code>ensemblRelease</code> arguments to
AnnotationHub and ensembldb. For a genome that is not supported on Ensembl
and/or AnnotationHub, a GFF/GTF (General Feature Format) file is required.
Generally, we recommend using a GTF (GFFv2) file here over a GFF3 file if
possible, although all GFF formats are supported. The function will
internally generate a <code>TxDb</code> containing transcript-to-gene mappings and
construct a <code>GRanges</code> object containing the genomic ranges (<code>rowRanges()</code>).</p></td>
    </tr>
    <tr>
      <th>transgeneNames</th>
      <td><p><code>character</code> or <code>NULL</code>. Vector indicating which
<code>assay()</code> rows denote transgenes (e.g. EGFP, TDTOMATO).</p></td>
    </tr>
    <tr>
      <th>spikeNames</th>
      <td><p><code>character</code> or <code>NULL</code>. Vector indicating which <code>assay()</code>
rows denote spike-in sequences (e.g. ERCCs).</p></td>
    </tr>
    <tr>
      <th>countsFromAbundance</th>
      <td><p><code>string</code>. Whether to generate estimated counts
using abundance estimates (<em>recommended by default</em>). <code>lengthScaledTPM</code> is
a suitable default, and counts are scaled using the average transcript
length over samples and then the library size. Refer to
<a href='http://www.rdocumentation.org/packages/tximport/topics/tximport'>tximport::tximport</a> for more information on this parameter, but it should
only ever be changed when loading some datasets at transcript level
(e.g. for DTU analsyis).</p></td>
    </tr>
    <tr>
      <th>vst</th>
      <td><p><code>boolean</code>. Calculate variance-stabilizing transformation using
<code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/varianceStabilizingTransformation'>DESeq2::varianceStabilizingTransformation()</a></code>. Recommended by default
for visualization.</p></td>
    </tr>
    <tr>
      <th>rlog</th>
      <td><p><code>boolean</code>. Calcualte regularized log transformation using
<code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/rlog'>DESeq2::rlog()</a></code>. This calculation is slow for large datasets and now
discouraged by default for visualization.</p></td>
    </tr>
    <tr>
      <th>interestingGroups</th>
      <td><p><code>character</code>. Groups of interest that define the
samples. If left unset, defaults to <code>sampleName</code>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>bcbioRNASeq</code>.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The <code>bcbioRNASeq()</code> generator function automatically imports RNA-seq counts,
metadata, and the program versions used from a <a href='https://bcbio-nextgen.readthedocs.io'>bcbio</a> RNA-seq run.</p>
    
    <h2 class="hasAnchor" id="sample-metadata"><a class="anchor" href="#sample-metadata"></a>Sample metadata</h2>

    
    <p>When loading a bcbio RNA-seq run, the sample metadata will be imported
automatically from the <code>project-summary.yaml</code> file in the final upload
directory. If you notice any typos in your metadata after completing the run,
these can be corrected by editing the YAML file.</p>
<p>Alternatively, you can pass in a sample metadata file into the
<code>bcbioRNASeq()</code> function call using the <code>sampleMetadataFile</code> argument. This
requires either a CSV or Excel spreadsheet.</p>
<p>The samples in the bcbio run must map to the <code>description</code> column. The values
provided in <code>description</code> must be unique. These values will be sanitized into
syntactically valid names (see <a href='http://www.rdocumentation.org/packages/basejump/topics/makeNames'>basejump::makeNames</a> for more information),
and assigned as the column names of the <code>bcbioRNASeq</code> object. The original
values are stored as the <code>sampleName</code> column in <code>colData()</code>, and are used for
all plotting functions. Do not attempt to set a <code>sampleID</code> column, as this is
used internally by the package.</p>
<p>Here is a minimal example of a properly formatted sample metadata file:</p>
<table class='table'>
<tr><td>description</td><td>genotype</td></tr>
<tr><td>sample1</td><td>wildtype</td></tr>
<tr><td>sample2</td><td>knockout</td></tr>
<tr><td>sample3</td><td>wildtype</td></tr>
<tr><td>sample4</td><td>knockout</td></tr>
</table>
    
    <h2 class="hasAnchor" id="valid-names"><a class="anchor" href="#valid-names"></a>Valid names</h2>

    
    <p>R is strict about values that are considered valid for use in <code>names()</code> and
<code>dimnames()</code> (i.e. <code>rownames()</code> and <code>colnames()</code>). Non-alphanumeric
characters, spaces, and <strong>dashes</strong> are not valid. Use either underscores or
periods in place of dashes when working in R. Also note that names should
<strong>not begin with a number</strong>, and will be prefixed with an <code>X</code> when sanitized.
Consult the documentation in the <code><a href='http://www.rdocumentation.org/packages/base/topics/make.names'>base::make.names()</a></code> function for more
information. We strongly recommend adhering to these conventions when
labeling samples, to help avoid unexpected downstream behavior in R due to
<code>dimnames()</code> mismatches.</p>
    
    <h2 class="hasAnchor" id="genome-build"><a class="anchor" href="#genome-build"></a>Genome build</h2>

    
    <p>Ensure that the organism and genome build used with bcio match correctly
here in the function call. In particular, for the legacy <em>Homo sapiens</em>
GRCh37/hg19 genome build, ensure that <code>genomeBuild = "GRCh37"</code>. Otherwise,
the genomic ranges set in <code>rowRanges()</code> will mismatch. It is recommended
for current projects that GRCh38/hg38 is used in place of GRCh37/hg19
if possible.</p>
    
    <h2 class="hasAnchor" id="deseq-"><a class="anchor" href="#deseq-"></a>DESeq2</h2>

    
    <p>DESeq2 is run automatically when <code>bcbioRNASeq()</code> is called. Internally, this
automatically slots normalized counts into <code>assays()</code>, and optionally
generates variance-stabilized <code>rlog</code> or <code>vst</code> counts, depending on the call.
When loading a dataset with a large number of samples (i.e. &gt; 50), we
recommend disabling the <code>rlog</code> transformation, since it can take a long time
to compute.</p>
    
    <h2 class="hasAnchor" id="remote-connections"><a class="anchor" href="#remote-connections"></a>Remote connections</h2>

    
    <p>When working on a local machine, it is possible to load bcbio run data over a
remote connection using <a href='https://github.com/osxfuse/osxfuse/wiki/SSHFS'>sshfs</a>. When loading a large number of samples, it
is preferable to call <code>bcbioRNASeq()</code> directly in R on the remote server, if
possible.</p>
    
    <h2 class="hasAnchor" id="automatic-metadata"><a class="anchor" href="#automatic-metadata"></a>Automatic metadata</h2>

    
    <p>The <code>metadata()</code> slot always contains:</p><ul>
<li><p>Object version.</p></li>
<li><p>bcbio data provenance information.</p></li>
<li><p>File paths and timestamps.</p></li>
<li><p>R session information.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><ul>
<li><p><code>.S4methods(class = "bcbioRNASeq")</code>.</p></li>
<li><p><code><a href='http://www.rdocumentation.org/packages/SummarizedExperiment/topics/RangedSummarizedExperiment-class'>SummarizedExperiment::SummarizedExperiment()</a></code>.</p></li>
<li><p><code><a href='http://www.rdocumentation.org/packages/methods/topics/new'>methods::initialize()</a></code>.</p></li>
<li><p><code><a href='http://www.rdocumentation.org/packages/methods/topics/validObject'>methods::validObject()</a></code>.</p></li>
<li><p><code><a href='http://www.rdocumentation.org/packages/BiocGenerics/topics/updateObject'>BiocGenerics::updateObject()</a></code>.</p></li>
</ul></div>
    <div class='dont-index'><p>Other S4 Generators: <code><a href='DESeqAnalysis.html'>DESeqAnalysis</a></code>,
  <code><a href='DESeqResultsTables.html'>DESeqResultsTables</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>uploadDir</span> <span class='kw'>&lt;-</span> <span class='fu'>system.file</span>(<span class='st'>"extdata/bcbio"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"bcbioRNASeq"</span>)

<span class='co'>## Gene level</span>
<span class='no'>object</span> <span class='kw'>&lt;-</span> <span class='fu'>bcbioRNASeq</span>(
    <span class='kw'>uploadDir</span> <span class='kw'>=</span> <span class='no'>uploadDir</span>,
    <span class='kw'>level</span> <span class='kw'>=</span> <span class='st'>"genes"</span>,
    <span class='kw'>caller</span> <span class='kw'>=</span> <span class='st'>"salmon"</span>,
    <span class='kw'>organism</span> <span class='kw'>=</span> <span class='st'>"Mus musculus"</span>,
    <span class='kw'>ensemblRelease</span> <span class='kw'>=</span> <span class='fl'>87L</span>
)</div><div class='output co'>#&gt; <span class='message'>Dated project directory: 2018-03-18_GSE65267-merged</span></div><div class='output co'>#&gt; <span class='message'>2 sample(s) detected:</span>
#&gt; <span class='message'>control_rep1, control_rep2</span></div><div class='output co'>#&gt; <span class='message'>Reading project-summary.yaml.</span></div><div class='output co'>#&gt; <span class='message'>Data versions are missing.</span></div><div class='output co'>#&gt; <span class='message'>Reading bcbio-nextgen.log.</span></div><div class='output co'>#&gt; <span class='message'>Importing using `read_lines()`.</span></div><div class='output co'>#&gt; <span class='message'>Reading bcbio-nextgen-commands.log.</span></div><div class='output co'>#&gt; <span class='message'>Importing using `read_lines()`.</span></div><div class='output co'>#&gt; <span class='message'>Getting sample metadata from YAML.</span></div><div class='output co'>#&gt; <span class='message'>Getting sample metrics from YAML.</span></div><div class='output co'>#&gt; <span class='message'>Reading salmon counts using tximport 1.9.12.</span></div><div class='output co'>#&gt; <span class='message'>Reading from quant.sf files.</span></div><div class='output co'>#&gt; <span class='message'>Scaling using lengthScaledTPM.</span></div><div class='output co'>#&gt; <span class='message'>1 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = col_character(),</span>
#&gt; <span class='message'>  Length = col_integer(),</span>
#&gt; <span class='message'>  EffectiveLength = col_double(),</span>
#&gt; <span class='message'>  TPM = col_double(),</span>
#&gt; <span class='message'>  NumReads = col_double()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>2 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = col_character(),</span>
#&gt; <span class='message'>  Length = col_integer(),</span>
#&gt; <span class='message'>  EffectiveLength = col_double(),</span>
#&gt; <span class='message'>  TPM = col_double(),</span>
#&gt; <span class='message'>  NumReads = col_double()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='message'>summarizing abundance</span></div><div class='output co'>#&gt; <span class='message'>summarizing counts</span></div><div class='output co'>#&gt; <span class='message'>summarizing length</span></div><div class='output co'>#&gt; <span class='warning'>Warning: bcbio GTF file missing:</span>
#&gt; <span class='warning'>  /n/app/bcbio/dev/genomes/Mmusculus/GRCm38_90/rnaseq/ref-transcripts.gtf</span></div><div class='output co'>#&gt; <span class='message'>Using `makeGRangesFromEnsembl()` for annotations.</span></div><div class='output co'>#&gt; <span class='message'>Making GRanges from Ensembl with AnnotationHub 2.13.8 (2018-10-12).</span></div><div class='output co'>#&gt; <span class='message'>Ensembl 87 EnsDb for Mus Musculus</span></div><div class='output co'>#&gt; <span class='message'>  - AnnotationHub: NULL</span>
#&gt; <span class='message'>  - Organism: "Mus musculus"</span>
#&gt; <span class='message'>  - Build: "GRCm38"</span>
#&gt; <span class='message'>  - Release: "87"</span>
#&gt; <span class='message'>  - Level: "genes"</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneName, geneBiotype, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Arranging by geneID.</span></div><div class='output co'>#&gt; <span class='message'>Generating DESeqDataSet with DESeq2 1.21.23.</span></div><div class='output co'>#&gt; <span class='message'>Calculating normalized counts.</span></div><div class='output co'>#&gt; <span class='message'>Calculating transformations.</span></div><div class='output co'>#&gt; <span class='message'>using pre-existing size factors</span></div><div class='output co'>#&gt; <span class='message'>estimating dispersions</span></div><div class='output co'>#&gt; <span class='message'>gene-wise dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>mean-dispersion relationship</span></div><div class='output co'>#&gt; <span class='message'>final dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>fitting model and testing</span></div><div class='output co'>#&gt; <span class='message'>Applying variance-stabilizing transformation.</span></div><div class='output co'>#&gt; <span class='message'>Calculating FPKM.</span></div><div class='input'><span class='fu'>print</span>(<span class='no'>object</span>)</div><div class='output co'>#&gt; bcbioRNASeq 0.99.0
#&gt; http://bioinformatics.sph.harvard.edu/bcbioRNASeq
#&gt; citation("bcbioRNASeq")
#&gt; uploadDir: /Users/mike/git/bcbioRNASeq-private/inst/extdata/bcbio
#&gt; dates(2): [bcbio] 2018-03-18; [R] 2018-10-14
#&gt; level: genes
#&gt; caller: salmon
#&gt; organism: Mus musculus
#&gt; annotationHub: AH53222
#&gt; ensemblRelease: 87
#&gt; genomeBuild: GRCm38
#&gt; interestingGroups: sampleName
#&gt; class: RangedSummarizedExperiment  dim: 100 2  metadata(26): version level ... wd sessionInfo assays(6): counts tpm ... vst fpkm rownames(100): ENSMUSG00000000001 ENSMUSG00000000003 ...   ENSMUSG00000062661 ENSMUSG00000074340 rowData names(7): broadClass description ... geneName seqCoordSystem colnames(2): control_rep1 control_rep2 colData names(26): averageInsertSize averageReadLength ... tissue   treatment</div><div class='input'>
<span class='co'>## Transcript level</span>
<span class='no'>object</span> <span class='kw'>&lt;-</span> <span class='fu'>bcbioRNASeq</span>(
    <span class='kw'>uploadDir</span> <span class='kw'>=</span> <span class='no'>uploadDir</span>,
    <span class='kw'>level</span> <span class='kw'>=</span> <span class='st'>"transcripts"</span>,
    <span class='kw'>caller</span> <span class='kw'>=</span> <span class='st'>"salmon"</span>,
    <span class='kw'>organism</span> <span class='kw'>=</span> <span class='st'>"Mus musculus"</span>,
    <span class='kw'>ensemblRelease</span> <span class='kw'>=</span> <span class='fl'>87L</span>
)</div><div class='output co'>#&gt; <span class='message'>Dated project directory: 2018-03-18_GSE65267-merged</span></div><div class='output co'>#&gt; <span class='message'>2 sample(s) detected:</span>
#&gt; <span class='message'>control_rep1, control_rep2</span></div><div class='output co'>#&gt; <span class='message'>Reading project-summary.yaml.</span></div><div class='output co'>#&gt; <span class='message'>Data versions are missing.</span></div><div class='output co'>#&gt; <span class='message'>Reading bcbio-nextgen.log.</span></div><div class='output co'>#&gt; <span class='message'>Importing using `read_lines()`.</span></div><div class='output co'>#&gt; <span class='message'>Reading bcbio-nextgen-commands.log.</span></div><div class='output co'>#&gt; <span class='message'>Importing using `read_lines()`.</span></div><div class='output co'>#&gt; <span class='message'>Getting sample metadata from YAML.</span></div><div class='output co'>#&gt; <span class='message'>Getting sample metrics from YAML.</span></div><div class='output co'>#&gt; <span class='message'>Reading salmon counts using tximport 1.9.12.</span></div><div class='output co'>#&gt; <span class='message'>Reading from quant.sf files.</span></div><div class='output co'>#&gt; <span class='message'>Scaling using lengthScaledTPM.</span></div><div class='output co'>#&gt; <span class='message'>1 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = col_character(),</span>
#&gt; <span class='message'>  Length = col_integer(),</span>
#&gt; <span class='message'>  EffectiveLength = col_double(),</span>
#&gt; <span class='message'>  TPM = col_double(),</span>
#&gt; <span class='message'>  NumReads = col_double()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; <span class='message'>2 </span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  Name = col_character(),</span>
#&gt; <span class='message'>  Length = col_integer(),</span>
#&gt; <span class='message'>  EffectiveLength = col_double(),</span>
#&gt; <span class='message'>  TPM = col_double(),</span>
#&gt; <span class='message'>  NumReads = col_double()</span>
#&gt; <span class='message'>)</span></div><div class='output co'>#&gt; </div><div class='output co'>#&gt; <span class='warning'>Warning: bcbio GTF file missing:</span>
#&gt; <span class='warning'>  /n/app/bcbio/dev/genomes/Mmusculus/GRCm38_90/rnaseq/ref-transcripts.gtf</span></div><div class='output co'>#&gt; <span class='message'>Using `makeGRangesFromEnsembl()` for annotations.</span></div><div class='output co'>#&gt; <span class='message'>Making GRanges from Ensembl with AnnotationHub 2.13.8 (2018-10-12).</span></div><div class='output co'>#&gt; <span class='message'>Ensembl 87 EnsDb for Mus Musculus</span></div><div class='output co'>#&gt; <span class='message'>  - AnnotationHub: NULL</span>
#&gt; <span class='message'>  - Organism: "Mus musculus"</span>
#&gt; <span class='message'>  - Build: "GRCm38"</span>
#&gt; <span class='message'>  - Release: "87"</span>
#&gt; <span class='message'>  - Level: "transcripts"</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneName, transcriptBiotype, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Arranging by transcriptID.</span></div><div class='output co'>#&gt; <span class='error'>Error in checkClasses(object = rowData, expected = list(broadClass = "factor",     description = "factor", entrezID = c("AsIs", "list"), geneBiotype = "factor",     geneID = "character", geneName = "factor", seqCoordSystem = "factor"),     subset = TRUE): Class checks failed.</span>
#&gt; <span class='error'>Run `updateObject()` to update your object.</span>
#&gt; <span class='error'>    broadClass    description       entrezID    geneBiotype         geneID </span>
#&gt; <span class='error'>          TRUE           TRUE           TRUE           TRUE          FALSE </span>
#&gt; <span class='error'>      geneName seqCoordSystem </span>
#&gt; <span class='error'>          TRUE           TRUE </span></div><div class='input'><span class='fu'>print</span>(<span class='no'>object</span>)</div><div class='output co'>#&gt; bcbioRNASeq 0.99.0
#&gt; http://bioinformatics.sph.harvard.edu/bcbioRNASeq
#&gt; citation("bcbioRNASeq")
#&gt; uploadDir: /Users/mike/git/bcbioRNASeq-private/inst/extdata/bcbio
#&gt; dates(2): [bcbio] 2018-03-18; [R] 2018-10-14
#&gt; level: genes
#&gt; caller: salmon
#&gt; organism: Mus musculus
#&gt; annotationHub: AH53222
#&gt; ensemblRelease: 87
#&gt; genomeBuild: GRCm38
#&gt; interestingGroups: sampleName
#&gt; class: RangedSummarizedExperiment  dim: 100 2  metadata(26): version level ... wd sessionInfo assays(6): counts tpm ... vst fpkm rownames(100): ENSMUSG00000000001 ENSMUSG00000000003 ...   ENSMUSG00000062661 ENSMUSG00000074340 rowData names(7): broadClass description ... geneName seqCoordSystem colnames(2): control_rep1 control_rep2 colData names(26): averageInsertSize averageReadLength ... tissue   treatment</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#sample-metadata">Sample metadata</a></li>

      <li><a href="#valid-names">Valid names</a></li>

      <li><a href="#genome-build">Genome build</a></li>

      <li><a href="#deseq-">DESeq2</a></li>

      <li><a href="#remote-connections">Remote connections</a></li>

      <li><a href="#automatic-metadata">Automatic metadata</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh, Lorena Pantano, Rory Kirchner, Victor Barrera

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io'>Lorena Pantano</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/chapmanb'>Brad Chapman</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/mistrm82'>Meeta Mistry</a>, <a href='https://github.com/rkhetani'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson'>John Hutchinson</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

