<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Count Matrix Accessors — counts • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bcbioRNASeq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Count Matrix Accessors</h1>
    </div>

    
    <p>By default, <code>counts()</code> returns the raw counts. Normalized counts, including
transcripts per million (TPM) can be accessed using the <code>normalized</code>
argument.</p>
    

    <pre class="usage"><span class='co'># S4 method for bcbioRNADataSet</span>
<span class='fu'>counts</span>(<span class='no'>object</span>, <span class='kw'>normalized</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>normalized</th>
      <td><p>Select raw counts (<code>FALSE</code>), DESeq2 normalized counts
(<code>TRUE</code>), or additional normalization methods:</p><ul>
<li><p><code>tpm</code>: Transcripts per million.</p></li>
<li><p><code>tmm</code>: Trimmed mean of M-values (edgeR).</p></li>
<li><p><code>rlog</code>: Regularized log transformation (<code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/rlog'>DESeq2::rlog()</a></code>).</p></li>
<li><p><code>vst</code>: Variance stabilizing transformation
(<code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/varianceStabilizingTransformation'>DESeq2::varianceStabilizingTransformation()</a></code>).</p></li>
</ul></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>matrix.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>data</span>(<span class='no'>bcb</span>)
<span class='co'># Raw counts</span>
<span class='fu'>counts</span>(<span class='no'>bcb</span>, <span class='kw'>normalized</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>) <span class='kw'>%&gt;%</span> <span class='no'>glimpse</span></div><div class='output co'>#&gt;  num [1:505, 1:4] 11.611 399.73 1712.747 677.519 0.999 ...
#&gt;  - attr(*, "dimnames")=List of 2
#&gt;   ..$ : chr [1:505] "ENSMUSG00000002459" "ENSMUSG00000004768" "ENSMUSG00000005886" "ENSMUSG00000016918" ...
#&gt;   ..$ : chr [1:4] "group1_1" "group1_2" "group2_1" "group2_2"</div><div class='input'>
<span class='co'># DESeq2 normalized counts</span>
<span class='fu'>counts</span>(<span class='no'>bcb</span>, <span class='kw'>normalized</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>) <span class='kw'>%&gt;%</span> <span class='no'>glimpse</span></div><div class='output co'>#&gt;  num [1:505, 1:4] 13.2 440 1884.4 745.8 1.1 ...
#&gt;  - attr(*, "dimnames")=List of 2
#&gt;   ..$ : chr [1:505] "ENSMUSG00000002459" "ENSMUSG00000004768" "ENSMUSG00000005886" "ENSMUSG00000016918" ...
#&gt;   ..$ : chr [1:4] "group1_1" "group1_2" "group2_1" "group2_2"</div><div class='input'>
<span class='co'># TPM</span>
<span class='fu'>counts</span>(<span class='no'>bcb</span>, <span class='kw'>normalized</span> <span class='kw'>=</span> <span class='st'>"tpm"</span>) <span class='kw'>%&gt;%</span> <span class='no'>glimpse</span></div><div class='output co'>#&gt;  num [1:505, 1:4] 0.3396 5.7356 8.6836 7.005 0.0392 ...
#&gt;  - attr(*, "dimnames")=List of 2
#&gt;   ..$ : chr [1:505] "ENSMUSG00000002459" "ENSMUSG00000004768" "ENSMUSG00000005886" "ENSMUSG00000016918" ...
#&gt;   ..$ : chr [1:4] "group1_1" "group1_2" "group2_1" "group2_2"</div><div class='input'>
<span class='co'># TMM</span>
<span class='fu'>counts</span>(<span class='no'>bcb</span>, <span class='kw'>normalized</span> <span class='kw'>=</span> <span class='st'>"tmm"</span>) <span class='kw'>%&gt;%</span> <span class='no'>glimpse</span></div><div class='output co'>#&gt;  num [1:505, 1:4] 28.63 985.7 4223.48 1670.7 2.46 ...
#&gt;  - attr(*, "dimnames")=List of 2
#&gt;   ..$ : chr [1:505] "ENSMUSG00000002459" "ENSMUSG00000004768" "ENSMUSG00000005886" "ENSMUSG00000016918" ...
#&gt;   ..$ : chr [1:4] "group1_1" "group1_2" "group2_1" "group2_2"</div><div class='input'>
<span class='co'># rlog</span>
<span class='fu'>counts</span>(<span class='no'>bcb</span>, <span class='kw'>normalized</span> <span class='kw'>=</span> <span class='st'>"rlog"</span>) <span class='kw'>%&gt;%</span> <span class='no'>glimpse</span></div><div class='output co'>#&gt;  num [1:505, 1:4] 3.42 8.67 10.85 9.37 -1.71 ...
#&gt;  - attr(*, "dimnames")=List of 2
#&gt;   ..$ : chr [1:505] "ENSMUSG00000002459" "ENSMUSG00000004768" "ENSMUSG00000005886" "ENSMUSG00000016918" ...
#&gt;   ..$ : chr [1:4] "group1_1" "group1_2" "group2_1" "group2_2"
#&gt;  - attr(*, "betaPriorVar")= num 0.184
#&gt;  - attr(*, "intercept")= num [1:505, 1] 3.36 8.42 10.76 8.93 -1.72 ...</div><div class='input'>
<span class='co'># VST</span>
<span class='fu'>counts</span>(<span class='no'>bcb</span>, <span class='kw'>normalized</span> <span class='kw'>=</span> <span class='st'>"vst"</span>) <span class='kw'>%&gt;%</span> <span class='no'>glimpse</span></div><div class='output co'>#&gt;  num [1:505, 1:4] 6.45 9.08 10.96 9.73 5.95 ...
#&gt;  - attr(*, "dimnames")=List of 2
#&gt;   ..$ : chr [1:505] "ENSMUSG00000002459" "ENSMUSG00000004768" "ENSMUSG00000005886" "ENSMUSG00000016918" ...
#&gt;   ..$ : chr [1:4] "group1_1" "group1_2" "group2_1" "group2_2"</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Michael J. Steinbaugh, Lorena Pantano, Rory Kirchner, Victor Barrera, John Hutchinson, Harvard Chan Bioinformatics Core.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
