<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Count Matrix Accessors — counts • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Count Matrix Accessors — counts" />

<meta property="og:description" content="By default, returns the raw counts. Normalized counts in a variety of formats
can be accessed using the normalized argument." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioRNASeq</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bcbioRNASeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Count Matrix Accessors</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioRNASeq/blob/master/R/counts-methods.R'><code>R/counts-methods.R</code></a></small>
    <div class="hidden name"><code>counts.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>By default, returns the raw counts. Normalized counts in a variety of formats
can be accessed using the <code>normalized</code> argument.</p>
    
    </div>

    <pre class="usage"><span class='fu'>counts</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='fu'>counts</span>(<span class='no'>object</span>, <span class='no'>...</span>) <span class='kw'>&lt;-</span> <span class='no'>value</span>

<span class='co'># S4 method for bcbioRNASeq</span>
<span class='fu'>counts</span>(<span class='no'>object</span>, <span class='kw'>normalized</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    <tr>
      <th>value</th>
      <td><p>Value to assign.</p></td>
    </tr>
    <tr>
      <th>normalized</th>
      <td><p><code>string</code> or <code>boolean</code>. Which normalization method to apply:</p><ul>
<li><p><code>FALSE</code>: Raw counts.</p><ul>
<li><p>When using a <a href='https://bioconductor.org/packages/release/bioc/html/tximport.html'>tximport</a>-compatible caller, these are length scaled
by default (see <code>countsFromAbundance</code> argument).</p></li>
<li><p>When using a <a href='http://bioinf.wehi.edu.au/featureCounts/'>featureCounts</a>-compatible caller, these are <code>integer</code>.</p></li>
</ul></li>
</ul>

<p><a href='https://bioconductor.org/packages/release/bioc/html/tximport.html'>tximport</a> caller-specific normalizations:</p><ul>
<li><p><code>"tpm"</code>: <strong>T</strong>ranscripts <strong>p</strong>er <strong>m</strong>illion.</p></li>
</ul>

<p>Additional gene-level-specific normalizations:</p><ul>
<li><p><code>TRUE</code>: Size factor-adjusted counts.<br />
See <code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/sizeFactors'>DESeq2::sizeFactors()</a></code> for more information.</p></li>
<li><p><code>"vst"</code>: <strong>V</strong>ariance-<strong>s</strong>tabilizing <strong>t</strong>ransformation.<br />
Requires <code>vst = TRUE</code> to be set during <code><a href='bcbioRNASeq.html'>bcbioRNASeq()</a></code> call.<br />
See <code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/varianceStabilizingTransformation'>DESeq2::varianceStabilizingTransformation()</a></code> for more information.</p></li>
<li><p><code>"rlog"</code>: <strong>R</strong>egularized <strong>log</strong> transformation (log2).<br />
Requires <code>rlog = TRUE</code> to be set during <code><a href='bcbioRNASeq.html'>bcbioRNASeq()</a></code> call.<br />
See <code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/rlog'>DESeq2::rlog()</a></code> for more information.</p></li>
<li><p><code>"tmm"</code>: <strong>T</strong>rimmed <strong>m</strong>ean of <strong>M</strong>-values.<br />
See <code><a href='http://www.rdocumentation.org/packages/edgeR/topics/calcNormFactors'>edgeR::calcNormFactors()</a></code> for more information.</p></li>
<li><p><code>"rle"</code>: <strong>R</strong>elative <strong>l</strong>og <strong>e</strong>xpression transformation.</p></li>
<li><p><code>"fpkm"</code>: <strong>F</strong>ragments <strong>p</strong>er <strong>k</strong>ilobase per <strong>m</strong>illion mapped
fragments.<br />
Requires annotations in <code>rowRanges()</code> with defined <code>width</code>; otherwise,
will be skipped during the <code><a href='bcbioRNASeq.html'>bcbioRNASeq()</a></code> load call.<br />
See <code><a href='http://www.rdocumentation.org/packages/BiocGenerics/topics/start'>BiocGenerics::width()</a></code> and <code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/fpkm'>DESeq2::fpkm()</a></code> for more information.</p></li>
</ul></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>matrix</code>.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <ul>
<li><p>TMM: Robinson and Oshlack (2010).</p></li>
<li><p>RLE: Anders and Huber (2010).</p></li>
</ul>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><ul>
<li><p><code><a href='http://www.rdocumentation.org/packages/tximport/topics/tximport'>tximport::tximport()</a></code>.</p></li>
<li><p><code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/counts'>DESeq2::counts()</a></code>.</p></li>
<li><p><code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/sizeFactors'>DESeq2::sizeFactors()</a></code>.</p></li>
<li><p><code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/varianceStabilizingTransformation'>DESeq2::varianceStabilizingTransformation()</a></code>.</p></li>
<li><p><code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/rlog'>DESeq2::rlog()</a></code>.</p></li>
<li><p><code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/fpkm'>DESeq2::fpkm()</a></code>.</p></li>
<li><p><code><a href='http://www.rdocumentation.org/packages/edgeR/topics/calcNormFactors'>edgeR::calcNormFactors()</a></code>.</p></li>
<li><p><code><a href='tmm.html'>tmm()</a></code>, <code><a href='relativeLogExpression.html'>relativeLogExpression()</a></code>.</p></li>
</ul></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>data</span>(<span class='no'>bcb_small</span>)
<span class='fu'>counts</span>(<span class='no'>bcb_small</span>) <span class='kw'>%&gt;%</span> <span class='fu'>summary</span>()</div><div class='output co'>#&gt;   control_rep1       control_rep2       control_rep3       fa_day7_rep1     
#&gt;  Min.   :       0   Min.   :       0   Min.   :       0   Min.   :       1  
#&gt;  1st Qu.:     419   1st Qu.:     904   1st Qu.:     317   1st Qu.:     579  
#&gt;  Median :    1389   Median :    3397   Median :    1315   Median :    1725  
#&gt;  Mean   :   43568   Mean   :   86384   Mean   :   41167   Mean   :   39351  
#&gt;  3rd Qu.:    4692   3rd Qu.:   11303   3rd Qu.:    4125   3rd Qu.:    5373  
#&gt;  Max.   :11767239   Max.   :22627793   Max.   :11608372   Max.   :10623916  
#&gt;   fa_day7_rep2      fa_day7_rep3     
#&gt;  Min.   :      0   Min.   :       0  
#&gt;  1st Qu.:    501   1st Qu.:     769  
#&gt;  Median :   1557   Median :    1904  
#&gt;  Mean   :  34610   Mean   :   42438  
#&gt;  3rd Qu.:   4840   3rd Qu.:    5304  
#&gt;  Max.   :9676254   Max.   :12145225  </div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh, Lorena Pantano

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io'>Lorena Pantano</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/chapmanb'>Brad Chapman</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/mistrm82'>Meeta Mistry</a>, <a href='https://github.com/rkhetani'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson'>John Hutchinson</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

