<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>General arguments — general • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="General arguments — general" />

<meta property="og:description" content="General arguments" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioRNASeq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bcbioRNASeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>General arguments</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioRNASeq/blob/master/R/general.R'><code>R/general.R</code></a></small>
    <div class="hidden name"><code>general.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>General arguments</p>
    
    </div>

    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>i</th>
      <td><p>An integer or numeric scalar.</p></td>
    </tr>
    <tr>
      <th>value</th>
      <td><p>Value to assign.</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    <tr>
      <th>alpha</th>
      <td><p><code>scalar numeric</code>. Adjusted P value ("alpha") cutoff.</p></td>
    </tr>
    <tr>
      <th>color</th>
      <td><p><code>ggproto</code>/<code>ScaleDiscrete</code> or <code>NULL</code>. Desired ggplot2 color
scale. Must supply discrete values. When set to <code>NULL</code>, the default ggplot2
color palette will be used. If manual color definitions are desired, we
recommend using <code><a href='https://rdrr.io/pkg/ggplot2/man/scale_manual.html'>ggplot2::scale_color_manual()</a></code>.
To set the discrete color palette globally, use
<code><a href='https://rdrr.io/r/base/options.html'>options(bcbio.discrete.color = scale_color_viridis_d())</a></code>.</p></td>
    </tr>
    <tr>
      <th>counts</th>
      <td><p><code>matrix</code>. Normalized counts.</p></td>
    </tr>
    <tr>
      <th>dir</th>
      <td><p><code>string</code>. Local directory path.</p></td>
    </tr>
    <tr>
      <th>direction</th>
      <td><p><code>string</code>. Plot "<code>both</code>", "<code>up</code>", or "<code>down</code>" directions.</p></td>
    </tr>
    <tr>
      <th>fill</th>
      <td><p><code>ggproto</code>/<code>ScaleDiscrete</code> or <code>NULL</code>. Desired ggplot2 fill scale.
Must supply discrete values. When set to <code>NULL</code>, the default ggplot2 color
palette will be used. If manual color definitions are desired, we recommend
using <code><a href='https://rdrr.io/pkg/ggplot2/man/scale_manual.html'>ggplot2::scale_fill_manual()</a></code>.
To set the discrete fill palette globally, use
<code><a href='https://rdrr.io/r/base/options.html'>options(bcbio.discrete.fill = scale_fill_viridis_d())</a></code>.</p></td>
    </tr>
    <tr>
      <th>flip</th>
      <td><p><code>boolean</code>. Flip x and y axes. Recommended for quality control
plots containing many samples.</p></td>
    </tr>
    <tr>
      <th>gene2symbol</th>
      <td><p><code>data.frame</code>. Gene-to-symbol mappings. Must contain the
columns <code>geneID</code> and <code>geneName</code>.</p></td>
    </tr>
    <tr>
      <th>genes</th>
      <td><p><code>character</code>. Genes to include. These must match the rownames of
the object. It is best practice to use the stable gene identifiers from
Ensembl (e.g. "ENSG00000000003") and not the gene symbols.</p></td>
    </tr>
    <tr>
      <th>headerLevel</th>
      <td><p><code>scalar integer</code> (<code>1</code>-<code>7</code>). R Markdown header level.</p></td>
    </tr>
    <tr>
      <th>interestingGroups</th>
      <td><p><code>character</code>. Groups of interest that define the
samples. If left unset, defaults to <code>sampleName</code>.</p></td>
    </tr>
    <tr>
      <th>label</th>
      <td><p><code>boolean</code>. Superimpose sample text labels on the plot.</p></td>
    </tr>
    <tr>
      <th>legend</th>
      <td><p><code>boolean</code>. Show plot legend.</p></td>
    </tr>
    <tr>
      <th>lfcThreshold</th>
      <td><p><code>scalar numeric</code>. Log fold change ratio (base 2) cutoff
threshold.</p></td>
    </tr>
    <tr>
      <th>limit</th>
      <td><p><code>scalar numeric</code>. Threshold to denote on the plot, using a
dashed line.</p></td>
    </tr>
    <tr>
      <th>minCounts</th>
      <td><p><code>scalar integer</code>. Minimum number of counts per gene in the
counts matrix.</p></td>
    </tr>
    <tr>
      <th>normalized</th>
      <td><p><code>string</code>. Which normalization method to apply:</p><ul>
<li><p>"<code>tpm</code>": Transcripts per million (tximport).</p></li>
<li><p>"<code>tmm</code>": edgeR trimmed mean of M-values. Calculated on the fly.</p></li>
<li><p>"<code>rlog</code>": DESeq2 <strong>log2</strong> regularized log transformation.</p></li>
<li><p>"<code>vst</code>": DESeq2 <strong>log2</strong> variance stabilizing transformation.</p></li>
</ul></td>
    </tr>
    <tr>
      <th>ntop</th>
      <td><p><code>scalar integer</code>. Number of top genes to label.</p></td>
    </tr>
    <tr>
      <th>passLimit</th>
      <td><p><code>scalar numeric</code>. Threshold to plot pass color marker.</p></td>
    </tr>
    <tr>
      <th>pointColor</th>
      <td><p><code>string</code>. Default point color for the plot.</p></td>
    </tr>
    <tr>
      <th>results</th>
      <td><p><code>DESeqResults</code>.</p></td>
    </tr>
    <tr>
      <th>return</th>
      <td><p><code>string</code>. Object class to return. Uses <code><a href='https://rdrr.io/r/base/match.arg.html'>match.arg()</a></code> internally
and picks the first item in the vector by default.</p></td>
    </tr>
    <tr>
      <th>samples</th>
      <td><p><code>character</code> or <code>NULL</code>. Samples to include.</p></td>
    </tr>
    <tr>
      <th>sigPointColor</th>
      <td><p><code>character</code>. Color names for labeling upregulated and
downregulated genes. Also supports a character string for labeling DEGs
with the same color, regardless of direction.</p></td>
    </tr>
    <tr>
      <th>subtitle</th>
      <td><p><code>string</code> or <code>NULL</code>. Plot subtitle.</p></td>
    </tr>
    <tr>
      <th>title</th>
      <td><p><code>string</code> or <code>NULL</code>. Plot title.</p></td>
    </tr>
    <tr>
      <th>warnLimit</th>
      <td><p><code>scalar numeric</code>. Threshold to plot warning color marker.</p></td>
    </tr>
    <tr>
      <th>withDimnames</th>
      <td><p><code>boolean</code>. Whether dimnames should be applied to
extracted assay elements.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>No value.</p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
          </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io'>Lorena Pantano</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/chapmanb'>Brad Chapman</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/mistrm82'>Meeta Mistry</a>, <a href='https://github.com/rkhetani'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson'>John Hutchinson</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9000.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

