<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ bcbioRnaseq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bcbioRnaseq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRnaseq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9">
    <div class="page-header">
      <h1>
        Reference
        <small>version&nbsp;0.0.25</small>
      </h1>
    </div>

    <div class="contents">
      <table class="ref-index">

      <colgroup>
        <col class="alias" />
        <col class="title" />
      </colgroup>

      <tbody>
        <tr>
          <th colspan="2">
            <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
            <p class="section-desc"></p>
          </th>
        </tr>
        <tr>
          <!--  -->
          <td>
            <p><code><a href="aggregate_replicates.html">aggregate_replicates</a></code> </p>
          </td>
          <td><p>Aggregate Lane-Split Technical Replicates</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="alpha_summary.html">alpha_summary</a></code> </p>
          </td>
          <td><p>Print Summary Statistics of Alpha Level Cutoffs</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="bcbio.html">bcbio</a></code> <code><a href="bcbio.html">bcbio<-</a></code> </p>
          </td>
          <td><p><code>bcbioRNADataSet</code> Caller Slot Accessor</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p></p>
          </td>
          <td><p><code>bcbioRNADataSet</code></p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p></p>
          </td>
          <td><p><code>bcbioRNAResults</code></p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p></p>
          </td>
          <td><p>bcbioRnaseq</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="counts.html">counts</a></code> </p>
          </td>
          <td><p>Count Matrix Accessors</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="downloads.html">downloads</a></code> </p>
          </td>
          <td><p>File downloads for RMarkdown templates</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="gene2symbol.html">gene2symbol</a></code> </p>
          </td>
          <td><p>Gene Identifier to Symbol Conversion</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="load_run.html">load_run</a></code> </p>
          </td>
          <td><p>Load <code>bcbio</code> Run Data</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="melt_log10.html">melt_log10</a></code> </p>
          </td>
          <td><p>Melt Count Matrix to Long Format and log10 Transform</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="metadata_table.html">metadata_table</a></code> </p>
          </td>
          <td><p>Metadata Table</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="metrics.html">metrics</a></code> </p>
          </td>
          <td><p>Sample Metrics</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_5_3_bias.html">plot_5_3_bias</a></code> <code><a href="plot_5_3_bias.html">.plot_5_3_bias</a></code> </p>
          </td>
          <td><p>Plot 5'-&gt;3' Bias</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_correlation_heatmap.html">plot_correlation_heatmap</a></code> </p>
          </td>
          <td><p>Plot Correlation Heatmap</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_count_density.html">plot_count_density</a></code> <code><a href="plot_count_density.html">.plot_count_density</a></code> </p>
          </td>
          <td><p>Plot Count Density</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_counts_per_gene.html">plot_counts_per_gene</a></code> <code><a href="plot_counts_per_gene.html">.plot_counts_per_gene</a></code> </p>
          </td>
          <td><p>Plot Counts Per Gene</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_deg_heatmap.html">plot_deg_heatmap</a></code> <code><a href="plot_deg_heatmap.html">.plot_deg_heatmap</a></code> </p>
          </td>
          <td><p>Differentially Expressed Gene Heatmap</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_dispersion.html">plot_dispersion</a></code> </p>
          </td>
          <td><p>Plot Dispersion Estimates</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_exonic_mapping_rate.html">plot_exonic_mapping_rate</a></code> <code><a href="plot_exonic_mapping_rate.html">.plot_exonic_mapping_rate</a></code> </p>
          </td>
          <td><p>Plot Exonic Mapping Rate</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_gender_markers.html">plot_gender_markers</a></code> </p>
          </td>
          <td><p>Plot Sexually Dimorphic Gender Markers</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_gene_detection_saturation.html">plot_gene_detection_saturation</a></code> <code><a href="plot_gene_detection_saturation.html">.plot_gene_detection_saturation</a></code> </p>
          </td>
          <td><p>Plot Gene Detection Saturation</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_gene_heatmap.html">plot_gene_heatmap</a></code> <code><a href="plot_gene_heatmap.html">.plot_gene_heatmap</a></code> </p>
          </td>
          <td><p>Gene Heatmap</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_gene.html">plot_gene</a></code> </p>
          </td>
          <td><p>Plot Individual Genes</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_genes_detected.html">plot_genes_detected</a></code> <code><a href="plot_genes_detected.html">.plot_genes_detected</a></code> </p>
          </td>
          <td><p>Plot Genes Detected</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_intronic_mapping_rate.html">plot_intronic_mapping_rate</a></code> <code><a href="plot_intronic_mapping_rate.html">.plot_intronic_mapping_rate</a></code> </p>
          </td>
          <td><p>Plot Intronic Mapping Rate</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_ma.html">plot_ma</a></code> <code><a href="plot_ma.html">.plot_ma</a></code> </p>
          </td>
          <td><p>Plot Mean Average</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_mapped_reads.html">plot_mapped_reads</a></code> <code><a href="plot_mapped_reads.html">.plot_mapped_reads</a></code> </p>
          </td>
          <td><p>Plot Mapped Reads</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_mapping_rate.html">plot_mapping_rate</a></code> <code><a href="plot_mapping_rate.html">.plot_mapping_rate</a></code> </p>
          </td>
          <td><p>Plot Mapping Rate</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_mean_sd.html">plot_mean_sd</a></code> </p>
          </td>
          <td><p>Plot Row Standard Deviations vs. Row Means</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_mirna_counts.html">plot_mirna_counts</a></code> </p>
          </td>
          <td><p>Plot of Total miRNA Counts</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_pattern.html">plot_pattern</a></code> </p>
          </td>
          <td><p>Make Groups of Genes Using Expression Profile</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_pca_covariates.html">plot_pca_covariates</a></code> </p>
          </td>
          <td><p>Find Correlation Between Principal Components (PCs) and Covariates</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_pca.html">plot_pca</a></code> </p>
          </td>
          <td><p>Sample PCA Plot for Transformed Data</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_rrna_mapping_rate.html">plot_rrna_mapping_rate</a></code> <code><a href="plot_rrna_mapping_rate.html">.plot_rrna_mapping_rate</a></code> </p>
          </td>
          <td><p>Plot rRNA Mapping Rate</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_size_distribution.html">plot_size_distribution</a></code> </p>
          </td>
          <td><p>Plot Size Distribution of Small RNA-Seq Data</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_srna_clusters.html">plot_srna_clusters</a></code> </p>
          </td>
          <td><p>Cluster Small RNA Samples</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_total_reads.html">plot_total_reads</a></code> <code><a href="plot_total_reads.html">.plot_total_reads</a></code> </p>
          </td>
          <td><p>Plot Total Reads</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_volcano.html">plot_volcano</a></code> <code><a href="plot_volcano.html">.plot_volcano</a></code> </p>
          </td>
          <td><p>Plot Volcano</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="results_tables.html">results_tables</a></code> </p>
          </td>
          <td><p>Differential expression results tables</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="tmm.html">tmm</a></code> </p>
          </td>
          <td><p>Trimmed Mean of M-Values (TMM) Normalization</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="top_tables.html">top_tables</a></code> </p>
          </td>
          <td><p>Top tables of differential expression results</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="tpm.html">tpm</a></code> </p>
          </td>
          <td><p>Transcripts Per Million (TPM)</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="txi.html">txi</a></code> </p>
          </td>
          <td><p><code>tximport</code> Accessor Shortcut</p></td>
        </tr>
      </tbody>
      </table>
    </div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-all-functions">All functions</a></li>
    </ul>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Michael J. Steinbaugh, Lorena Pantano, Rory Kirchner, Victor Barrera, John Hutchinson, Harvard Chan Bioinformatics Core.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
