<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Plot sexually dimorphic gender marker genes — plotGenderMarkers • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Plot sexually dimorphic gender marker genes — plotGenderMarkers" />

<meta property="og:description" content="This is a convenience function that quickly plots known sexually dimorphic
genes, and overall gene experssion for the X and Y chromosomes." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioRNASeq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.25</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bcbioRNASeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot sexually dimorphic gender marker genes</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioRNASeq/blob/master/R/plotGenderMarkers-methods.R'><code>R/plotGenderMarkers-methods.R</code></a></small>
    <div class="hidden name"><code>plotGenderMarkers.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This is a convenience function that quickly plots known sexually dimorphic
genes, and overall gene experssion for the X and Y chromosomes.</p>
    
    </div>

    <pre class="usage"><span class='fu'>plotGenderMarkers</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S4 method for bcbioRNASeq</span>
<span class='fu'>plotGenderMarkers</span>(<span class='no'>object</span>, <span class='kw'>normalized</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"vst"</span>,
  <span class='st'>"rlog"</span>, <span class='st'>"tmm"</span>, <span class='st'>"rle"</span>, <span class='st'>"tpm"</span>, <span class='st'>"fpkm"</span>), <span class='kw'>interestingGroups</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>trans</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"identity"</span>, <span class='st'>"log2"</span>, <span class='st'>"log10"</span>), <span class='kw'>medianLine</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>color</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/options.html'>getOption</a></span>(<span class='kw'>x</span> <span class='kw'>=</span> <span class='st'>"acid.color.discrete"</span>, <span class='kw'>default</span> <span class='kw'>=</span>
  <span class='kw pkg'>acidplots</span><span class='kw ns'>::</span><span class='fu'><a href='https://rdrr.io/pkg/acidplots/man/synesthesia.html'>scale_colour_synesthesia_d</a></span>()), <span class='kw'>legend</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/options.html'>getOption</a></span>(<span class='kw'>x</span> <span class='kw'>=</span>
  <span class='st'>"acid.legend"</span>, <span class='kw'>default</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>), <span class='kw'>style</span> <span class='kw'>=</span> <span class='st'>"wide"</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>normalized</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code>.
Type of normalized counts to visualize:</p><ul>
<li><p><code>tpm</code> Transcripts per million.</p></li>
<li><p><code>tmm</code>: edgeR trimmed mean of M-values.</p></li>
<li><p><code>rlog</code>: DESeq2 log2 regularized log transformation.</p></li>
<li><p><code>vst</code>: DESeq2 log2 variance stabilizing transformation.</p></li>
</ul></td>
    </tr>
    <tr>
      <th>interestingGroups</th>
      <td><p><code>character</code>.
Groups of interest to use for visualization.
Corresponds to factors describing the columns of the object.</p></td>
    </tr>
    <tr>
      <th>trans</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code>.
Name of the axis scale transformation to apply.</p>
<p>For more information:</p><pre>help(topic = "scale_x_continuous", package = "ggplot2")
</pre></td>
    </tr>
    <tr>
      <th>medianLine</th>
      <td><p><code><a href='https://rdrr.io/r/base/logical.html'>logical(1)</a></code>.
Include median line for each group. Disabled if samples are colored by
sample name.</p></td>
    </tr>
    <tr>
      <th>color</th>
      <td><p><code>ScaleDiscrete</code>.
Desired ggplot2 color scale. Must supply discrete values. When set <code>NULL</code>,
the default ggplot2 color palette will be used. If manual color definitions
are desired, we recommend using <code><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>ggplot2::scale_color_manual()</a></code>.</p>
<p>To set the discrete color palette globally, use:</p><pre>options(acid.color.discrete = ggplot2::scale_color_viridis_d())
</pre></td>
    </tr>
    <tr>
      <th>legend</th>
      <td><p><code><a href='https://rdrr.io/r/base/logical.html'>logical(1)</a></code>.
Include plot legend.</p></td>
    </tr>
    <tr>
      <th>style</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code>.
Plot style.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>ggplot</code>.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>Updated 2019-08-07.</p>
    
    <h2 class="hasAnchor" id="methods-by-class-"><a class="anchor" href="#methods-by-class-"></a>Methods (by class)</h2>

    

<ul>
<li><p><code>SingleCellExperiment</code>: Applies <code>pseudobulk()</code> calculation to average
gene expression at sample level prior to plotting.</p></li>
</ul>

    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='plotCounts.html'>plotCounts()</a></code>.</p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>bcb</span>)
<span class='fu'>plotGenderMarkers</span>(<span class='no'>bcb</span>)</div><div class='output co'>#&gt; <span class='error'>Error in (function (object, genes, strict = TRUE) {    validObject(object)    assert(is(object, "Gene2Symbol"), isCharacter(genes), isFlag(strict))    if (any(genes %in% rownames(object))) {        table &lt;- rownames(object)    }    else if (any(genes %in% object[["geneName"]])) {        assert(matchesUniqueGeneNames(object, genes))        table &lt;- object[["geneName"]]    }    else if (any(genes %in% object[["geneID"]])) {        table &lt;- object[["geneID"]]    }    else {        stop(sprintf("All genes failed to map: %s.", toString(genes,             width = 100L)))    }    match &lt;- match(x = genes, table = table)    names(match) &lt;- genes    if (isTRUE(strict)) {        fun &lt;- stop    }    else {        fun &lt;- message    }    unmapped &lt;- which(is.na(match))    if (length(unmapped) &gt; 0L) {        fun(sprintf("Some genes failed to map: %s.", toString(genes[unmapped],             width = 200L)))    }    mapped &lt;- na.omit(match)    assert(hasLength(mapped))    mapped})(object = new("Gene2Symbol", rownames = c("ENSMUSG00000000001", "ENSMUSG00000000003", "ENSMUSG00000000028", "ENSMUSG00000000049", "ENSMUSG00000000058", "ENSMUSG00000000078", "ENSMUSG00000000085", "ENSMUSG00000000088", "ENSMUSG00000000093", "ENSMUSG00000000094", "ENSMUSG00000000120", "ENSMUSG00000000125", "ENSMUSG00000000126", "ENSMUSG00000000127", "ENSMUSG00000000149", "ENSMUSG00000000157", "ENSMUSG00000000159", "ENSMUSG00000000167", "ENSMUSG00000000171", "ENSMUSG00000000182", "ENSMUSG00000000183", "ENSMUSG00000000184", "ENSMUSG00000000197", "ENSMUSG00000000202", "ENSMUSG00000000204", "ENSMUSG00000000214", "ENSMUSG00000000215", "ENSMUSG00000000216", "ENSMUSG00000000223", "ENSMUSG00000000247", "ENSMUSG00000000248", "ENSMUSG00000000253", "ENSMUSG00000000275", "ENSMUSG00000000276", "ENSMUSG00000000278", "ENSMUSG00000000282", "ENSMUSG00000000290", "ENSMUSG00000000295", "ENSMUSG00000000296", "ENSMUSG00000000301", "ENSMUSG00000000303", "ENSMUSG00000000305", "ENSMUSG00000000308", "ENSMUSG00000000317", "ENSMUSG00000000318", "ENSMUSG00000000320", "ENSMUSG00000000326", "ENSMUSG00000000339", "ENSMUSG00000000340", "ENSMUSG00000000346", "ENSMUSG00000000355", "ENSMUSG00000000359", "ENSMUSG00000000374", "ENSMUSG00000000378", "ENSMUSG00000000384", "ENSMUSG00000000385", "ENSMUSG00000000392", "ENSMUSG00000000402", "ENSMUSG00000000411", "ENSMUSG00000000420", "ENSMUSG00000000435", "ENSMUSG00000000439", "ENSMUSG00000000440", "ENSMUSG00000000441", "ENSMUSG00000000489", "ENSMUSG00000000530", "ENSMUSG00000000531", "ENSMUSG00000000532", "ENSMUSG00000000560", "ENSMUSG00000000562", "ENSMUSG00000000567", "ENSMUSG00000000581", "ENSMUSG00000000594", "ENSMUSG00000000605", "ENSMUSG00000000617", "ENSMUSG00000000627", "ENSMUSG00000000628", "ENSMUSG00000000631", "ENSMUSG00000000632", "ENSMUSG00000000673", "ENSMUSG00000020152", "ENSMUSG00000020333", "ENSMUSG00000020676", "ENSMUSG00000020875", "ENSMUSG00000026315", "ENSMUSG00000026535", "ENSMUSG00000028494", "ENSMUSG00000029231", "ENSMUSG00000029534", "ENSMUSG00000029730", "ENSMUSG00000032050", "ENSMUSG00000035352", "ENSMUSG00000035385", "ENSMUSG00000038257", "ENSMUSG00000044807", "ENSMUSG00000048583", "ENSMUSG00000055022", "ENSMUSG00000061689", "ENSMUSG00000062661", "ENSMUSG00000074340"), nrows = 100L, listData = list(geneID = c("ENSMUSG00000000001", "ENSMUSG00000000003", "ENSMUSG00000000028", "ENSMUSG00000000049", "ENSMUSG00000000058", "ENSMUSG00000000078", "ENSMUSG00000000085", "ENSMUSG00000000088", "ENSMUSG00000000093", "ENSMUSG00000000094", "ENSMUSG00000000120", "ENSMUSG00000000125", "ENSMUSG00000000126", "ENSMUSG00000000127", "ENSMUSG00000000149", "ENSMUSG00000000157", "ENSMUSG00000000159", "ENSMUSG00000000167", "ENSMUSG00000000171", "ENSMUSG00000000182", "ENSMUSG00000000183", "ENSMUSG00000000184", "ENSMUSG00000000197", "ENSMUSG00000000202", "ENSMUSG00000000204", "ENSMUSG00000000214", "ENSMUSG00000000215", "ENSMUSG00000000216", "ENSMUSG00000000223", "ENSMUSG00000000247", "ENSMUSG00000000248", "ENSMUSG00000000253", "ENSMUSG00000000275", "ENSMUSG00000000276", "ENSMUSG00000000278", "ENSMUSG00000000282", "ENSMUSG00000000290", "ENSMUSG00000000295", "ENSMUSG00000000296", "ENSMUSG00000000301", "ENSMUSG00000000303", "ENSMUSG00000000305", "ENSMUSG00000000308", "ENSMUSG00000000317", "ENSMUSG00000000318", "ENSMUSG00000000320", "ENSMUSG00000000326", "ENSMUSG00000000339", "ENSMUSG00000000340", "ENSMUSG00000000346", "ENSMUSG00000000355", "ENSMUSG00000000359", "ENSMUSG00000000374", "ENSMUSG00000000378", "ENSMUSG00000000384", "ENSMUSG00000000385", "ENSMUSG00000000392", "ENSMUSG00000000402", "ENSMUSG00000000411", "ENSMUSG00000000420", "ENSMUSG00000000435", "ENSMUSG00000000439", "ENSMUSG00000000440", "ENSMUSG00000000441", "ENSMUSG00000000489", "ENSMUSG00000000530", "ENSMUSG00000000531", "ENSMUSG00000000532", "ENSMUSG00000000560", "ENSMUSG00000000562", "ENSMUSG00000000567", "ENSMUSG00000000581", "ENSMUSG00000000594", "ENSMUSG00000000605", "ENSMUSG00000000617", "ENSMUSG00000000627", "ENSMUSG00000000628", "ENSMUSG00000000631", "ENSMUSG00000000632", "ENSMUSG00000000673", "ENSMUSG00000020152", "ENSMUSG00000020333", "ENSMUSG00000020676", "ENSMUSG00000020875", "ENSMUSG00000026315", "ENSMUSG00000026535", "ENSMUSG00000028494", "ENSMUSG00000029231", "ENSMUSG00000029534", "ENSMUSG00000029730", "ENSMUSG00000032050", "ENSMUSG00000035352", "ENSMUSG00000035385", "ENSMUSG00000038257", "ENSMUSG00000044807", "ENSMUSG00000048583", "ENSMUSG00000055022", "ENSMUSG00000061689", "ENSMUSG00000062661", "ENSMUSG00000074340"), geneName = c("Gnai3", "Pbsn", "Cdc45", "Apoh", "Cav2", "Klf6", "Scmh1", "Cox5a", "Tbx2", "Tbx4", "Ngfr", "Wnt3", "Wnt9a", "Fer", "Gna12", "Itgb2l", "Igsf5", "Pih1d2", "Sdhd", "Fgf23", "Fgf6", "Ccnd2", "Nalcn", "Btbd17", "Slfn4", "Th", "Ins2", "Scnn1g", "Drp2", "Lhx2", "Clec2g", "Gmpr", "Trim25", "Dgke", "Scpep1", "Mnt", "Itgb2", "Hddc2", "Tpd52l1", "Pemt", "Cdh1", "Cdh4", "Ckmt1", "Bcl6b", "Clec10a", "Alox12", "Comt", "Rtca", "Dbt", "Dazap2", "Mcts1", "Rem1", "Trappc10", "Ccm2", "Tbrg4", "Tmprss2", "Fap", "Egfl6", "Tssk3", "Galnt1", "Myf5", "Mkrn2", "Pparg", "Raf1", "Pdgfb", "Acvrl1", "Grasp", "Acvr1b", "Gabra2", "Adora3", "Sox9", "C1d", "Gm2a", "Clcn4", "Grm6", "Sema4f", "Hk2", "Myo18a", "Sez6", "Haao", "Actr2", "Acsl6", "Ccl11", "Hoxb9", "Serpinb8", "Ifi202b", "Plin2", "Pdgfra", "St7", "Mcm7", "Rdx", "Ccl12", "Ccl2", "Glra3", "Zfp354c", "Igf2", "Cntn1", "Dlgap4", "Ncs1", "Ovgp1")), elementType = "ANY",     elementMetadata = NULL, metadata = list(package = "freerange",         version = structure(list(c(0L, 2L, 0L)), class = c("package_version",         "numeric_version")), date = structure(18099, class = "Date"),         organism = "Mus musculus", genomeBuild = "GRCm38", ensemblRelease = 90L,         ensembldb = structure(list(name = c("Db type", "Type of Gene ID",         "Supporting package", "Db created by", "script_version",         "Creation time", "ensembl_version", "ensembl_host", "Organism",         "taxonomy_id", "genome_build", "DBSCHEMAVERSION"), value = c("EnsDb",         "Ensembl Gene ID", "ensembldb", "ensembldb package from Bioconductor",         "0.3.4", "Tue Jan  1 21:15:02 2019", "90", "localhost",         "Mus musculus", "10090", "GRCm38", "2.1")), class = "data.frame", row.names = c(NA,         -12L)), level = "genes", id = "AH57770", format = "makeUnique")),     genes = c("ENSMUSG00000086503", "ENSMUSG00000069045", "ENSMUSG00000056673",     "ENSMUSG00000068457"), strict = FALSE): All genes failed to map: ENSMUSG00000086503, ENSMUSG00000069045, ENSMUSG00000056673, ENSMUSG00000068457.</span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#methods-by-class-">Methods (by class)</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Michael Steinbaugh</p>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io/'>Lorena Pantano</a>, <a href='https://github.com/roryk/'>Rory Kirchner</a>, <a href='https://github.com/vbarrera/'>Victor Barrera</a>, <a href='https://github.com/chapmanb/'>Brad Chapman</a>, <a href='https://github.com/marypiper/'>Mary Piper</a>, <a href='https://github.com/mistrm82/'>Meeta Mistry</a>, <a href='https://github.com/rkhetani/'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford/'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann/'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson/'>John Hutchinson</a>, <a href='https://github.com/sjhosui/'>Shannan Ho Sui</a>, <a href='https://bioinformatics.sph.harvard.edu/'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9100.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


