<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Gene Heatmap — plot_gene_heatmap • bcbioRnaseq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bcbioRnaseq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRnaseq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Gene Heatmap</h1>
    </div>

    
    <p>These functions facilitate heatmap plotting of a specified set of genes. By
default, row- and column-wise hierarchical clustering is performed using the
Ward method, but this behavior can be overrided by setting <code>cluster_rows</code> or
<code>cluster_cols</code> to <code>FALSE</code>. When column clustering is disabled, the columns
are sorted by the interesting groups (<code>interesting_groups</code>) specified in the
<a href='bcbioRNADataSet-class.html'>bcbioRNADataSet</a> and then the sample names.</p>
    

    <pre class="usage"><span class='fu'>plot_gene_heatmap</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S4 method for bcbioRNADataSet</span>
<span class='fu'>plot_gene_heatmap</span>(<span class='no'>object</span>, <span class='no'>genes</span>,
  <span class='kw'>clustering_method</span> <span class='kw'>=</span> <span class='st'>"ward.D2"</span>, <span class='kw'>cluster_rows</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>cluster_cols</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>scale</span> <span class='kw'>=</span> <span class='st'>"row"</span>, <span class='kw'>interesting_groups</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>annotation</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>title</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    <tr>
      <th>genes</th>
      <td><p>Character vector of specific gene identifiers to plot.</p></td>
    </tr>
    <tr>
      <th>clustering_method</th>
      <td><p>Hierarchical clustering method. Accepts the same
values as <code><a href='http://www.rdocumentation.org/packages/stats/topics/hclust'>stats::hclust()</a></code>.</p></td>
    </tr>
    <tr>
      <th>cluster_rows</th>
      <td><p>Use hierarchical clustering to arrange rows.</p></td>
    </tr>
    <tr>
      <th>cluster_cols</th>
      <td><p>Use hierarchical clustering to arrange columns.</p></td>
    </tr>
    <tr>
      <th>scale</th>
      <td><p>Character indicating if the values should be centered and scaled
in either the <code>row</code> direction, <code>column</code> direction, or <code>none</code>.</p></td>
    </tr>
    <tr>
      <th>interesting_groups</th>
      <td><p><em>Optional</em>. Interesting groups to label with bars
above heatmap. If <code>NULL</code>, defaults to <code>interesting_groups</code> defined in the
<a href='bcbioRNADataSet-class.html'>bcbioRNADataSet</a>.</p></td>
    </tr>
    <tr>
      <th>annotation</th>
      <td><p><em>Optional</em>. Alternative annotation to use. Useful when
plotting more than one column.</p></td>
    </tr>
    <tr>
      <th>title</th>
      <td><p><em>Optional</em>. Text to include in plot title.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Graphical output only.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p><code><a href='http://www.rdocumentation.org/packages/pheatmap/topics/pheatmap'>pheatmap::pheatmap()</a></code>.</p>
<p>Other Heatmaps: <code><a href='plot_correlation_heatmap.html'>plot_correlation_heatmap</a></code>,
  <code><a href='plot_deg_heatmap.html'>plot_deg_heatmap</a></code></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>data</span>(<span class='no'>bcb</span>)
<span class='fu'>plot_gene_heatmap</span>(<span class='no'>bcb</span>)</div><div class='output co'>#&gt; <span class='error'>Error in (function (classes, fdef, mtable) {    methods &lt;- .findInheritedMethods(classes, fdef, mtable)    if (length(methods) == 1L)         return(methods[[1L]])    else if (length(methods) == 0L) {        cnames &lt;- paste0("\"", vapply(classes, as.character,             ""), "\"", collapse = ", ")        stop(gettextf("unable to find an inherited method for function %s for signature %s",             sQuote(fdef@generic), sQuote(cnames)), domain = NA)    }    else stop("Internal error in finding inherited methods; didn't return a unique method",         domain = NA)})(list("function", structure("missing", package = "methods")),     structure(function (x, i, ...)     standardGeneric("assay"), generic = structure("assay", package = "SummarizedExperiment"), package = "SummarizedExperiment", group = list(), valueClass = character(0), signature = c("x",     "i"), default = `\001NULL\001`, skeleton = (function (x,         i, ...)     stop("invalid call in method dispatch to 'assay' (no default method)",         domain = NA))(x, i, ...), class = structure("standardGeneric", package = "methods")),     &lt;environment&gt;): unable to find an inherited method for function ‘assay’ for signature ‘"function", "missing"’</span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Michael J. Steinbaugh, Lorena Pantano, Rory Kirchner, Victor Barrera, John Hutchinson, Harvard Chan Bioinformatics Core.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
