<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Differential Expression Results Tables — resultsTables • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  <link href="../extra.css" rel="stylesheet">
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bcbioRNASeq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Differential Expression Results Tables</h1>
    </div>

    
    <p>Differential Expression Results Tables</p>
    

    <pre class="usage"><span class='fu'>resultsTables</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S4 method for DESeqResults</span>
<span class='fu'>resultsTables</span>(<span class='no'>object</span>, <span class='kw'>lfc</span> <span class='kw'>=</span> <span class='fl'>0L</span>, <span class='kw'>annotable</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>summary</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>headerLevel</span> <span class='kw'>=</span> <span class='fl'>2L</span>, <span class='kw'>write</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>dir</span> <span class='kw'>=</span> <span class='fu'>getwd</span>(),
  <span class='kw'>dropboxDir</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>rdsToken</span> <span class='kw'>=</span> <span class='fl'>NA</span>, <span class='kw'>quiet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    <tr>
      <th>lfc</th>
      <td><p>Log fold change ratio (base 2) cutoff. Does not apply to
statistical hypothesis testing, only gene filtering in the results tables.
See <code>results()</code> for additional information about using <code>lfcThreshold</code> and
<code>altHypothesis</code> to set an alternative hypothesis based on expected fold
changes.</p></td>
    </tr>
    <tr>
      <th>annotable</th>
      <td><p>Join Ensembl gene annotations to the results. Apply gene
identifier to symbol mappings. A previously saved annotable data.frame
is recommended. Alternatively if set <code>NULL</code>, then gene annotations will
not be added to the results.</p></td>
    </tr>
    <tr>
      <th>summary</th>
      <td><p>Show summary statistics.</p></td>
    </tr>
    <tr>
      <th>headerLevel</th>
      <td><p>R Markdown header level. Applies only when
<code>summary = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>write</th>
      <td><p>Write CSV files to disk.</p></td>
    </tr>
    <tr>
      <th>dir</th>
      <td><p>Local directory path where to write the results tables.</p></td>
    </tr>
    <tr>
      <th>dropboxDir</th>
      <td><p>Dropbox directory path where to archive the results tables
for permanent storage (e.g. Stem Cell Commons). When this option is
enabled, unique links per file are generated internally with the rdrop2
package.</p></td>
    </tr>
    <tr>
      <th>rdsToken</th>
      <td><p>RDS file token to use for Dropbox authentication.</p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p>If <code>TRUE</code>, suppress any status messages and/or progress bars.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Results list.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>load</span>(<span class='fu'>system.file</span>(<span class='st'>"extdata/bcb.rda"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"bcbioRNASeq"</span>))
<span class='fu'>load</span>(<span class='fu'>system.file</span>(<span class='st'>"extdata/res.rda"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"bcbioRNASeq"</span>))

<span class='co'># Can use a stashed annotable for improved speed</span>
<span class='no'>annotable</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='annotable.html'>annotable</a></span>(<span class='no'>bcb</span>)

<span class='no'>resTbl</span> <span class='kw'>&lt;-</span> <span class='fu'>resultsTables</span>(
    <span class='no'>res</span>,
    <span class='kw'>lfc</span> <span class='kw'>=</span> <span class='fl'>0.25</span>,
    <span class='kw'>annotable</span> <span class='kw'>=</span> <span class='no'>annotable</span>,
    <span class='kw'>summary</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
    <span class='kw'>headerLevel</span> <span class='kw'>=</span> <span class='fl'>2L</span>,
    <span class='kw'>write</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
    <span class='kw'>dir</span> <span class='kw'>=</span> <span class='st'>"resultsTables"</span>,
    <span class='kw'>dropboxDir</span> <span class='kw'>=</span> <span class='fu'>file.path</span>(<span class='st'>"bcbioRNASeq_examples"</span>, <span class='st'>"resultsTables"</span>),
    <span class='kw'>rdsToken</span> <span class='kw'>=</span> <span class='fu'>system.file</span>(<span class='st'>"extdata/token.rds"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"bcbioRNASeq"</span>)
)</div><div class='output co'>#&gt; 
#&gt; 
#&gt; ## Summary statistics
#&gt; 
#&gt; 
#&gt; - 505 genes in counts matrix
#&gt; - Base mean &gt; 0: 304 genes (non-zero)
#&gt; - Base mean &gt; 1: 213 genes
#&gt; - Alpha cutoff: 0.1
#&gt; - LFC cutoff: 0.25 (applied in tables only)
#&gt; - DEG pass alpha: 4 genes
#&gt; - DEG LFC up: 2 genes
#&gt; - DEG LFC down: 2 genes
#&gt; </div><div class='output co'>#&gt; <span class='message'>Dropbox: Michael Steinbaugh &lt;mjsteinbaugh@protonmail.com&gt;</span></div><div class='output co'>#&gt; <span class='message'>/bcbioRNASeq_examples/resultsTables/group_ko_vs_ctrl_all.csv.gz was successfully deleted</span></div><div class='output co'>#&gt; <span class='message'>File /Users/mike/GitHub/hbc/bcbioRNASeq/docs/reference/resultsTables/group_ko_vs_ctrl_all.csv.gz uploaded as /bcbioRNASeq_examples/resultsTables/group_ko_vs_ctrl_all.csv.gz successfully at 2018-02-19T22:15:51Z</span></div><div class='output co'>#&gt; <span class='message'>/bcbioRNASeq_examples/resultsTables/group_ko_vs_ctrl_deg.csv.gz was successfully deleted</span></div><div class='output co'>#&gt; <span class='message'>File /Users/mike/GitHub/hbc/bcbioRNASeq/docs/reference/resultsTables/group_ko_vs_ctrl_deg.csv.gz uploaded as /bcbioRNASeq_examples/resultsTables/group_ko_vs_ctrl_deg.csv.gz successfully at 2018-02-19T22:15:54Z</span></div><div class='output co'>#&gt; <span class='message'>/bcbioRNASeq_examples/resultsTables/group_ko_vs_ctrl_deg_lfc_up.csv.gz was successfully deleted</span></div><div class='output co'>#&gt; <span class='message'>File /Users/mike/GitHub/hbc/bcbioRNASeq/docs/reference/resultsTables/group_ko_vs_ctrl_deg_lfc_up.csv.gz uploaded as /bcbioRNASeq_examples/resultsTables/group_ko_vs_ctrl_deg_lfc_up.csv.gz successfully at 2018-02-19T22:15:58Z</span></div><div class='output co'>#&gt; <span class='message'>/bcbioRNASeq_examples/resultsTables/group_ko_vs_ctrl_deg_lfc_down.csv.gz was successfully deleted</span></div><div class='output co'>#&gt; <span class='message'>File /Users/mike/GitHub/hbc/bcbioRNASeq/docs/reference/resultsTables/group_ko_vs_ctrl_deg_lfc_down.csv.gz uploaded as /bcbioRNASeq_examples/resultsTables/group_ko_vs_ctrl_deg_lfc_down.csv.gz successfully at 2018-02-19T22:16:03Z</span></div><div class='output co'>#&gt; 
#&gt; 
#&gt; ## Results tables
#&gt; 
#&gt; 
#&gt; - [`group_ko_vs_ctrl_all.csv.gz`](https://www.dropbox.com/s/upev0uc9e9qj0ny/group_ko_vs_ctrl_all.csv.gz?dl=0): All genes, sorted by Ensembl identifier.
#&gt; - [`group_ko_vs_ctrl_deg.csv.gz`](https://www.dropbox.com/s/4416e9a7422g7yx/group_ko_vs_ctrl_deg.csv.gz?dl=0): Genes that pass the alpha (FDR) cutoff.
#&gt; - [`group_ko_vs_ctrl_deg_lfc_up.csv.gz`](https://www.dropbox.com/s/mdr0h0ofbc68nbo/group_ko_vs_ctrl_deg_lfc_up.csv.gz?dl=0): Upregulated DEG; positive log2 fold change.
#&gt; - [`group_ko_vs_ctrl_deg_lfc_down.csv.gz`](https://www.dropbox.com/s/dj2inosd3y1aln4/group_ko_vs_ctrl_deg_lfc_down.csv.gz?dl=0): Downregulated DEG; negative log2 fold change.
#&gt; </div><div class='input'><span class='fu'>names</span>(<span class='no'>resTbl</span>)</div><div class='output co'>#&gt;  [1] "contrast"     "alpha"        "lfc"          "all"          "deg"         
#&gt;  [6] "degLFC"       "degLFCUp"     "degLFCDown"   "localFiles"   "dropboxFiles"</div><div class='input'><span class='fu'>dir</span>(<span class='st'>"resultsTables"</span>)</div><div class='output co'>#&gt; [1] "group_ko_vs_ctrl_all.csv.gz"         
#&gt; [2] "group_ko_vs_ctrl_deg_lfc_down.csv.gz"
#&gt; [3] "group_ko_vs_ctrl_deg_lfc_up.csv.gz"  
#&gt; [4] "group_ko_vs_ctrl_deg.csv.gz"         </div><div class='input'><span class='fu'>unlink</span>(<span class='st'>"resultsTables"</span>, <span class='kw'>recursive</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io'>Lorena Pantano</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/mistrm82'>Meeta Mistry</a>, <a href='https://github.com/rkhetani'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson'>John Hutchinson</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
