<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Differential expression results tables — resultsTables • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Differential expression results tables — resultsTables" />

<meta property="og:description" content="Differential expression results tables" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioRNASeq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/bcbioRNASeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Differential expression results tables</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioRNASeq/blob/master/R/resultsTables-methods.R'><code>R/resultsTables-methods.R</code></a></small>
    <div class="hidden name"><code>resultsTables.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Differential expression results tables</p>
    
    </div>

    <pre class="usage"><span class='fu'>resultsTables</span>(<span class='no'>results</span>, <span class='no'>counts</span>, <span class='no'>...</span>)

<span class='co'># S4 method for DESeqResults,DESeqDataSet</span>
<span class='fu'>resultsTables</span>(<span class='no'>results</span>, <span class='no'>counts</span>, <span class='no'>alpha</span>,
  <span class='kw'>lfcThreshold</span> <span class='kw'>=</span> <span class='fl'>0L</span>, <span class='kw'>summary</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>write</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>headerLevel</span> <span class='kw'>=</span> <span class='fl'>2L</span>, <span class='kw'>dir</span> <span class='kw'>=</span> <span class='st'>"."</span>, <span class='kw'>dropboxDir</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>rdsToken</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)

<span class='co'># S4 method for DESeqResults,missingOrNULL</span>
<span class='fu'>resultsTables</span>(<span class='no'>results</span>,
  <span class='kw'>counts</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>alpha</span>, <span class='kw'>lfcThreshold</span> <span class='kw'>=</span> <span class='fl'>0L</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>results</th>
      <td><p><code>DESeqResults</code>.</p></td>
    </tr>
    <tr>
      <th>counts</th>
      <td><p><code>DESeqDataSet</code>. Corresponding object containing the counts that
were used to generate the <code>DESeqResults</code> object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    <tr>
      <th>alpha</th>
      <td><p><code>scalar numeric</code>. Adjusted P value ("alpha") cutoff.</p></td>
    </tr>
    <tr>
      <th>lfcThreshold</th>
      <td><p><code>scalar numeric</code>. Log fold change ratio (base 2) cutoff
threshold.</p></td>
    </tr>
    <tr>
      <th>summary</th>
      <td><p><code>boolean</code>. Show summary statistics.</p></td>
    </tr>
    <tr>
      <th>write</th>
      <td><p><code>boolean</code>. Write CSV files to disk.</p></td>
    </tr>
    <tr>
      <th>headerLevel</th>
      <td><p><code>scalar integer</code> (<code>1</code>-<code>7</code>). R Markdown header level.</p></td>
    </tr>
    <tr>
      <th>dir</th>
      <td><p><code>string</code>. Local directory path.</p></td>
    </tr>
    <tr>
      <th>dropboxDir</th>
      <td><p><code>string</code> or <code>NULL</code>. Dropbox directory path where to archive
the results tables for permanent storage (e.g. Stem Cell Commons). When
this option is enabled, unique links per file are generated internally with
the rdrop2 package. Note that local files are written to <code><a href='https://rdrr.io/r/base/tempfile.html'>base::tempdir()</a></code>
and the <code>dir</code> argument is ignored, if this is enabled.</p></td>
    </tr>
    <tr>
      <th>rdsToken</th>
      <td><p><code>string</code> or <code>NULL</code>. RDS file token to use for Dropbox
authentication. If set <code>NULL</code> and <code>dropboxDir</code> is defined, then an
interactive prompt will appear requesting authorization.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>list</code> containing modified <code>DESeqResults</code> return, including
additional gene-level metadata and normalized counts.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>Log fold change cutoff threshold ("<code>lfcThreshold</code>") does not apply to
statistical hypothesis testing, only gene filtering in the results tables.
See <code><a href='https://rdrr.io/pkg/DESeq2/man/results.html'>DESeq2::results()</a></code> for additional information about using
<code>lfcThreshold</code> and <code>altHypothesis</code> to set an alternative hypothesis based
on expected fold changes.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other Differential Expression Functions: <code><a href='alphaSummary.html'>alphaSummary</a></code>,
  <code><a href='contrastName.html'>contrastName</a></code>, <code><a href='plotDEGHeatmap.html'>plotDEGHeatmap</a></code>,
  <code><a href='plotDEGPCA.html'>plotDEGPCA</a></code>, <code><a href='plotMA.html'>plotMA</a></code>,
  <code><a href='plotVolcano.html'>plotVolcano</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># DESeqResults, DESeqDataSet ====</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>resultsTables</span>(<span class='kw'>results</span> <span class='kw'>=</span> <span class='no'>res_small</span>, <span class='kw'>counts</span> <span class='kw'>=</span> <span class='no'>dds_small</span>)</div><div class='output co'>#&gt; 
#&gt; 
#&gt; ## treatment folic acid vs control
#&gt; 
#&gt; 
#&gt; 
#&gt; ### Summary statistics
#&gt; 
#&gt; 
#&gt; - 500 genes in counts matrix
#&gt; - Base mean &gt; 0: 500 genes (non-zero)
#&gt; - Base mean &gt; 1: 500 genes
#&gt; - Alpha: 0.1
#&gt; - LFC threshold: 0
#&gt; - DEG pass alpha: 254 genes
#&gt; - DEG LFC up: 115 genes
#&gt; - DEG LFC down: 139 genes
#&gt; </div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "deg"          "degLFC"       "degLFCUp"     "degLFCDown"   "all"         
#&gt; [6] "contrast"     "alpha"        "lfcThreshold"</div><div class='input'><span class='no'>x</span><span class='kw'>[[</span><span class='st'>"deg"</span>]] <span class='kw'>%&gt;%</span> <span class='fu'><a href='https://rdrr.io/r/utils/str.html'>str</a></span>()</div><div class='output co'>#&gt; 'data.frame':	254 obs. of  44 variables:
#&gt;  $ baseMean                                     : num  12562 69340 3560 3544 3002 ...
#&gt;  $ log2FoldChange                               : num  8.27 5.36 4.08 3.56 3.74 ...
#&gt;  $ lfcSE                                        : num  0.274 0.286 0.283 0.255 0.284 ...
#&gt;  $ stat                                         : num  29.9 18.7 14.4 14 13.2 ...
#&gt;  $ pvalue                                       : num  7.87e-197 3.65e-78 4.35e-47 3.13e-44 1.65e-39 ...
#&gt;  $ padj                                         : num  3.89e-194 9.03e-76 7.16e-45 3.87e-42 1.63e-37 ...
#&gt;  $ geneID                                       : chr  "ENSMUSG00000024164" "ENSMUSG00000029304" "ENSMUSG00000003617" "ENSMUSG00000022037" ...
#&gt;  $ seqnames                                     : Factor w/ 86 levels "3","X","16","7",..: 11 12 1 14 5 13 15 6 11 18 ...
#&gt;  $ start                                        : int  57203970 104435118 19957054 65968483 46735080 32384633 117277331 136872435 34263209 60803848 ...
#&gt;  $ end                                          : int  57228136 104441050 20009145 65981547 46779578 32388252 117282321 136875823 34269418 60812646 ...
#&gt;  $ width                                        : int  24167 5933 52092 13065 44499 3620 4991 3389 6210 8799 ...
#&gt;  $ strand                                       : Factor w/ 3 levels "+","-","*": 2 1 1 1 1 2 2 2 1 1 ...
#&gt;  $ broadClass                                   : Factor w/ 8 levels "coding","ig",..: 1 1 1 1 1 1 1 1 1 1 ...
#&gt;  $ description                                  : Factor w/ 48505 levels " 1 isoform 3  [Source:RefSeq peptide;Acc:NP_001177633]",..: 3427 42477 2783 3125 7251 9435 9628 9919 7362 2461 ...
#&gt;  $ geneBiotype                                  : Factor w/ 44 levels "3prime_overlapping_ncRNA",..: 22 22 22 22 22 22 22 22 22 22 ...
#&gt;  $ geneName                                     : Factor w/ 52861 levels "0610005C13Rik",..: 8340 48497 9898 9647 35979 38133 38589 39104 35851 9012 ...
#&gt;  $ seqCoordSystem                               : Factor w/ 2 levels "chromosome","scaffold": 1 1 1 1 1 1 1 1 1 1 ...
#&gt;  $ baseVar                                      : num  1.88e+08 6.00e+09 1.37e+07 1.17e+07 8.87e+06 ...
#&gt;  $ allZero                                      : logi  FALSE FALSE FALSE FALSE FALSE FALSE ...
#&gt;  $ dispGeneEst                                  : num  0.0124 0.0533 0.0398 0.0201 0.0416 ...
#&gt;  $ dispGeneIter                                 : num  14 3 7 8 7 8 3 8 8 4 ...
#&gt;  $ dispFit                                      : num  0.0856 0.0839 0.0909 0.0909 0.0922 ...
#&gt;  $ dispersion                                   : num  0.0527 0.0658 0.0628 0.0502 0.0636 ...
#&gt;  $ dispIter                                     : int  10 9 7 11 7 10 9 12 8 5 ...
#&gt;  $ dispOutlier                                  : logi  FALSE FALSE FALSE FALSE FALSE FALSE ...
#&gt;  $ dispMAP                                      : num  0.0527 0.0658 0.0628 0.0502 0.0636 ...
#&gt;  $ Intercept                                    : num  5.9 11.39 8.42 8.96 8.51 ...
#&gt;  $ treatment_folic_acid_vs_control              : num  8.72 5.66 4.31 3.72 3.95 ...
#&gt;  $ SE_Intercept                                 : num  0.22 0.214 0.214 0.19 0.215 ...
#&gt;  $ SE_treatment_folic_acid_vs_control           : num  0.291 0.303 0.299 0.267 0.301 ...
#&gt;  $ WaldStatistic_Intercept                      : num  26.9 53.1 39.4 47.1 39.6 ...
#&gt;  $ WaldStatistic_treatment_folic_acid_vs_control: num  29.9 18.7 14.4 14 13.2 ...
#&gt;  $ WaldPvalue_Intercept                         : num  6.94e-159 0.00 0.00 0.00 0.00 ...
#&gt;  $ WaldPvalue_treatment_folic_acid_vs_control   : num  7.87e-197 3.65e-78 4.35e-47 3.13e-44 1.65e-39 ...
#&gt;  $ betaConv                                     : logi  TRUE TRUE TRUE TRUE TRUE TRUE ...
#&gt;  $ betaIter                                     : num  2 3 3 2 3 4 3 2 3 3 ...
#&gt;  $ deviance                                     : num  81.6 116 85 84.9 84.3 ...
#&gt;  $ maxCooks                                     : num  0.541 0.363 0.246 0.186 0.313 ...
#&gt;  $ control_rep1                                 : num  78.1 2313.6 301.6 512.8 334.5 ...
#&gt;  $ control_rep2                                 : num  46.8 2998.4 344.7 536.8 452.1 ...
#&gt;  $ control_rep3                                 : num  56.2 2733.3 379.9 445.7 301.7 ...
#&gt;  $ fa_day7_rep1                                 : num  25465 182255 8797 7941 4727 ...
#&gt;  $ fa_day7_rep2                                 : num  25910 121258 6227 5552 5293 ...
#&gt;  $ fa_day7_rep3                                 : num  23817 104478 5311 6276 6902 ...</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Michael Steinbaugh</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://lpantano.github.io'>Lorena Pantano</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/chapmanb'>Brad Chapman</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/mistrm82'>Meeta Mistry</a>, <a href='https://github.com/rkhetani'>Radhika Khetani</a>, <a href='https://github.com/KayleighRutherford'>Kayleigh Rutherford</a>, <a href='https://github.com/ohofmann'>Oliver Hofmann</a>, <a href='https://github.com/jnhutchinson'>John Hutchinson</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9000.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

