<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Bracket-Based Subsetting — subset • bcbioRNASeq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bcbioRNASeq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioRNASeq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Bracket-Based Subsetting</h1>
    </div>

    
    <p>Extract genes by row and samples by column from a <a href='bcbioRNADataSet-class.html'>bcbioRNADataSet</a>. The
internal DESeqDataSet and count transformations are rescaled automatically.</p>
    

    <pre class="usage"># S4 method for bcbioRNADataSet,ANY,ANY,ANY
[(x, i, j, ..., drop = FALSE)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>object from which to extract element(s) or in which to replace element(s).</p></td>
    </tr>
    <tr>
      <th>i</th>
      <td><p>indices specifying elements to extract or replace.  Indices are
    <code>numeric</code> or <code>character</code> vectors or empty (missing) or
    <code>NULL</code>.  Numeric values are coerced to integer as by
    <code>as.integer</code> (and hence truncated towards zero).
    Character vectors will be matched to the <code>names</code> of the
    object (or for matrices/arrays, the <code>dimnames</code>):
    see &#8216;Character indices&#8217; below for further details.</p>
<p>For <code>[</code>-indexing only: <code>i</code>, <code>j</code>, <code>&#8230;</code> can be
    logical vectors, indicating elements/slices to select.  Such vectors
    are recycled if necessary to match the corresponding extent.
    <code>i</code>, <code>j</code>, <code>&#8230;</code> can also be negative integers,
    indicating elements/slices to leave out of the selection.</p>
<p>When indexing arrays by <code>[</code> a single argument <code>i</code> can be a
    matrix with as many columns as there are dimensions of <code>x</code>; the
    result is then a vector with elements corresponding to the sets of
    indices in each row of <code>i</code>.</p>
<p>An index value of <code>NULL</code> is treated as if it were <code>integer(0)</code>.</p></td>
    </tr>
    <tr>
      <th>j</th>
      <td><p>indices specifying elements to extract or replace.  Indices are
    <code>numeric</code> or <code>character</code> vectors or empty (missing) or
    <code>NULL</code>.  Numeric values are coerced to integer as by
    <code>as.integer</code> (and hence truncated towards zero).
    Character vectors will be matched to the <code>names</code> of the
    object (or for matrices/arrays, the <code>dimnames</code>):
    see &#8216;Character indices&#8217; below for further details.</p>
<p>For <code>[</code>-indexing only: <code>i</code>, <code>j</code>, <code>&#8230;</code> can be
    logical vectors, indicating elements/slices to select.  Such vectors
    are recycled if necessary to match the corresponding extent.
    <code>i</code>, <code>j</code>, <code>&#8230;</code> can also be negative integers,
    indicating elements/slices to leave out of the selection.</p>
<p>When indexing arrays by <code>[</code> a single argument <code>i</code> can be a
    matrix with as many columns as there are dimensions of <code>x</code>; the
    result is then a vector with elements corresponding to the sets of
    indices in each row of <code>i</code>.</p>
<p>An index value of <code>NULL</code> is treated as if it were <code>integer(0)</code>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    <tr>
      <th>drop</th>
      <td><p>For matrices and arrays.  If <code>TRUE</code> the result is
    coerced to the lowest possible dimension (see the examples).  This
    only works for extracting elements, not for the replacement.  See
    <code>drop</code> for further details.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><a href='bcbioRNADataSet-class.html'>bcbioRNADataSet</a>.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p><code>help("[", "base")</code>.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>data</span>(<span class='no'>bcb</span>)
<span class='no'>genes</span> <span class='kw'>&lt;-</span> <span class='fl'>1L</span>:<span class='fl'>50L</span>
<span class='no'>samples</span> <span class='kw'>&lt;-</span> <span class='fu'>c</span>(<span class='st'>"group1_1"</span>, <span class='st'>"group1_2"</span>)

<span class='co'># Subset by sample name</span>
<span class='no'>bcb</span>[, <span class='no'>samples</span>]</div><div class='output co'>#&gt; <span class='message'>using just counts from tximport</span></div><div class='output co'>#&gt; <span class='message'>estimating size factors</span></div><div class='output co'>#&gt; <span class='message'>estimating dispersions</span></div><div class='output co'>#&gt; <span class='message'>gene-wise dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>mean-dispersion relationship</span></div><div class='output co'>#&gt; <span class='message'>final dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>fitting model and testing</span></div><div class='output co'>#&gt; <span class='message'>Performing rlog transformation</span></div><div class='output co'>#&gt; <span class='message'>Performing variance stabilizing transformation</span></div><div class='output co'>#&gt; class: bcbioRNADataSet 
#&gt; dim: 505 2 
#&gt; metadata(25): version uploadDir ... sessionInfo missingGenes
#&gt; assays(6): raw normalized ... rlog vst
#&gt; rownames(505): ENSMUSG00000002459 ENSMUSG00000004768 ...
#&gt;   ENSMUSG00000105982 ENSMUSG00000109048
#&gt; rowData names(0):
#&gt; colnames(2): group1_1 group1_2
#&gt; colData names(3): sampleID sampleName group</div><div class='input'>
<span class='co'># Subset by gene list</span>
<span class='no'>bcb</span>[<span class='no'>genes</span>, ]</div><div class='output co'>#&gt; <span class='message'>using just counts from tximport</span></div><div class='output co'>#&gt; <span class='message'>estimating size factors</span></div><div class='output co'>#&gt; <span class='message'>estimating dispersions</span></div><div class='output co'>#&gt; <span class='message'>gene-wise dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>mean-dispersion relationship</span></div><div class='output co'>#&gt; <span class='message'>final dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>fitting model and testing</span></div><div class='output co'>#&gt; <span class='message'>Performing rlog transformation</span></div><div class='output co'>#&gt; <span class='message'>Performing variance stabilizing transformation</span></div><div class='output co'>#&gt; class: bcbioRNADataSet 
#&gt; dim: 50 4 
#&gt; metadata(25): version uploadDir ... sessionInfo missingGenes
#&gt; assays(6): raw normalized ... rlog vst
#&gt; rownames(50): ENSMUSG00000002459 ENSMUSG00000004768 ...
#&gt;   ENSMUSG00000026154 ENSMUSG00000026155
#&gt; rowData names(0):
#&gt; colnames(4): group1_1 group1_2 group2_1 group2_2
#&gt; colData names(3): sampleID sampleName group</div><div class='input'>
<span class='co'># Subset by both genes and samples</span>
<span class='no'>bcb</span>[<span class='no'>genes</span>, <span class='no'>samples</span>]</div><div class='output co'>#&gt; <span class='message'>using just counts from tximport</span></div><div class='output co'>#&gt; <span class='message'>estimating size factors</span></div><div class='output co'>#&gt; <span class='message'>estimating dispersions</span></div><div class='output co'>#&gt; <span class='message'>gene-wise dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>mean-dispersion relationship</span></div><div class='output co'>#&gt; <span class='message'>final dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>fitting model and testing</span></div><div class='output co'>#&gt; <span class='message'>Performing rlog transformation</span></div><div class='output co'>#&gt; <span class='message'>Performing variance stabilizing transformation</span></div><div class='output co'>#&gt; class: bcbioRNADataSet 
#&gt; dim: 50 2 
#&gt; metadata(25): version uploadDir ... sessionInfo missingGenes
#&gt; assays(6): raw normalized ... rlog vst
#&gt; rownames(50): ENSMUSG00000002459 ENSMUSG00000004768 ...
#&gt;   ENSMUSG00000026154 ENSMUSG00000026155
#&gt; rowData names(0):
#&gt; colnames(2): group1_1 group1_2
#&gt; colData names(3): sampleID sampleName group</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Lorena Pantano, Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Michael J. Steinbaugh, Lorena Pantano, Rory Kirchner, Victor Barrera, John Hutchinson, Harvard Chan Bioinformatics Core.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
