setClassUnion(name = "missingOrNULL", members = c("missing", "NULL"))



# bcbioRNASeq ==================================================================
# TODO Improve the object documentation here.
# TODO Consider adding countsFromAbundance check against tximport formals.
#' bcbio RNA-Seq Data Set
#'
#' `bcbioRNASeq` is an S4 class that extends `RangedSummarizedExperiment`, and
#' is designed to store a [bcbio](https://bcbio-nextgen.readthedocs.org) RNA-seq
#' analysis.
#'
#' @note `bcbioRNASeq` extended `SummarizedExperiment` prior to v0.2.0, where we
#'   migrated to `RangedSummarizedExperiment`.
#'
#' @section Automatic metadata:
#'
#' The [metadata()] slot always contains:
#'
#' - Object version.
#' - bcbio data provenance information.
#' - File paths and timestamps.
#' - R session information.
#'
#' @family S4 Classes
#' @author Michael Steinbaugh, Lorena Pantano
#' @export
#'
#' @seealso [bcbioRNASeq()].
setClass(Class = "bcbioRNASeq", contains = "RangedSummarizedExperiment")

setValidity(
    Class = "bcbioRNASeq",
    method = function(object) {
        # Return invalid for all objects older than v0.2.
        stopifnot(metadata(object)[["version"]] >= 0.2)

        assert_is_all_of(object, "RangedSummarizedExperiment")
        assert_has_dimnames(object)

        # Metadata -------------------------------------------------------------
        # Require the user to update to Bioconductor version.
        metadata <- metadata(object)

        # Detect legacy metrics.
        if (is.data.frame(metadata[["metrics"]])) {
            stop(paste(
                "`metrics` saved in `metadata()` instead of `colData()`.",
                updateMessage
            ))
        }

        # Check that interesting groups defined in metadata are valid.
        assert_is_subset(
            x = metadata[["interestingGroups"]],
            y = colnames(colData(object))
        )

        # Error on legacy slot detection.
        legacyMetadata <- c(
            "design",
            "devtoolsSessionInfo",
            "ensemblVersion",
            "gtf",
            "gtfFile",
            "missingGenes",
            "programs",
            "rowRangesMetadata",
            "template",
            "utilsSessionInfo",
            "yamlFile"
        )
        intersect <- intersect(names(metadata), legacyMetadata)
        if (length(intersect)) {
            stop(paste(
                paste(
                    "Legacy metadata slots:",
                    toString(sort(intersect))
                ),
                updateMessage,
                sep = "\n"
            ))
        }

        # Class checks (order independent).
        requiredMetadata <- list(
            allSamples = "logical",
            bcbioCommandsLog = "character",
            bcbioLog = "character",
            call = "call",
            caller = "character",
            countsFromAbundance = "character",
            dataVersions = "DataFrame",
            date = "Date",
            ensemblRelease = "integer",
            genomeBuild = "character",
            gffFile = "character",
            interestingGroups = "character",
            lanes = "integer",
            level = "character",
            organism = "character",
            programVersions = "DataFrame",
            projectDir = "character",
            runDate = "Date",
            sampleDirs = "character",
            sampleMetadataFile = "character",
            tx2gene = "Tx2Gene",
            sessionInfo = "session_info",
            uploadDir = "character",
            version = "package_version",
            wd = "character",
            yaml = "list"
        )
        classChecks <- invisible(mapply(
            name <- names(requiredMetadata),
            expected <- requiredMetadata,
            MoreArgs = list(metadata = metadata),
            FUN = function(name, expected, metadata) {
                actual <- class(metadata[[name]])
                if (!length(intersect(expected, actual))) {
                    FALSE
                } else {
                    TRUE
                }
            },
            SIMPLIFY = TRUE,
            USE.NAMES = TRUE
        ))
        if (!all(classChecks)) {
            stop(paste(
                "Metadata class checks failed.",
                updateMessage,
                printString(classChecks),
                sep = "\n"
            ))
        }

        # Additional assert checks.
        match.arg(
            arg = metadata[["caller"]],
            choices = validCallers
        )
        match.arg(
            arg = metadata[["level"]],
            choices = validLevels
        )
        if (is.character(metadata[["countsFromAbundance"]])) {
            match.arg(
                arg = metadata[["countsFromAbundance"]],
                choices = eval(formals(tximport)[["countsFromAbundance"]])
            )
        }

        # Assays ---------------------------------------------------------------
        assayNames <- assayNames(object)
        assert_is_subset(requiredAssays, assayNames)
        # Check that all assays are matrices.
        # Note that in previous versions, we slotted `DESeqDataSet` and
        # `DESeqTransform`, which can result in metadata mismatches because
        # those objects contain their own `colData()` and `rowData()`.
        assayCheck <- vapply(
            X = assays(object),
            FUN = is.matrix,
            FUN.VALUE = logical(1L),
            USE.NAMES = TRUE
        )
        if (!all(assayCheck)) {
            stop(paste(
                paste(
                    "Assays that are not matrix:",
                    toString(names(assayCheck[!assayCheck]))
                ),
                updateMessage,
                sep = "\n"
            ))
        }

        # Caller-specific checks.
        caller <- metadata[["caller"]]
        if (caller %in% tximportCallers) {
            assert_is_subset(tximportAssays, assayNames)
        } else if (caller %in% featureCountsCallers) {
            assert_is_subset(featureCountsAssays, assayNames)
        }

        # Check for average transcript length matrix, if necessary.
        if (
            metadata[["caller"]] %in% tximportCallers &&
            metadata[["countsFromAbundance"]] == "no"
        ) {
            assert_is_subset("avgTxLength", assayNames)
        }

        # Row data -------------------------------------------------------------
        assert_is_all_of(rowRanges(object), "GRanges")
        assert_is_all_of(rowData(object), "DataFrame")

        # Column data ----------------------------------------------------------
        # Error on legacy metrics columns.
        assert_are_disjoint_sets(
            x = colnames(colData(object)),
            y = legacyMetricsCols
        )

        TRUE
    }
)



# DESeqAnalysis ================================================================
#' DESeq2 Analysis Container
#'
#' Class containing all elements generated during differential expression
#' analysis with DESeq2. This class is essentially a `list` with validity checks
#' to ensure `DESeqTransform` and `DESeqResults` correspond to the
#' `DESeqDataSet`.
#'
#' @section DESeqDataSet:
#'
#' We recommend generating the `DESeqDataSet` by coercion from `bcbioRNASeq`
#' object using `as(dds, "bcbioRNASeq")`. Don't use the [DESeq2::DESeqDataSet()]
#' or [DESeq2::DESeqDataSetFromMatrix()] constructors to generate the
#' `DESeqDataSet` object.
#'
#' @section DESeqResults:
#'
#' Don't modify any of the `DESeqResults` objects manually. This includes
#' rearranging the rows or dropping genes without adjusted P values. We'll take
#' care of this automatically in supported methods.
#'
#' @family S4 Classes
#' @author Michael Steinbaugh
#' @export
#'
#' @slot data `DESeqDataSet`.
#' @slot transform `DESeqTransform`.
#' @slot results `list`. One or more unshrunken `DESeqResults`.
#' @slot lfcShrink `list`. One or more shrunken `DESeqResults`.
#'
#' @seealso [DESeqAnalysis()].
setClass(
    Class = "DESeqAnalysis",
    slots = c(
        data = "DESeqDataSet",
        transform = "DESeqTransform",
        results = "list",
        lfcShrink = "list"
    ),
    prototype = list(
        lfcShrink = list()
    )
)

setValidity(
    Class = "DESeqAnalysis",
    method = function(object) {
        # Require valid dimnames for data set.
        assertHasValidDimnames(object@data)

        # Require gene-to-symbol mappings.
        assert_is_subset(
            x = c("geneID", "geneName"),
            y = colnames(rowData(object@data))
        )

        # DESeqDataSet and DESeqTransform must match.
        assert_are_identical(
            x = dimnames(object@data),
            y = dimnames(object@transform)
        )

        # DESeqDataSet and DESeqResults must match.
        invisible(lapply(
            X = object@results,
            FUN = function(res) {
                assert_are_identical(
                    x = rownames(res),
                    y = rownames(object@data)
                )
            }
        ))

        # DESeqResults and lfcShrink rownames must match, if shrinkage has been
        # calculated.
        if (length(object@lfcShrink) > 0L) {
            invisible(mapply(
                unshrunken = object@results,
                shrunken = object@lfcShrink,
                FUN = function(unshrunken, shrunken) {
                    assert_are_identical(
                        x = rownames(unshrunken),
                        y = rownames(shrunken)
                    )
                }
            ))
        }

        TRUE
    }
)



# DESeqResultsTables ===========================================================
#' DESeq2 Differential Expression Results Tables
#'
#' `DESeqResults` object with `DataFrame` subsets and file path information.
#'
#' @family S4 Classes
#' @author Michael Steinbaugh
#' @export
#'
#' @slot all `DESeqResults`. Original unmodified `DESeqResults`. Should contain
#'   all genes, including those with `NA` adjusted *P* values.
#' @slot deg `DataFrame`. Subset containing genes that pass adjusted *P* value
#'   and log2 fold change cutoffs.
#' @slot degUp `DataFrame`. Directional subset containing only upregulated
#'   genes.
#' @slot degDown `DataFrame`. Directional subset containing only downregulated
#'   genes.
#' @slot localFiles `list`. Local file paths.
#' @slot dropboxFiles `list`. Dropbox file paths.
#'
#' @seealso [DESeqResultsTables()].
setClass(
    Class = "DESeqResultsTables",
    slots = c(
        all = "DESeqResults",
        deg = "DataFrame",
        degUp = "DataFrame",
        degDown = "DataFrame",
        localFiles = "character",
        dropboxFiles = "list"
    ),
    # Consider setting an initialize method instead.
    prototype = list(
        localFiles = character(),
        dropboxFiles = list()
    )
)

setValidity(
    Class = "DESeqResultsTables",
    method = function(object) {
        assert_is_a_string(contrastName(object@all))
        assert_is_a_number(metadata(object@all)[["alpha"]])
        assert_is_a_number(metadata(object@all)[["lfcThreshold"]])
        assert_is_subset(
            x = rownames(object@degUp),
            y = rownames(object@all)
        )
        assert_is_subset(
            x = rownames(object@degDown),
            y = rownames(object@all)
        )
        assert_are_disjoint_sets(
            x = rownames(object@degUp),
            y = rownames(object@degDown)
        )

        # Require that the subsets match the `DESeqResults` summary.
        match <- removeNA(str_match(
            string = capture.output(summary(object@all)),
            pattern = "^LFC.*\\s\\:\\s([0-9]+).*"
        ))
        assert_are_identical(
            x = nrow(object@degUp),
            y = as.integer(match[1, 2])
        )
        assert_are_identical(
            x = nrow(object@degDown),
            y = as.integer(match[2, 2])
        )

        # Object can have either local or Dropbox file paths, but not both.
        stopifnot(!all(
            length(object@localFiles) > 0L,
            length(object@dropboxFiles) > 0L
        ))

        TRUE
    }
)
