#' bcbioRNASeq Object Class
#'
#' S4 class containing RNA-seq counts and metadata generated by
#' [bcbio](https://bcbio-nextgen.readthedocs.org/).
#'
#' `bcbioRNASeq` is a subclass of `SummarizedExperiment` designed to store
#' an RNA-seq analysis. This class contains raw read counts and transcripts per
#' million (TPM) generated by [tximport()].
#'
# Additional slots can be defined by setting the representation, if necessary.
# This is useful for including alternate count matrices (e.g. featureCounts).
#'
#' The `metadata` accessor contains:
#'
#' - Sample summary statistics (a.k.a. metrics).
#' - Ensembl annotations.
#' - Server run paths.
#' - R local environment information, including [sessionInfo()].
#'
#' @slot bcbio [SimpleList] containing additional bcbio run data with dimensions
#' that don't match the count matrix. This is currently used to store STAR
#' featureCounts aligned counts.
#'
#' @author Lorena Pantano, Michael Steinbaugh
#' @export
#'
#' @seealso
#' - [SummarizedExperiment::SummarizedExperiment].
#' - `.S4methods(class = "bcbioRNASeq")`.
#'
#' @examples
#' uploadDir <- system.file("extdata/bcbio", package = "bcbioRNASeq")
#'
#' # We're suppressing warnings about missing log files in our minimal
#' # dataset, which end up in the Travis CI logs. This isn't recommended
#' # for a typical bcbio run.
#' bcb <- suppressWarnings(loadRNASeq(uploadDir))
#' bcb
#'
#' # Check validity
#' validObject(bcb)
bcbioRNASeq <- setClass(
    "bcbioRNASeq",
    contains = "SummarizedExperiment",
    slots = c(bcbio = "SimpleList")
)



# Validity =====================================================================
setValidity("bcbioRNASeq", function(object) {
    updateMsg <- paste(
        "Run `updateObject()` to update your bcbioRNASeq object",
        "to the latest version."
    )

    # SummarizedExperiment internal structure
    assert_is_all_of(object, "SummarizedExperiment")
    assert_has_dimnames(object)
    stopifnot(all(vapply(
        X = c(
            "assays",
            "bcbio",
            "colData",
            "elementMetadata",
            "metadata",
            "NAMES"
        ),
        FUN = function(slot) {
            .hasSlot(object, slot)
        },
        FUN.VALUE = logical(1L)
    )))

    # Required count matrices
    # Note that `rlog` and `vst` DESeqTransform objects are optional
    assert_is_subset(
        c("raw", "normalized", "tpm", "tmm"),
        names(assays(object))
    )

    # Metadata
    metadata <- metadata(object)
    requiredMetadata <- list(
        "version" = "package_version",
        "uploadDir" = "character",
        "sampleDirs" = "character",
        "projectDir" = "character",
        "template" = "character",
        "runDate" = "Date",
        "interestingGroups" = "character",
        "organism" = "character",
        "genomeBuild" = "character",
        "ensemblVersion" = c("integer", "NULL"),
        "annotable" = c("data.frame", "NULL"),
        "tx2gene" = "data.frame",
        "lanes" = "integer",
        "yaml" = "list",
        "metrics" = "data.frame",
        "sampleMetadataFile" = c("character", "NULL"),
        "dataVersions" = "tbl_df",
        "programVersions" = "tbl_df",
        "bcbioLog" = "character",
        "bcbioCommandsLog" = "character",
        "allSamples" = "logical",
        "design" = "formula",
        # Allow user to set `Inf`, which isn't an integer
        "transformationLimit" = c("integer", "numeric"),
        "date" = "Date",
        "wd" = "character",
        "utilsSessionInfo" = "sessionInfo",
        "devtoolsSessionInfo" = "session_info",
        "unannotatedGenes" = c("character", "NULL")
    )

    # Inform the user about renamed metadata slots
    legacyMetadata <- c(
        "gtf",
        "missingGenes",
        "programs"
    )
    intersect <- intersect(names(metadata), legacyMetadata)
    if (length(intersect) > 0L) {
        abort(paste(
            paste(
                "Legacy metadata slots:",
                toString(sort(intersect))
            ),
            updateMsg,
            sep = "\n"
        ))
    }

    # Integrity check the classes of required metadata
    classChecks <- invisible(vapply(
        X = seq_along(requiredMetadata),
        FUN = function(a) {
            name <- names(requiredMetadata)[[a]]
            actual <- class(metadata[[name]])
            expected <- requiredMetadata[[a]]
            if (!length(intersect(expected, actual))) {
                warn(paste(
                    name, "is not", toString(expected)
                ))
                FALSE
            } else {
                TRUE
            }
        },
        FUN.VALUE = logical(1L),
        USE.NAMES = FALSE
    ))
    if (!all(classChecks)) {
        abort(paste("Metadata class checks failed.", updateMsg, sep = "\n"))
    }

    annotable <- metadata[["annotable"]]
    if (!is.null(annotable)) {
        assert_is_annotable(annotable)
    }
    tx2gene <- metadata[["tx2gene"]]
    assert_is_tx2gene(tx2gene)

    TRUE
})
