#' `bcbioRNASeq`
#'
#' S4 class containing RNA-seq counts and metadata generated by
#' [bcbio](https://bcbio-nextgen.readthedocs.org/).
#'
#' `bcbioRNASeq` is a subclass of `SummarizedExperiment` designed to store
#' an RNA-seq analysis. This class contains raw read counts and transcripts per
#' million (TPM) generated by [tximport()].
#'
# Additional slots can be defined by setting the representation, if necessary.
# This is useful for including alternate count matrices (e.g. featureCounts).
#'
#' The `metadata` accessor contains:
#'
#' - Sample summary statistics (a.k.a. metrics).
#' - Ensembl annotations.
#' - Server run paths.
#' - R local environment information, including [sessionInfo()].
#'
#' @slot bcbio [SimpleList] containing additional bcbio run data with dimensions
#' that don't match the count matrix. This is currently used to store STAR
#' featureCounts aligned counts.
#'
#' @author Lorena Pantano, Michael Steinbaugh
#' @export
#'
#' @seealso
#' - [SummarizedExperiment::SummarizedExperiment].
#' - `.S4methods(class = "bcbioRNASeq")`.
#'
#' @examples
#' uploadDir <- system.file("extdata/bcbio", package = "bcbioRNASeq")
#' bcb <- loadRNASeq(uploadDir)
bcbioRNASeq <- setClass(
    "bcbioRNASeq",
    contains = "SummarizedExperiment",
    slots = c(bcbio = "SimpleList")
)
setValidity("bcbioRNASeq", function(object) TRUE)



#' `bcbioRNADataSet`
#'
#' This class has been superceded by [bcbioRNASeq] and will be formally
#' deprecated in a future release.
#'
#' @author Lorena Pantano, Michael Steinbaugh
#' @keywords internal
#'
#' @slot callers [SimpleList] containing additional bcbio run data with
#'   dimensions that don't match the count matrix.
#'
#' @export
bcbioRNADataSet <- setClass(
    "bcbioRNADataSet",
    contains = "SummarizedExperiment",
    slots = c(callers = "SimpleList")
)
setValidity("bcbioRNADataSet", function(object) TRUE)



# DESeq2
setClassUnion(
    name = "DESeq",
    members = c("DESeqDataSet", "DESeqTransform")
)
