#' bcbioRNADataSet object accessors
#'
#' This method is used to access count matrices with differing dimensions from
#' the object. By default, we store quasi-aligned (a.k.a. lightweight) gene
#' expression counts generated by
#' [salmon](https://combine-lab.github.io/salmon/) into in a
#' [SummarizedExperiment::SummarizedExperiment] using [tximport()].
#'
#' Any additional transformations of those original counts, including [tpm()],
#' [rlog()], or [tmm()] are stored alongside the raw counts. Reads be accessed
#' using the `counts()` generic.
#'
#' Additional count matrices of interest include:
#'
#' - Aligned counts generated by [STAR](https://github.com/alexdobin/STAR) and
#'   [featureCounts](http://bioinf.wehi.edu.au/featureCounts/).
#' - Transcript-level gene expression, for
#'   [sleuth](http://pachterlab.github.io/sleuth) analysis.
#' - miRNA expression counts.
#'
#' Since these matrices don't necessarily contain the same number of columns and
#' rows as the counts imported with [tximport()], it is preferable to store this
#' information as alternative callers.
#'
#' @rdname bcbio
#' @docType methods
#'
#' @param object Object.
#' @param type Type of count data to retrieve.
#' @param value An integer matrix or other object.
#' @param ... Additional arguments.
#'
#' @return Count matrix.
#' @export
setMethod("bcbio", "bcbioRNADataSet", function(object, type = "counts") {
    if (type == "counts") {
        assays(object)[["counts"]]
    } else if (type %in% names(slot(object, "callers"))) {
        slot(object, "callers")[[type]]
    } else {
        stop(paste(type, "not found"))
    }
})



#' @rdname bcbio
#' @export
setMethod(
    "bcbio<-",
    signature(object = "bcbioRNADataSet", value = "ANY"),
    function(object, type = "counts", value) {
        if (type == "counts") {
            assays(object)[["counts"]] <- value
            validObject(object)
        } else {
            slot(object, "callers")[[type]] <- value
        }
        object
    })
