#' Differential Expression Results Tables
#'
#' @rdname resultsTables
#' @name resultsTables
#' @author Michael Steinbaugh
#'
#' @inheritParams AllGenerics
#'
#' @param lfc Log fold change ratio (base 2) cutoff. Does not apply to
#'   statistical hypothesis testing, only gene filtering in the results tables.
#'   See [results()] for additional information about using `lfcThreshold` and
#'   `altHypothesis` to set an alternative hypothesis based on expected fold
#'   changes.
#' @param annotable Join Ensembl gene annotations to the results. Apply gene
#'   identifier to symbol mappings. If `TRUE`, the function will attempt to
#'   detect the organism from the gene identifiers (rownames; "ensgene" column)
#'   and automatically obtain the latest annotations from Ensembl using
#'   [annotable()]. If set `FALSE`/`NULL`, then gene annotations will
#'   not be added to the results. This is useful when working with a poorly
#'   annotated genome. Alternatively, a previously saved annotable [data.frame]
#'   can be passed in.
#' @param summary Show summary statistics.
#' @param headerLevel R Markdown header level. Applies only when
#'   `summary = TRUE`.
#' @param write Write CSV files to disk.
#' @param dir Local directory path where to write the results tables.
#' @param dropboxDir Dropbox directory path where to archive the results tables
#'   for permanent storage (e.g. Stem Cell Commons). When this option is
#'   enabled, unique links per file are generated internally with the rdrop2
#'   package.
#' @param rdsToken RDS file token to use for Dropbox authentication.
#' @param quiet If `TRUE`, suppress any status messages and/or progress bars.
#'
#' @return Results [list].
#'
#' @examples
#' load(system.file(
#'     file.path("extdata", "bcb.rda"),
#'     package = "bcbioRNASeq"))
#' load(system.file(
#'     file.path("extdata", "res.rda"),
#'     package = "bcbioRNASeq"))
#'
#' # Can use a stashed annotable for improved speed
#' annotable <- annotable(bcb)
#'
#' resTbl <- resultsTables(
#'     res,
#'     lfc = 0.25,
#'     annotable = annotable,
#'     summary = TRUE,
#'     headerLevel = 2L,
#'     write = TRUE,
#'     dir = "resultsTables",
#'     dropboxDir = file.path("bcbioRNASeq_examples", "resultsTables"),
#'     rdsToken = system.file("token.rds", package = "bcbioRNASeq")
#' )
#' names(resTbl)
#' dir("resultsTables")
#' unlink("resultsTables", recursive = TRUE)
NULL



# Constructors =================================================================
#' Markdown List of Results Files
#'
#' Enables looping of results contrast file links for RMarkdown.
#'
#' @author Michael Steinbaugh
#' @keywords internal
#'
#' @importFrom bcbioBase mdList
#'
#' @param resTbl List of results tables generated by [resultsTables()].
#' @param dir Output directory.
#'
#' @return [writeLines()].
#' @noRd
.mdResultsTables <- function(
    resTbl,
    headerLevel = 2L) {
    # Prioritze `dropboxFiles` over `localFiles` for path return
    if ("dropboxFiles" %in% names(resTbl)) {
        paths <- vapply(
            X = resTbl[["dropboxFiles"]],
            FUN = function(x) {
                x[["url"]]
            },
            FUN.VALUE = "character")
        basenames <- basename(paths) %>%
            gsub("\\?.*$", "", .)
    } else if ("localFiles" %in% names(resTbl)) {
        paths <- resTbl[["localFiles"]]
        basenames <- basename(paths)
    } else {
        abort("`resTbl` must contain `localFiles` or `dropboxFiles")
    }
    names(basenames) <- names(paths)

    mdHeader("Results tables", level = headerLevel, asis = TRUE)
    mdList(c(
        paste0(
            "[`",basenames[["all"]], "`]",
            "(", paths[["all"]], "): ",
            "All genes, sorted by Ensembl identifier."),
        paste0(
            "[`", basenames[["deg"]], "`]",
            "(", paths[["deg"]], "): ",
            "Genes that pass the alpha (FDR) cutoff."),
        paste0(
            "[`", basenames[["degLFCUp"]], "`]",
            "(", paths[["degLFCUp"]], "): ",
            "Upregulated DEG; positive log2 fold change."),
        paste0(
            "[`", basenames[["degLFCDown"]], "`]",
            "(", paths[["degLFCDown"]], "): ",
            "Downregulated DEG; negative log2 fold change.")
    ), asis = TRUE)
}



#' @importFrom bcbioBase annotable camel checkAnnotable copyToDropbox mdHeader
#'   mdList sanitizeAnnotable snake
#' @importFrom dplyr arrange desc left_join
#' @importFrom readr write_csv
#' @importFrom rlang !! sym
#' @importFrom S4Vectors metadata
#' @importFrom tibble rownames_to_column
.resultsTables.DESeqResults <- function(  # nolint
    object,
    lfc = 0L,
    annotable = TRUE,
    summary = TRUE,
    headerLevel = 2L,
    write = FALSE,
    dir = getwd(),
    dropboxDir = NULL,
    rdsToken = NA,
    quiet = FALSE) {
    # Parameter integrity checks ===============================================
    # lfc
    if (!(is.numeric(lfc) && length(lfc) == 1L)) {
        abort("`lfc` must be a numeric string")
    }
    # annotable
    if (!(is.logical(annotable) ||
          is.data.frame(annotable) ||
          is.null(annotable))) {
        abort("`annotable` must be a logical vector, data.frame, or NULL")
    }
    # summary
    if (!is.logical(summary)) {
        abort("`summary` must be a logical vector")
    }
    # headerLevel
    if (!(is.numeric(headerLevel) && length(headerLevel) == 1L)) {
        abort("`headerLevel` must be a numeric string")
    }
    # write
    if (!is.logical(write)) {
        abort("`write` must be a logical vector")
    }
    # dir
    if (!is_string(dir)) {
        abort("`dir` must be a string")
    }
    # dropboxDir
    if (!(is_string(dropboxDir) || is.null(dropboxDir))) {
        abort("`dropboxDir` must be a string or NULL")
    }
    # quiet
    if (!is.logical(quiet)) {
        abort("`quiet` must be a logical vector")
    }

    # Extract internal parameters from DESeqResults object =====================
    contrast <- .resContrastName(object)
    fileStem <- snake(contrast)
    # Alpha level, slotted in `DESeqResults` metadata
    alpha <- metadata(object)[["alpha"]]

    # Prepare the results tables ===============================================
    all <- object %>%
        as.data.frame() %>%
        rownames_to_column("ensgene") %>%
        as("tibble") %>%
        camel(strict = FALSE) %>%
        arrange(!!sym("ensgene"))

    # Add Ensembl gene annotations (annotable), if desired
    if (isTRUE(annotable)) {
        # Match genome against the first gene identifier by default
        organism <- rownames(object) %>%
            .[[1L]] %>%
            detectOrganism()
        annotable <- annotable(organism, quiet = quiet)
    }
    if (is.data.frame(annotable)) {
        checkAnnotable(annotable)
        # Drop the nested lists (e.g. entrez), otherwise the CSVs will fail to
        # save when `write = TRUE`.
        annotable <- sanitizeAnnotable(annotable)
        all <- left_join(all, annotable, by = "ensgene")
    }

    # Check for overall gene expression with base mean
    baseMeanGt0 <- all %>%
        arrange(desc(!!sym("baseMean"))) %>%
        .[.[["baseMean"]] > 0L, , drop = FALSE]
    baseMeanGt1 <- baseMeanGt0 %>%
        .[.[["baseMean"]] > 1L, , drop = FALSE]

    # All DEG tables are sorted by BH adjusted P value
    deg <- all %>%
        .[!is.na(.[["padj"]]), , drop = FALSE] %>%
        .[.[["padj"]] < alpha, , drop = FALSE] %>%
        arrange(!!sym("padj"))
    degLFC <- deg %>%
        .[.[["log2FoldChange"]] > lfc |
            .[["log2FoldChange"]] < -lfc, , drop = FALSE]
    degLFCUp <- degLFC %>%
        .[.[["log2FoldChange"]] > 0L, , drop = FALSE]
    degLFCDown <- degLFC %>%
        .[.[["log2FoldChange"]] < 0L, , drop = FALSE]

    resTbl <- list(
        "contrast" = contrast,
        # Cutoffs
        "alpha" = alpha,
        "lfc" = lfc,
        # Tibbles
        "all" = all,
        "deg" = deg,
        "degLFC" = degLFC,
        "degLFCUp" = degLFCUp,
        "degLFCDown" = degLFCDown)

    if (isTRUE(summary)) {
        mdHeader("Summary statistics", level = headerLevel, asis = TRUE)
        mdList(c(
            paste(nrow(all), "genes in counts matrix"),
            paste("Base mean > 0:", nrow(baseMeanGt0), "genes (non-zero)"),
            paste("Base mean > 1:", nrow(baseMeanGt1), "genes"),
            paste("Alpha cutoff:", alpha),
            paste("LFC cutoff:", lfc, "(applied in tables only)"),
            paste("DEG pass alpha:", nrow(deg), "genes"),
            paste("DEG LFC up:", nrow(degLFCUp), "genes"),
            paste("DEG LFC down:", nrow(degLFCDown), "genes")
        ), asis = TRUE)
    }

    if (isTRUE(write)) {
        tibbles <- c("all", "deg", "degLFCUp", "degLFCDown")

        # Local files (required) ===============================================
        # Standardize the output directory
        dir.create(dir, recursive = TRUE, showWarnings = FALSE)
        dir <- normalizePath(dir)

        # Local file paths
        localFiles <- file.path(
            dir,
            paste0(fileStem, "_", snake(tibbles), ".csv.gz")
        )
        names(localFiles) <- tibbles

        # Write the results tibbles to local directory
        invisible(lapply(
            X = seq_along(localFiles),
            FUN = function(a) {
                write_csv(
                    x = get(tibbles[[a]]),
                    path = localFiles[[a]]
                )
            }
        ))

        # Check that writes were successful
        if (!all(vapply(localFiles, file.exists, logical(1L)))) {
            abort("Results table CSVs failed to write to disk")
        }

        # Update the resTbl list with the file paths
        resTbl[["localFiles"]] <- localFiles

        # Copy to Dropbox (optional) ===========================================
        if (is.character(dropboxDir)) {
            dropboxFiles <- copyToDropbox(
                files = localFiles,
                dir = dropboxDir,
                rdsToken = rdsToken)
            resTbl[["dropboxFiles"]] <- dropboxFiles
        }

        # Output file information in Markdown format
        .mdResultsTables(resTbl, headerLevel = headerLevel)
    }

    resTbl
}



# Methods ======================================================================
#' @rdname resultsTables
#' @export
setMethod(
    "resultsTables",
    signature("DESeqResults"),
    .resultsTables.DESeqResults)
