# TODO Deprecate `gene2symbol` in favor of stashed `rowRanges`



#' Differentially Expressed Gene Heatmap
#'
#' This function is a simplified version of [plotHeatmap()] that is
#' optimized for handling a `DESeqResults` object rather a gene vector. All of
#' the optional parameters for [plotHeatmap()] are also available to this
#' function.
#'
#' To adjust the annotation columns, modify the `colData` of the `counts`
#' argument, which must contain a `SummarizedExperiment` (e.g. `DESeqTransform`,
#' `DESeqDataSet`).
#'
#' @name plotDEGHeatmap
#' @family Differential Expression Functions
#' @author Michael Steinbaugh
#'
#' @inherit bcbioBase::plotHeatmap
#'
#' @inheritParams general
#' @param results `DESeqResults` object.
#' @param ... Passthrough arguments to [plotHeatmap()] `SummarizedExperiment`
#'   method.
#'
#' @seealso
#' - `help("plotHeatmap", "bcbioBase")`.
#' - `findMethod("plotHeatmap", "SummarizedExperiment")`.
#'
#' @examples
#' # DESeqResults, matrix ====
#' counts <- assay(rld_small)
#' gene2symbol <- gene2symbol(rld_small)
#' plotDEGHeatmap(
#'     results = res_small,
#'     counts = counts,
#'     gene2symbol = gene2symbol
#' )
#'
#' # DESeqResults, SummarizedExperiment ====
#' plotDEGHeatmap(
#'     results = res_small,
#'     counts = rld_small
#' )
#'
#' # DESeqResults, DESeqDataSet ====
#' # This always uses normalized counts
#' # Using default ggplot2 colors
#' plotDEGHeatmap(
#'     results = res_small,
#'     counts = dds_small,
#'     color = NULL,
#'     legendColor = NULL
#' )
#'
#' # DESeqResults, bcbioRNASeq ====
#' plotDEGHeatmap(
#'     results = res_small,
#'     counts = bcb_small
#' )
NULL



# Methods ======================================================================
#' @rdname plotDEGHeatmap
#' @export
setMethod(
    "plotDEGHeatmap",
    signature(
        results = "DESeqResults",
        counts = "matrix"
    ),
    function(
        results,
        counts,
        lfc = 0L,
        gene2symbol = NULL,
        title = TRUE,
        ...
    ) {
        validObject(results)
        assert_are_identical(rownames(results), rownames(counts))
        assert_is_a_number(lfc)
        assert_all_are_non_negative(lfc)

        # Title
        if (isTRUE(title)) {
            title <- contrastName(results)
        } else if (!is_a_string(title)) {
            title <- NULL
        }

        # Alpha level
        alpha <- metadata(results)[["alpha"]]
        assert_is_a_number(alpha)

        data <- results %>%
            as.data.frame() %>%
            camel() %>%
            # Keep genes that pass alpha cutoff
            .[!is.na(.[["padj"]]), , drop = FALSE] %>%
            .[.[["padj"]] < alpha, , drop = FALSE] %>%
            # Keep genes that pass log2 fold change cutoff
            .[!is.na(.[["log2FoldChange"]]), , drop = FALSE] %>%
            .[.[["log2FoldChange"]] > lfc |
                  .[["log2FoldChange"]] < -lfc, , drop = FALSE]
        assert_has_rows(data)

        # Subset the counts matrix to only contain DEGs
        deg <- rownames(data)
        counts <- counts[deg, , drop = FALSE]

        # Remap gene identifier rows to symbols
        if (is.data.frame(gene2symbol)) {
            rownames <- convertGenesToSymbols(
                object = rownames(counts),
                gene2symbol = gene2symbol
            )
            rownames(counts) <- rownames
        }

        plotHeatmap(
            object = counts,
            title = title,
            ...
        )
    }
)



#' @rdname plotDEGHeatmap
#' @export
setMethod(
    "plotDEGHeatmap",
    signature(
        results = "DESeqResults",
        counts = "SummarizedExperiment"
    ),
    function(results, counts, ...) {
        validObject(counts)
        gene2symbol <- gene2symbol(counts)
        counts <- assay(counts)
        plotDEGHeatmap(
            results = results,
            counts = assay(counts),
            gene2symbol = gene2symbol,
            ...
        )
    }
)



#' @rdname plotDEGHeatmap
#' @export
setMethod(
    "plotDEGHeatmap",
    signature(
        results = "DESeqResults",
        counts = "DESeqDataSet"
    ),
    function(results, counts, ...) {
        validObject(counts)
        gene2symbol <- gene2symbol(counts)
        counts <- counts(counts, normalized = TRUE)
        plotDEGHeatmap(
            results = results,
            counts = counts,
            gene2symbol = gene2symbol,
            ...
        )
    }
)



#' @rdname plotDEGHeatmap
#' @export
setMethod(
    "plotDEGHeatmap",
    signature(
        results = "DESeqResults",
        counts = "bcbioRNASeq"
    ),
    function(
        results,
        counts,
        normalized = "rlog",
        ...
    ) {
        validObject(counts)
        gene2symbol <- gene2symbol(counts)
        counts <- counts(counts, normalized = normalized)
        plotDEGHeatmap(
            results = results,
            counts = counts,
            gene2symbol = gene2symbol,
            ...
        )
    }
)
