#' Top Tables of Differential Expression Results
#'
#' @rdname topTables
#' @name topTables
#' @author Michael Steinbaugh
#'
#' @param object DESeq2 results tables list generated by [resultsTables()].
#' @param n Number genes to report.
#' @param coding Whether to only return coding genes.
#'
#' @return Top table kables, for knit report.
#'
#' @examples
#' load(system.file(
#'     file.path("inst", "extdata", "bcb.rda"),
#'     package = "bcbioRNASeq"))
#' load(system.file(
#'     file.path("inst", "extdata", "res.rda"),
#'     package = "bcbioRNASeq"))
#'
#' annotable <- annotable(bcb)
#' resTbl <- resultsTables(
#'     res,
#'     annotable = annotable,
#'     summary = FALSE,
#'     write = FALSE)
#' topTables(resTbl)
NULL



# Constructors ====
#' @importFrom basejump fixNA
#' @importFrom dplyr filter mutate rename
#' @importFrom S4Vectors head
#' @importFrom tibble remove_rownames
.subsetTop <- function(
    df,
    n = 50,
    coding = FALSE) {
    if (isTRUE(coding)) {
        if (!"broadClass" %in% colnames(df)) {
            stop("'coding = TRUE' argument requires 'broadClass' column",
                 call. = FALSE)
        }
        df <- df %>%
            .[.[["broadClass"]] == "coding", , drop = FALSE]
    }
    if (!nrow(df) | !is.data.frame(df)) {
        return(NULL)
    }
    keepCols <- c(
        "ensgene",
        "baseMean",
        "lfc",
        "padj",
        "symbol",
        "description")
    df %>%
        head(n = n) %>%
        rename(lfc = .data[["log2FoldChange"]]) %>%
        mutate(
            baseMean = round(.data[["baseMean"]]),
            lfc = format(.data[["lfc"]], digits = 3),
            padj = format(.data[["padj"]], digits = 3, scientific = TRUE),
            # Remove symbol information in description, if present
            description = gsub(
                x = .data[["description"]],
                pattern = " \\[.+\\]$",
                replacement = "")
        ) %>%
        .[, which(colnames(.) %in% keepCols)] %>%
        remove_rownames() %>%
        fixNA()
}



# Methods ====
#' @rdname topTables
#' @importFrom knitr kable
#' @export
setMethod(
    "topTables",
    signature("list"),
    function(
        object,
        n = 50,
        coding = FALSE) {
        up <- .subsetTop(
            object[["degLFCUp"]],
            n = n,
            coding = coding)
        down <- .subsetTop(
            object[["degLFCDown"]],
            n = n,
            coding = coding)
        contrastName <- object[["contrast"]]
        if (!is.null(up)) {
            show(kable(
                up,
                caption = paste(contrastName, "(upregulated)")
            ))
        }
        if (!is.null(down)) {
            show(kable(
                down,
                caption = paste(contrastName, "(downregulated)")
            ))
        }
    })
