#' Plot Counts Per Gene
#'
#' Generally, we expect similar count spreads for all genes between samples
#' unless the library sizes or total RNA expression are different.
#'
#' @section TMM:
#' We recommend visualizing counts normalized with the Trimmed Mean of M-Values
#' (TMM) method here (Robinson, et al., 2010). TMM normalization equates the
#' overall expression levels of genes between samples under the assumption that
#' the majority of them are not differentially expressed. Therefore, by
#' normalizing for total RNA expression by sample, we expect the spread of the
#' TMM-normalized counts per gene to be similar for every sample.
#'
#' @name plotCountsPerGene
#' @family Quality Control Functions
#' @author Michael Steinbaugh, Rory Kirchner, Victor Barrera
#' @importFrom basejump plotCountsPerGene
#' @inherit basejump::plotCountsPerGene
#' @export
#'
#' @inheritParams general
#'
#' @examples
#' plotCountsPerGene(bcb_small)
NULL



.plotCountsPerGene.bcbioRNASeq <-  # nolint
    function(
        object,
        normalized = c("tmm", "vst", "rlog", "tpm", "rle")
    ) {
        validObject(object)
        normalized <- match.arg(normalized)

        # Coerce to RSE.
        rse <- as(object, "RangedSummarizedExperiment")
        counts <- counts(object, normalized = normalized)
        assays(rse) <- list(counts)
        assayNames(rse) <- normalized

        # Set the counts axis label.
        countsAxisLabel <- paste(normalized, "counts")
        trans <- .normalizedTrans(normalized)
        if (trans != "identity") {
            countsAxisLabel <- paste(trans, countsAxisLabel)
        }

        do.call(
            what = plotCountsPerGene,
            args = matchArgsToDoCall(
                args = list(
                    object = rse,
                    trans = trans,
                    countsAxisLabel = countsAxisLabel
                ),
                removeFormals = "normalized"
            )
        )
    }
f1 <- formals(.plotCountsPerGene.bcbioRNASeq)
f2 <- methodFormals("plotCountsPerGene", "SummarizedExperiment")
f2 <- f2[setdiff(names(f2), c(names(f1), "assay"))]
f <- c(f1, f2)
formals(.plotCountsPerGene.bcbioRNASeq) <- f



#' @rdname plotCountsPerGene
#' @export
setMethod(
    f = "plotCountsPerGene",
    signature = signature("bcbioRNASeq"),
    definition = .plotCountsPerGene.bcbioRNASeq
)
