#' Print summary statistics of alpha level cutoffs
#'
#' @author Michael Steinbaugh
#'
#' @param dds \linkS4class{DESeqDataSet}.
#' @param alpha Numeric vector of desired alpha cutoffs.
#' @param ... Passthrough parameters to [DESeq2::results()].
#'
#' @return Printed summary.
#' @export
alpha_summary <- function(
    dds,
    alpha = c(0.1, 0.05, 0.01, 1e-3, 1e-6),
    ...) {
    name <- deparse(substitute(dds))
    print(name)
    lapply(seq_along(alpha), function(a) {
        writeLines(
            paste(name,
                  paste("alpha", alpha[a], sep = " = "),
                  sep = " : "))
        results(dds, alpha = alpha[a], ...) %>% summary
    }) %>% invisible
}



#' Metadata table
#'
#' Returns a subset of metadata columns of interest used for knit reports. These
#' "interesting group" columns are defined as `intgroup` in the bcbio-nextgen
#' run object.
#'
#' @author Michael Steinbaugh
#'
#' @param run bcbio-nextgen run.
#'
#' @return Data frame containing only the columns of interest.
#' @export
metadata_table <- function(run) {
    run$metadata %>%
        remove_rownames %>%
        kable(caption = "Sample metadata")
}



#' Differential expression results tables
#'
#' @author Michael Steinbaugh
#'
#' @param run bcbio-nextgen run.
#' @param res \linkS4class{DESeqResults}.
#' @param lfc Log fold change ratio (base 2) cutoff. Does not apply to
#'   statistical hypothesis testing, only gene filtering in the results tables.
#'   See [DESeq2::results()] for additional information about using
#'   `lfcThreshold` and `altHypothesis` to set an alternative hypothesis based
#'   on expected fold changes.
#'
#' @return Results list.
#' @export
#'
#' @examples
#' \dontrun{
#' res_tables(run, res, lfc = 0.25)
#' }
res_tables <- function(
    run,
    res,
    lfc = 0) {
    check_run(run)
    check_res(res)
    import_tidy_verbs()

    name <- deparse(substitute(res))
    contrast <- res_contrast_name(res)
    contrast_name <- contrast %>% gsub(" ", "_", .)
    alpha <- res@metadata$alpha
    annotations <- gene_level_annotations(run)

    all <- res %>%
        as.data.frame %>%
        rownames_to_column("ensembl_gene_id") %>%
        as_tibble %>%
        snake %>%
        left_join(annotations, by = "ensembl_gene_id") %>%
        arrange(!!sym("ensembl_gene_id"))

    # Check for overall gene expression with base mean
    base_mean_gt0 <- all %>%
        arrange(desc(!!sym("base_mean"))) %>%
        filter(.data$base_mean > 0)
    base_mean_gt1 <- base_mean_gt0 %>%
        filter(.data$base_mean > 1)

    # All DEG tables are sorted by BH adjusted P value
    deg <- all %>%
        filter(.data$padj < alpha) %>%
        arrange(!!sym("padj"))
    deg_lfc <- deg %>%
        .[which(.$log2_fold_change > lfc | .$log2_fold_change < -lfc), ]
    deg_lfc_up <- deg_lfc %>%
        filter(.data$log2_fold_change > 0)
    deg_lfc_down <- deg_lfc %>%
        filter(.data$log2_fold_change < 0)

    writeLines(c(
        paste(name, "differential expression tables"),
        paste(nrow(all), "gene annotations"),
        "cutoffs applied",
        paste("    alpha:", alpha),
        paste("    lfc:  ", lfc, "(tables only)"),
        "base mean",
        paste("    > 0:", nrow(base_mean_gt0), "genes (detected)"),
        paste("    > 1:", nrow(base_mean_gt1), "genes"),
        "pass cutoffs",
        paste("    alpha:     ", nrow(deg), "genes"),
        paste("    + lfc up:  ", nrow(deg_lfc_up), "genes"),
        paste("    + lfc down:", nrow(deg_lfc_down), "genes"),
        ""
    ))

    # Write the CSV files to `results/`
    de_dir <- file.path("results", "de")
    dir.create(de_dir, recursive = TRUE, showWarnings = FALSE)

    write_csv(
        all,
        file.path(de_dir,
                  paste(name,
                        contrast_name, "all_genes.csv.gz",
                        sep = "_")))
    write_csv(
        deg,
        file.path(de_dir,
                  paste(name,
                        contrast_name,
                        "deg.csv.gz",
                        sep = "_")))
    write_csv(
        deg_lfc_up,
        file.path(de_dir,
                  paste(name,
                        contrast_name,
                        "deg_lfc_up.csv.gz",
                        sep = "_")))
    write_csv(
        deg_lfc_down,
        file.path(de_dir,
                  paste(name,
                        contrast_name,
                        "deg_lfc_down.csv.gz",
                        sep = "_")))

    list(name = name,
         contrast = contrast,
         alpha = alpha,
         lfc = lfc,
         all = all,
         deg = deg,
         deg_lfc = deg_lfc,
         deg_lfc_up = deg_lfc_up,
         deg_lfc_down = deg_lfc_down,
         res = res)
}



#' Top tables of differential expression results
#'
#' @author Michael Steinbaugh
#'
#' @param res_tbl Results tables generated by \code{\link{res_tables}}.
#' @param n Number genes to report.
#'
#' @return Top tables list.
#' @export
top_tables <- function(res_tbl, n = 50) {
    subset_top <- function(df) {
        discard <- c("description", "lfc_se", "pvalue", "stat")
        df <- df[, setdiff(colnames(df), discard)] %>%
            remove_rownames %>%
            # [top_n()] defaults using last column in tbl
            top_n(n = 50)
        # Handle digits
        df$base_mean <- round(df$base_mean)
        df$log2_fold_change <- format(df$log2_fold_change, digits = 3)
        df$padj <- format(df$padj, digits = 3, scientific = TRUE)
        df
    }

    up <- subset_top(res_tbl$deg_lfc_up)
    down <- subset_top(res_tbl$deg_lfc_down)

    # Captions
    name <- res_tbl$name
    contrast <- res_tbl$contrast
    sep <- " : "
    name_prefix <- paste(name, contrast, sep = sep)

    show(kable(up, caption = paste(name_prefix, "upregulated", sep = sep)))
    show(kable(down, caption = paste(name_prefix, "downregulated", sep = sep)))
}
